From d9b118162ddadfb0b16f6e1888db71004d361a6b Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 16 Jan 2023 09:04:54 +0200
Subject: [PATCH] show pll start print

Change-Id: I9b6ff994a6f96b20460e2e4a67c6769bf1594281
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/drivers/clock_control/clock_control_hl_pll.c b/drivers/clock_control/clock_control_hl_pll.c
index 7fd5c73..cd00e7c 100644
--- a/drivers/clock_control/clock_control_hl_pll.c
+++ b/drivers/clock_control/clock_control_hl_pll.c
@@ -956,7 +956,7 @@
 	int32_t ret = 0;
 	uint32_t idx;
 
-	HL_PLL_DBG("%s PLL init started", DEV_CFG(dev)->name);
+	HL_PLL_INF("%s", DEV_CFG(dev)->name);
 
 	/* Single die by default */
 	data->num_dies = 1;
