From 35771753a5399e0eae7f0494980ef58c0b9d51a6 Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Tue, 23 Jul 2024 12:58:53 +0300
Subject: [PATCH] [SW-0] WIP: reorder single arch

Change-Id: Id5e443adaeafb5591d7d0d44dfde7c501269658f
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/scripts/make/hl-common/Makefile b/scripts/make/hl-common/Makefile
index 6cd9c46..5272c9c 100644
--- a/scripts/make/hl-common/Makefile
+++ b/scripts/make/hl-common/Makefile
@@ -1,7 +1,7 @@
-HL_MULTI_ARCH_RELEASE := gaudi2 gaudi3 fs1
+HL_MULTI_ARCH_RELEASE := gaudi2 gaudi3
 HL_DEFAULT_DEV_ARCH := gaudi2_arm gaudi3_arc fs1_arc
 HL_SINGLE_ARCH_RELEASE :=
-HL_SECURITY_RELEASE :=
+HL_SECURITY_RELEASE := fs1
 
 RELEASE_JOB_FAKE_SEC := "fake_security=true"
 RELEASE_JOB_RUN_TEST_BIN := "run_test_on_bin=false"
