From 7f570612da0f801b704cfa0947d0859435a78f07 Mon Sep 17 00:00:00 2001
From: Tamal <tasaha@habana.ai>
Date: Sun, 25 Aug 2024 15:43:35 +0300
Subject: [PATCH] [SW-0]: fs1: io die binning test cases

Change-Id: I701afad50c99e31aaf976a5b278a512c38b9e270
Signed-off-by: Tamal <tasaha@habana.ai>
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index b0157e4..f94d74a 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -326,6 +326,8 @@
 		return -1;
 	}
 
+	io_die_binning_validation_step_1();
+
 	ret = fs1_redundancy_ctx_init(masks, die_id);
 
 	if (ret) {
@@ -339,6 +341,8 @@
 		return 1;
 	}
 
+	io_die_binning_validation_step_2();
+
 	/*
 	 * disable isolation to all IO0 engines.
 	 * TODO: adjust in respect to binning requirements
