From 9d243ff822513d2743886246570dd652d3772997 Mon Sep 17 00:00:00 2001
From: Sujoy Saraswati <ssaraswati@habana.ai>
Date: Wed, 04 Sep 2024 17:42:46 +0300
Subject: [PATCH] [Not for merge] tracing inside functions

Change-Id: Iac312116b66c5887d0871e7e55c397d5bbe8aa9e
---

diff --git a/backend/kernel/hpu_habana_cache.cpp b/backend/kernel/hpu_habana_cache.cpp
index d926ebc..20c1dac 100644
--- a/backend/kernel/hpu_habana_cache.cpp
+++ b/backend/kernel/hpu_habana_cache.cpp
@@ -1510,12 +1510,15 @@
     // it from being deallocated while the operation is still in flight.
     std::vector<synapse_helpers::device_ptr> inDevPtr;
     inDevPtr.reserve(num_inputs_);
-    for (auto& input : input_refs) {
-      if (input.isTensor()) {
-        at::Tensor tensor = input.toTensor();
-        ptRefs.push_back(std::move(tensor));
-        inDevPtr.push_back(reinterpret_cast<synapse_helpers::device_ptr>(
-            input.toTensor().storage().data_ptr().get()));
+    {
+      PT_EXTRA_TRACE("RecipeLauncher::Launch:input_tensor_loop");
+      for (auto& input : input_refs) {
+        if (input.isTensor()) {
+          at::Tensor tensor = input.toTensor();
+          ptRefs.push_back(std::move(tensor));
+          inDevPtr.push_back(reinterpret_cast<synapse_helpers::device_ptr>(
+              input.toTensor().storage().data_ptr().get()));
+        }
       }
     }
     if (dma_inputs.size() > 0) {
@@ -1616,128 +1619,132 @@
         PT_BRIDGE_DEBUG("Skipping recipe launch. empty recipe");
       }
     }
-    recipe_counter.increase();
+    {
+      PT_EXTRA_TRACE("Post launch");
+      recipe_counter.increase();
 
-    // register events for external tensors on compute
-    for (size_t i = 0; i < ext_events.size(); ++i) {
-      device.register_producer_on_stream(stream_handle, ext_events.at(i));
+      // register events for external tensors on compute
+      for (size_t i = 0; i < ext_events.size(); ++i) {
+        device.register_producer_on_stream(stream_handle, ext_events.at(i));
+      }
+
+      if (!(common::IsRecordStreamEnabled() &&
+            GET_ENV_FLAG_NEW(PT_HPU_USE_LAUNCH_RECORD_STREAM))) {
+        // Use wrapper for resources that must survive async part of the
+        // compute.
+        struct ResourceHolder {
+          std::shared_ptr<synapse_helpers::graph::recipe_handle> recipe_id_;
+          std::vector<at::Tensor> output_tensors_;
+          std::vector<at::Tensor> input_tensors_;
+          std::unique_ptr<synapse_helpers::device_ptr_lock> address_lock;
+          synapse_helpers::active_recipe_counter* recipe_counter_ptr;
+          size_t active_graph_key_;
+        };
+        auto resource_holder = std::shared_ptr<ResourceHolder>(
+            new ResourceHolder(), [](ResourceHolder* resource_holder) {
+              auto recipe_counter_ptr = resource_holder->recipe_counter_ptr;
+              auto recipe_handle = resource_holder->recipe_id_;
+              delete resource_holder;
+              recipe_counter_ptr->decrease_and_notify();
+              if (synapse_helpers::memory_reporter_enable() &&
+                  resource_holder->active_graph_key_ > 0) {
+                auto& device = HPURegistrar::get_device();
+                synapse_helpers::MemoryReporter* reporter =
+                    device.get_device_memory().get_memory_reporter();
+                reporter->getGraphStats()->removeLiveGraph(
+                    resource_holder->active_graph_key_);
+              }
+              towl::emitRecipeFinished(recipe_handle.get());
+              PT_LAZY_DEBUG("call decrease and notify of recipe_counter");
+            });
+        // recipe_id_ needs to be passed to done_cb to ensure its lifetime until
+        // corresponding recipe is finished on stream
+        const auto& recipe_ptr = recipe_;
+        resource_holder->recipe_id_ = recipe_ptr;
+        if (not common::IsRecordStreamNoHolderEnabled()) {
+          resource_holder->output_tensors_ = outPtRefs;
+          resource_holder->address_lock = std::move(address_lock);
+          resource_holder->input_tensors_ = ptRefs;
+        }
+        resource_holder->recipe_counter_ptr = &recipe_counter;
+        resource_holder->active_graph_key_ = active_graph_key_;
+        // ResourceHolder could be used directly as callback, if we would only
+        // implement operator(), but copying of ResourceHolder would result in
+        // copying of all shared_ptr stored inside (including std::vector). To
+        // make sharing more lightweight we hide ResourceHolder behind one
+        // shared_ptr. This indirection allows us to maintain only one shared
+        // reference.
+        auto cleanup_callback = [resource_holder]() mutable {
+          resource_holder.reset();
+        };
+
+        device.register_producer_on_stream(
+            std::move(outDevPtr), stream_handle, cleanup_callback);
+        // Launch collective ops
+        collective_kernels_info_->Launch(true, cleanup_callback);
+      } else {
+        // Use wrapper for resources that must survive async part of the
+        // compute.
+        struct ResourceHolder {
+          std::shared_ptr<synapse_helpers::graph::recipe_handle> recipe_id_;
+          std::unique_ptr<synapse_helpers::device_ptr_lock> address_lock;
+          synapse_helpers::active_recipe_counter* recipe_counter_ptr;
+          size_t active_graph_key_;
+        };
+        auto resource_holder = std::shared_ptr<ResourceHolder>(
+            new ResourceHolder(), [](ResourceHolder* resource_holder) {
+              auto recipe_counter_ptr = resource_holder->recipe_counter_ptr;
+              auto recipe_handle = resource_holder->recipe_id_;
+              delete resource_holder;
+              recipe_counter_ptr->decrease_and_notify();
+              if (synapse_helpers::memory_reporter_enable() &&
+                  resource_holder->active_graph_key_ > 0) {
+                auto& device = HPURegistrar::get_device();
+                synapse_helpers::MemoryReporter* reporter =
+                    device.get_device_memory().get_memory_reporter();
+                reporter->getGraphStats()->removeLiveGraph(
+                    resource_holder->active_graph_key_);
+              }
+              towl::emitRecipeFinished(recipe_handle.get());
+              PT_LAZY_DEBUG("call decrease and notify of recipe_counter");
+            });
+        // recipe_id_ needs to be passed to done_cb to ensure its lifetime until
+        // corresponding recipe is finished on stream
+        const auto& recipe_ptr = recipe_;
+        resource_holder->recipe_id_ = recipe_ptr;
+        if (not common::IsRecordStreamNoHolderEnabled()) {
+          resource_holder->address_lock = std::move(address_lock);
+        }
+        resource_holder->recipe_counter_ptr = &recipe_counter;
+        resource_holder->active_graph_key_ = active_graph_key_;
+        // ResourceHolder could be used directly as callback, if we would only
+        // implement operator(), but copying of ResourceHolder would result in
+        // copying of all shared_ptr stored inside (including std::vector). To
+        // make sharing more lightweight we hide ResourceHolder behind one
+        // shared_ptr. This indirection allows us to maintain only one shared
+        // reference.
+        auto cleanup_callback = [resource_holder]() mutable {
+          resource_holder.reset();
+        };
+
+        device.register_producer_on_stream(
+            std::move(outDevPtr), stream_handle, cleanup_callback);
+        for (auto data_ptr : inDevPtr) {
+          device.get_device_memory().recordStream(
+              // NOLINTNEXTLINE(performance-no-int-to-ptr)
+              reinterpret_cast<void*>(data_ptr),
+              hpu_stream);
+        }
+        for (auto data_ptr : outDevPtr) {
+          device.get_device_memory().recordStream(
+              // NOLINTNEXTLINE(performance-no-int-to-ptr)
+              reinterpret_cast<void*>(data_ptr),
+              hpu_stream);
+        }
+        // Launch collective ops
+        collective_kernels_info_->Launch(true, cleanup_callback);
+      }
     }
-
-    if (!(common::IsRecordStreamEnabled() &&
-          GET_ENV_FLAG_NEW(PT_HPU_USE_LAUNCH_RECORD_STREAM))) {
-      // Use wrapper for resources that must survive async part of the compute.
-      struct ResourceHolder {
-        std::shared_ptr<synapse_helpers::graph::recipe_handle> recipe_id_;
-        std::vector<at::Tensor> output_tensors_;
-        std::vector<at::Tensor> input_tensors_;
-        std::unique_ptr<synapse_helpers::device_ptr_lock> address_lock;
-        synapse_helpers::active_recipe_counter* recipe_counter_ptr;
-        size_t active_graph_key_;
-      };
-      auto resource_holder = std::shared_ptr<ResourceHolder>(
-          new ResourceHolder(), [](ResourceHolder* resource_holder) {
-            auto recipe_counter_ptr = resource_holder->recipe_counter_ptr;
-            auto recipe_handle = resource_holder->recipe_id_;
-            delete resource_holder;
-            recipe_counter_ptr->decrease_and_notify();
-            if (synapse_helpers::memory_reporter_enable() &&
-                resource_holder->active_graph_key_ > 0) {
-              auto& device = HPURegistrar::get_device();
-              synapse_helpers::MemoryReporter* reporter =
-                  device.get_device_memory().get_memory_reporter();
-              reporter->getGraphStats()->removeLiveGraph(
-                  resource_holder->active_graph_key_);
-            }
-            towl::emitRecipeFinished(recipe_handle.get());
-            PT_LAZY_DEBUG("call decrease and notify of recipe_counter");
-          });
-      // recipe_id_ needs to be passed to done_cb to ensure its lifetime until
-      // corresponding recipe is finished on stream
-      const auto& recipe_ptr = recipe_;
-      resource_holder->recipe_id_ = recipe_ptr;
-      if (not common::IsRecordStreamNoHolderEnabled()) {
-        resource_holder->output_tensors_ = outPtRefs;
-        resource_holder->address_lock = std::move(address_lock);
-        resource_holder->input_tensors_ = ptRefs;
-      }
-      resource_holder->recipe_counter_ptr = &recipe_counter;
-      resource_holder->active_graph_key_ = active_graph_key_;
-      // ResourceHolder could be used directly as callback, if we would only
-      // implement operator(), but copying of ResourceHolder would result in
-      // copying of all shared_ptr stored inside (including std::vector). To
-      // make sharing more lightweight we hide ResourceHolder behind one
-      // shared_ptr. This indirection allows us to maintain only one shared
-      // reference.
-      auto cleanup_callback = [resource_holder]() mutable {
-        resource_holder.reset();
-      };
-
-      device.register_producer_on_stream(
-          std::move(outDevPtr), stream_handle, cleanup_callback);
-      // Launch collective ops
-      collective_kernels_info_->Launch(true, cleanup_callback);
-    } else {
-      // Use wrapper for resources that must survive async part of the compute.
-      struct ResourceHolder {
-        std::shared_ptr<synapse_helpers::graph::recipe_handle> recipe_id_;
-        std::unique_ptr<synapse_helpers::device_ptr_lock> address_lock;
-        synapse_helpers::active_recipe_counter* recipe_counter_ptr;
-        size_t active_graph_key_;
-      };
-      auto resource_holder = std::shared_ptr<ResourceHolder>(
-          new ResourceHolder(), [](ResourceHolder* resource_holder) {
-            auto recipe_counter_ptr = resource_holder->recipe_counter_ptr;
-            auto recipe_handle = resource_holder->recipe_id_;
-            delete resource_holder;
-            recipe_counter_ptr->decrease_and_notify();
-            if (synapse_helpers::memory_reporter_enable() &&
-                resource_holder->active_graph_key_ > 0) {
-              auto& device = HPURegistrar::get_device();
-              synapse_helpers::MemoryReporter* reporter =
-                  device.get_device_memory().get_memory_reporter();
-              reporter->getGraphStats()->removeLiveGraph(
-                  resource_holder->active_graph_key_);
-            }
-            towl::emitRecipeFinished(recipe_handle.get());
-            PT_LAZY_DEBUG("call decrease and notify of recipe_counter");
-          });
-      // recipe_id_ needs to be passed to done_cb to ensure its lifetime until
-      // corresponding recipe is finished on stream
-      const auto& recipe_ptr = recipe_;
-      resource_holder->recipe_id_ = recipe_ptr;
-      if (not common::IsRecordStreamNoHolderEnabled()) {
-        resource_holder->address_lock = std::move(address_lock);
-      }
-      resource_holder->recipe_counter_ptr = &recipe_counter;
-      resource_holder->active_graph_key_ = active_graph_key_;
-      // ResourceHolder could be used directly as callback, if we would only
-      // implement operator(), but copying of ResourceHolder would result in
-      // copying of all shared_ptr stored inside (including std::vector). To
-      // make sharing more lightweight we hide ResourceHolder behind one
-      // shared_ptr. This indirection allows us to maintain only one shared
-      // reference.
-      auto cleanup_callback = [resource_holder]() mutable {
-        resource_holder.reset();
-      };
-
-      device.register_producer_on_stream(
-          std::move(outDevPtr), stream_handle, cleanup_callback);
-      for (auto data_ptr : inDevPtr) {
-        device.get_device_memory().recordStream(
-            // NOLINTNEXTLINE(performance-no-int-to-ptr)
-            reinterpret_cast<void*>(data_ptr),
-            hpu_stream);
-      }
-      for (auto data_ptr : outDevPtr) {
-        device.get_device_memory().recordStream(
-            // NOLINTNEXTLINE(performance-no-int-to-ptr)
-            reinterpret_cast<void*>(data_ptr),
-            hpu_stream);
-      }
-      // Launch collective ops
-      collective_kernels_info_->Launch(true, cleanup_callback);
-    }
-
   } else {
     std::vector<synapse_helpers::shared_event> ext_events;
     std::unique_ptr<synapse_helpers::device_ptr_lock> address_lock;
@@ -1784,7 +1791,6 @@
       reporter->getGraphStats()->removeLiveGraph(active_graph_key_);
     }
   }
-
   num_launches++;
   PT_BRIDGE_END;
 }
diff --git a/backend/profiling/profiling.h b/backend/profiling/profiling.h
index 2c95d9d..afb6d2e 100644
--- a/backend/profiling/profiling.h
+++ b/backend/profiling/profiling.h
@@ -137,9 +137,9 @@
 };
 
 namespace bridge {
-void trace_start(std::string_view id);
-void trace_end(std::string_view id);
-bool is_enabled(std::string_view id);
+void trace_start(std::string_view id, bool extra_trace = false);
+void trace_end(std::string_view id, bool extra_trace = false);
+bool is_enabled(std::string_view id, bool extra_trace = false);
 }; // namespace bridge
 
 }; // namespace profile
diff --git a/backend/profiling/trace_sources/bridge_logs_source.cpp b/backend/profiling/trace_sources/bridge_logs_source.cpp
index f956ee7..a4115e7 100644
--- a/backend/profiling/trace_sources/bridge_logs_source.cpp
+++ b/backend/profiling/trace_sources/bridge_logs_source.cpp
@@ -37,8 +37,8 @@
 struct BridgeLogsSourceImpl : public TraceSource {
   BridgeLogsSourceImpl() = default;
   ~BridgeLogsSourceImpl() override = default;
-  void log(std::string_view id, bool is_begin) {
-    if (enabled(id)) {
+  void log(std::string_view id, bool is_begin, bool extra_trace) {
+    if (extra_trace or enabled(id)) {
       int64_t dtime = NowMicros();
       pid_t tid = syscall(__NR_gettid);
       std::string event_id{id};
@@ -173,14 +173,14 @@
 }
 
 namespace bridge {
-void trace_start(std::string_view id) {
-  BridgeLogsSourceImpl::instance().log(id, true);
+void trace_start(std::string_view id, bool extra_trace) {
+  BridgeLogsSourceImpl::instance().log(id, true, extra_trace);
 }
-void trace_end(std::string_view id) {
-  BridgeLogsSourceImpl::instance().log(id, false);
+void trace_end(std::string_view id, bool extra_trace) {
+  BridgeLogsSourceImpl::instance().log(id, false, extra_trace);
 }
-bool is_enabled(std::string_view name) {
-  return BridgeLogsSourceImpl::instance().enabled(name);
+bool is_enabled(std::string_view name, bool extra_trace) {
+  return extra_trace or BridgeLogsSourceImpl::instance().enabled(name);
 }
 }; // namespace bridge
 }; // namespace profile
diff --git a/pytorch_helpers/habana_helpers/logging.h b/pytorch_helpers/habana_helpers/logging.h
index 98b6c17..c441821 100644
--- a/pytorch_helpers/habana_helpers/logging.h
+++ b/pytorch_helpers/habana_helpers/logging.h
@@ -263,14 +263,20 @@
   const std::string_view pName;
   const std::string_view name;
   bool isActive;
+  bool extra_trace;
 
  public:
   PTFuncLog(
       const std::string_view module,
       const std::string_view pn,
       const std::string_view n,
-      bool isActive)
-      : module(module), pName(pn), name(n), isActive(isActive) {
+      bool isActive,
+      bool extra_trace = false)
+      : module(module),
+        pName(pn),
+        name(n),
+        isActive(isActive),
+        extra_trace(extra_trace) {
     if (isActive) {
       HLLOG_TRACE(
           PT_TRACE,
@@ -283,7 +289,7 @@
               pName));
     }
     synapse_helpers::trace_start(name.data());
-    habana::profile::bridge::trace_start(name);
+    habana::profile::bridge::trace_start(name, extra_trace);
   }
   ~PTFuncLog() {
     if (isActive) {
@@ -293,7 +299,7 @@
               "[Rank:", Logger::get_rank(), "] ", module, ": end of ", pName));
     }
     synapse_helpers::trace_end(name.data());
-    habana::profile::bridge::trace_end(name);
+    habana::profile::bridge::trace_end(name, extra_trace);
   }
 };
 
@@ -426,11 +432,17 @@
     ptFuncLogger.emplace(#MOD, PNAME, NAME, isTraceLoggerEnabled); \
   }
 
+#define PT_MOD_EXTRA_TRACE(MOD, PNAME, NAME) \
+  std::optional<PTFuncLog> ptFuncLogger{};   \
+  ptFuncLogger.emplace(#MOD, PNAME, NAME, true, true);
+
 #define PT_MOD_TRACE(MOD, PNAME, NAME) PT_MOD_SCOPE(MOD, PNAME, NAME)
 
 #define PT_EAGER_TRACE PT_MOD_TRACE(PT_EAGER, __PRETTY_FUNCTION__, __FUNCTION__)
 #define PT_EAGER_TRACE_WITH_NAME(name) PT_MOD_TRACE(PT_EAGER, name, name)
 
+#define PT_EXTRA_TRACE(name) \
+  PT_MOD_EXTRA_TRACE(PT_BRIDGE, __PRETTY_FUNCTION__, name)
 #define PT_LAZY_TRACE PT_MOD_TRACE(PT_LAZY, __PRETTY_FUNCTION__, __FUNCTION__)
 #define PT_LAZY_TRACE_WITH_NAME(name) PT_MOD_TRACE(PT_LAZY, name, name)
 #define PT_BRIDGE_TRACE \
@@ -574,4 +586,4 @@
 inline std::string to_hexstring(const void* x) {
   return to_hexstring((std::uint64_t)x);
 }
-} // namespace
\ No newline at end of file
+} // namespace
