From 60a169146ce17e2356292e697ff572339bc28a14 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Thu, 17 Oct 2024 16:23:20 +0300
Subject: [PATCH] [SW-0] gaudi2: btl did not run - hw arc barrier

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I2d8923a58832430ba7d087ddbaf7440e80d91ef9
---

diff --git a/drivers/d2d/controller/d2d_controller.c b/drivers/d2d/controller/d2d_controller.c
index 6f69725..af579e0 100644
--- a/drivers/d2d/controller/d2d_controller.c
+++ b/drivers/d2d/controller/d2d_controller.c
@@ -324,6 +324,7 @@
 	for (die = drv_data->die_id; die < drv_cfg->regs_num; die++) {
 		single_die_mac_core_bcast(wval,
 				drv_cfg->base_addr[die] + D2D_MAC0_CORE0_CTRL_PKTZR_1);
+		HL_ARC_W_BARRIER();
 	}
 }
 
@@ -374,12 +375,15 @@
 	rc = d2d_phy_sw_train_ctrl_set(d2d_phy_dev, 0x1);
 	if (rc)
 		return rc;
+
 	rc = link_train_init(dev, MASTER_DIE_ID);
 	if (rc)
 		return rc;
+
 	rc = d2d_phy_sw_link_train(d2d_phy_dev);
 	if (rc)
 		return rc;
+
 	/* Handoff training control back to HW */
 	return d2d_phy_sw_train_ctrl_set(d2d_phy_dev, 0x0);
 }
@@ -943,6 +947,8 @@
 	}
 
 	d2d_controller_drv_traffic_en_set(dev, true);
+	HL_ARC_W_BARRIER();
+
 #endif
 	rc = d2d_controller_retrain_cfg();
 	if (rc) {
diff --git a/drivers/d2d/phy/d2d_phy.c b/drivers/d2d/phy/d2d_phy.c
index 20484ae..612c6d1 100644
--- a/drivers/d2d/phy/d2d_phy.c
+++ b/drivers/d2d/phy/d2d_phy.c
@@ -105,6 +105,7 @@
 		/* clear MCU memory write pointer, disable write protection */
 		single_die_dphy_bcast(0x0, drv_cfg->base_addr[die] +
 					DPHY0_CH_BCAST_PHYCTRLREG_PROGMEMWPTR);
+		HL_ARC_W_BARRIER();
 		/* broadcast PHY FW */
 		for (dw_idx = 0; dw_idx < D2D_PHY_FW_DW_NUM; dw_idx++) {
 			single_die_dphy_bcast(d2d_phy_fw_arr[dw_idx], drv_cfg->base_addr[die] +
@@ -139,6 +140,8 @@
 					drv_cfg->base_addr[die] + DPHY0_CTRL_INIT_PLL_CTRL2);
 		single_die_dphy_bcast(DPHY_CTRL_INIT_PLL_CTRL3_VAL,
 					drv_cfg->base_addr[die] + DPHY0_CTRL_INIT_PLL_CTRL3);
+		HL_ARC_W_BARRIER();
+
 	}
 }
 
@@ -1154,24 +1157,40 @@
 	int rc;
 
 	d2d_phy_drv_dphy_sys_reset_set(dev, true);
+	HL_ARC_W_BARRIER();
 	d2d_io_cfg(dev);
+	HL_ARC_W_BARRIER();
 	d2d_phy_drv_dphy_sys_reset_set(dev, false);
+	HL_ARC_W_BARRIER();
+
 	rc = d2d_phy_drv_pll_init(dev, false, pll_freq_mhz);
 	if (rc)
 		return rc;
+	HL_ARC_W_BARRIER();
+
 #if defined(CONFIG_D2D_POST_SYS_RST_DCC_DIS)
 	rc = dphy_phy_status_wait(dev);
 	if (rc)
 		return rc;
 	LOG_DBG("Applying post-DCC reset WA");
 	d2d_phy_drv_dphy_sys_reset_set(dev, true);
+	HL_ARC_W_BARRIER();
+
 	/* disable dcc ckmeas , set phyinit_en again */
 	phy_init_ctrl_cfg(dev, true, false, true, false);
+	HL_ARC_W_BARRIER();
+
 	phy_pll_en_cfg(dev, false);
+	HL_ARC_W_BARRIER();
+
 	/* wait for internal logic to clear deglitch buffer */
 	k_busy_wait(1);
 	d2d_phy_drv_dphy_sys_reset_set(dev, false);
+	HL_ARC_W_BARRIER();
+
 	phy_pll_en_cfg(dev, true);
+	HL_ARC_W_BARRIER();
+
 	rc = dphy_pll_and_ofstcal_wait_for_lock(dev, 0x1, false, DPHY_PLL_LOCK_TIMEOUT_US);
 	if (rc)
 		return rc;
@@ -2781,6 +2800,8 @@
 					REKTIMER_VAL);
 		single_die_dphy_bcast(val, drv_cfg->base_addr[die] +
 						DPHY0_CH_BCAST_PHYCTRLREG_PHYUPDTCTRLSTS);
+		HL_ARC_W_BARRIER();
+
 	}
 	LOG_DBG("%s VT adaptive receivers and data eye monitors", en ? "Enabled" : "Disabled");
 }
