From 23f958fc0035f9af5ec35bd27382c1048e7caa9a Mon Sep 17 00:00:00 2001
From: Tadeusz Lipinski <tlipinski@habana.ai>
Date: Wed, 30 Oct 2024 16:07:27 +0200
Subject: [PATCH] [u22+py311 testing] do not merge

Change-Id: I0301385f5d1679b32260b7b5e17d6905d6178e75
---

diff --git a/.devops/cd/build/Makefile b/.devops/cd/build/Makefile
index 59d2afc..8c0d2ea 100644
--- a/.devops/cd/build/Makefile
+++ b/.devops/cd/build/Makefile
@@ -22,10 +22,10 @@
 CACHE_DIR ?= $(HOME)/.cd_cache/pytorch/cache
 CCACHE_CONF ?= $(CURDIR)/ccache.conf
 
-DOCKER_IMAGE ?= artifactory-kfs.habana-labs.com/devops-docker-local/habana-builder-pt:$(BUILD_OS)-$(RELEASE_VERSION)
+DOCKER_IMAGE ?= artifactory-kfs.habana-labs.com/devops-docker-local/habana-builder-pt:ubuntu22.04-1.19.0-207
 DOCKER_IMAGE_NDA ?= artifactory-kfs.habana-labs.com/devops-docker-local/habana-tools-cd:nda
 
-SW_ARTIFACT_ROOT_URL ?= https://artifactory-kfs.habana-labs.com/artifactory/$(SW_META_REPO)/$(RELEASE_VERSION)/$(RELEASE_BUILD_ID)/$(subst .,,$(BUILD_OS))/binary
+SW_ARTIFACT_ROOT_URL ?= https://artifactory-kfs.habana-labs.com/artifactory/sw-build-generic-preprod-local/$(RELEASE_VERSION)/$(RELEASE_BUILD_ID)/$(subst .,,$(BUILD_OS))/binary
 
 PT_VERSION ?= $(shell jq -rj '.pt_versions.current.version' $(SRC_DIR)/pytorch-integration/.devops/build_profiles/profiles.json)
 
