From f2af105fb4b48c0b3821ec292bb5a1b15441fd0c Mon Sep 17 00:00:00 2001
From: Tadeusz Lipinski <tlipinski@habana.ai>
Date: Wed, 30 Oct 2024 23:16:01 +0200
Subject: [PATCH] [u22+py310] do not merge

Change-Id: I7c2d321f3ff567e16d9e74a4a2b1afde86d51641
---

diff --git a/tests/asic/PT-Suite/pt-run-docker-container.sh b/tests/asic/PT-Suite/pt-run-docker-container.sh
index 22fd326..d5ebbd2 100755
--- a/tests/asic/PT-Suite/pt-run-docker-container.sh
+++ b/tests/asic/PT-Suite/pt-run-docker-container.sh
@@ -45,7 +45,8 @@
 esac
 
 pt_version=$(get_pytorch_version "${PT_VERSION_TYPE}")
-pytorch_docker="${ARTIFACTORY_SERVER}/${ARTIFACTORY_DOCKER_REGISTRY}/${RELEASE_VERSION}/${HL_TEST_OS}/habanalabs/pytorch-installer-${pt_version}:${RELEASE_VERSION}-${RELEASE_BUILD_ID}"
+pytorch_docker="artifactory-kfs.habana-labs.com/staging_docker-local/1.19.0/ubuntu22.04/habanalabs/external/pytorch-installer-2.4.1:1.19.0-316"
+# pytorch_docker="${ARTIFACTORY_SERVER}/${ARTIFACTORY_DOCKER_REGISTRY}/${RELEASE_VERSION}/${HL_TEST_OS}/habanalabs/pytorch-installer-${pt_version}:${RELEASE_VERSION}-${RELEASE_BUILD_ID}"
 
 export PT_CONTAINER_NAME="HL_PT_DOCKER"
 export PYTORCH_TESTS_ROOT="/root/pytorch-training-tests"
