From a72d8bb533cb5db4a3fe0c3191997c025bfaa8fd Mon Sep 17 00:00:00 2001
From: Tadeusz Lipinski <tlipinski@habana.ai>
Date: Wed, 30 Oct 2024 20:59:13 +0200
Subject: [PATCH] [u22+py311testing] do not merge

Change-Id: I1a866cd4e98c4e6002a5710c67ac3310c399c95b
---

diff --git a/dockerfiles/pytorch/install_packages.sh b/dockerfiles/pytorch/install_packages.sh
index af21152..651f52f 100755
--- a/dockerfiles/pytorch/install_packages.sh
+++ b/dockerfiles/pytorch/install_packages.sh
@@ -1,7 +1,7 @@
 #!/bin/bash
 set -ex
 
-PT_BUILD_REPO="${PT_BUILD_REPO:-hl-generic-prod-local}"
+PT_BUILD_REPO="staging_hl-generic-prod-local"
 PYPI_REPO_NAME="${PYPI_REPO_NAME:-pypi-virtual}"
 
 pt_package_name="pytorch_modules-v${PT_VERSION}_${VERSION}_${REVISION}.tgz"
