From 86020e3ce3d56cbdf765578a7e1d4b5d3401e9a6 Mon Sep 17 00:00:00 2001
From: Sergei Shalnov <sergey.shalnov@habana.ai>
Date: Mon, 04 Nov 2024 10:51:48 +0200
Subject: [PATCH] [SW-201534] Fix license header part 1

Change-Id: Ib483c08bd9c7d67031d216ab9a3fe5c3e567ac12
---

diff --git a/habana_kernels/basic_kernels.h b/habana_kernels/basic_kernels.h
index 2f2fac9..25b1143 100644
--- a/habana_kernels/basic_kernels.h
+++ b/habana_kernels/basic_kernels.h
@@ -1,15 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020-2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include <perf_lib_layer_params.h>
 #include "backend/habana_operator.h"
diff --git a/habana_kernels/binary_inplace_kernels.h b/habana_kernels/binary_inplace_kernels.h
index 6023e9d..ac532f5 100644
--- a/habana_kernels/binary_inplace_kernels.h
+++ b/habana_kernels/binary_inplace_kernels.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 #include "habana_kernels/tensor_shape_kernels.h"
diff --git a/habana_kernels/binary_kernels.h b/habana_kernels/binary_kernels.h
index c609655..a005e33 100644
--- a/habana_kernels/binary_kernels.h
+++ b/habana_kernels/binary_kernels.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 #include "habana_kernels/tensor_shape_kernels.h"
diff --git a/habana_kernels/compare_kernels.h b/habana_kernels/compare_kernels.h
index 53e469b..e521a9b 100644
--- a/habana_kernels/compare_kernels.h
+++ b/habana_kernels/compare_kernels.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 #include "backend/helpers/tensor_utils.h"
diff --git a/habana_kernels/custom_op_kernel.h b/habana_kernels/custom_op_kernel.h
index a52d7e4..d930b90 100644
--- a/habana_kernels/custom_op_kernel.h
+++ b/habana_kernels/custom_op_kernel.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2021 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 #include "backend/habana_operator.h"
diff --git a/habana_kernels/embedding_kernels.h b/habana_kernels/embedding_kernels.h
index 86bb7ea..d25f5b6 100644
--- a/habana_kernels/embedding_kernels.h
+++ b/habana_kernels/embedding_kernels.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 namespace habana {
diff --git a/habana_kernels/fallback_helper.h b/habana_kernels/fallback_helper.h
index 5b7d8c2..5c5d536 100644
--- a/habana_kernels/fallback_helper.h
+++ b/habana_kernels/fallback_helper.h
@@ -1,13 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2021 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly
- * prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include <c10/util/Exception.h>
 #include <mutex>
 #include <unordered_map>
diff --git a/habana_kernels/hccl_kernels.h b/habana_kernels/hccl_kernels.h
index fe6f654..070566b 100644
--- a/habana_kernels/hccl_kernels.h
+++ b/habana_kernels/hccl_kernels.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 
diff --git a/habana_kernels/index_kernels.h b/habana_kernels/index_kernels.h
index fa2532c..79eedf3 100644
--- a/habana_kernels/index_kernels.h
+++ b/habana_kernels/index_kernels.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 namespace habana {
diff --git a/habana_kernels/instance_norm_utils.h b/habana_kernels/instance_norm_utils.h
index 32ca978..e817dbb 100644
--- a/habana_kernels/instance_norm_utils.h
+++ b/habana_kernels/instance_norm_utils.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 /*******************************************************************************
  * Core content of this file comes from:
  * pytorch-fork/torch/csrc/autograd/FunctionsManual.h
diff --git a/habana_kernels/kernel_input_checks.h b/habana_kernels/kernel_input_checks.h
index 34063df..5fc6009 100644
--- a/habana_kernels/kernel_input_checks.h
+++ b/habana_kernels/kernel_input_checks.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <ATen/core/TensorBody.h>
diff --git a/habana_kernels/kernel_utils.h b/habana_kernels/kernel_utils.h
index 8261ee3..cdec30c 100644
--- a/habana_kernels/kernel_utils.h
+++ b/habana_kernels/kernel_utils.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <c10/util/ArrayRef.h>
diff --git a/habana_kernels/nms_kernels.h b/habana_kernels/nms_kernels.h
index 1c65236..42c8570 100644
--- a/habana_kernels/nms_kernels.h
+++ b/habana_kernels/nms_kernels.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 namespace habana {
diff --git a/habana_kernels/nonzero_kernel.h b/habana_kernels/nonzero_kernel.h
index c74490f..cd77d72 100644
--- a/habana_kernels/nonzero_kernel.h
+++ b/habana_kernels/nonzero_kernel.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 namespace habana {
diff --git a/habana_kernels/norm_kernels.h b/habana_kernels/norm_kernels.h
index 5b1cba0..8e1cab3 100644
--- a/habana_kernels/norm_kernels.h
+++ b/habana_kernels/norm_kernels.h
@@ -1,15 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 #include "habana_kernels/index_kernels.h"
diff --git a/habana_kernels/shape_agnostic_helper.h b/habana_kernels/shape_agnostic_helper.h
index 9d85116..4859174 100644
--- a/habana_kernels/shape_agnostic_helper.h
+++ b/habana_kernels/shape_agnostic_helper.h
@@ -1,13 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2021 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly
- * prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include <c10/util/Exception.h>
 #include <mutex>
 #include <unordered_map>
diff --git a/habana_kernels/template_helpers.h b/habana_kernels/template_helpers.h
index 131fe06..0da095a 100644
--- a/habana_kernels/template_helpers.h
+++ b/habana_kernels/template_helpers.h
@@ -1,15 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <ATen/Tensor.h>
diff --git a/habana_kernels/tensor_shape_kernels.h b/habana_kernels/tensor_shape_kernels.h
index 8ac98b1..646b30f 100644
--- a/habana_kernels/tensor_shape_kernels.h
+++ b/habana_kernels/tensor_shape_kernels.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 
diff --git a/habana_kernels/topk_kernels.h b/habana_kernels/topk_kernels.h
index d0ce6c2..49afa7b 100644
--- a/habana_kernels/topk_kernels.h
+++ b/habana_kernels/topk_kernels.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "backend/habana_operator.h"
 namespace habana {
