From 1d53e44790ed52fda53eba44b2d392e17649342b Mon Sep 17 00:00:00 2001
From: Gennady Davidov <gdavidov@habana.ai>
Date: Mon, 04 Nov 2024 13:23:27 +0200
Subject: [PATCH] [SW-200181] DONT MERGE - ubuntu22.04 install python3.11 -

Change-Id: I85de8f57dfbeb7b9c3cc25466268ce4c118af8e0
Signed-off-by: Gennady Davidov <gdavidov@habana.ai>
---

diff --git a/templates/ansible/playbooks/cd.yml b/templates/ansible/playbooks/cd.yml
index 6d16c91..a1c7e3a 100644
--- a/templates/ansible/playbooks/cd.yml
+++ b/templates/ansible/playbooks/cd.yml
@@ -51,4 +51,3 @@
     - include_role:
         name: tools/xrd
       when: ansible_os_family != "RedHat"
-
diff --git a/templates/ansible/roles/kernel/tasks/setup-ubuntu.yml b/templates/ansible/roles/kernel/tasks/setup-ubuntu.yml
index 11bee21..71c35d2 100644
--- a/templates/ansible/roles/kernel/tasks/setup-ubuntu.yml
+++ b/templates/ansible/roles/kernel/tasks/setup-ubuntu.yml
@@ -1,5 +1,4 @@
 ---
-
 - name: Install kernel packages - deb
   package:
     name: "{{ kernel_package }}"
diff --git a/templates/base/ubuntu/scripts/common.sh b/templates/base/ubuntu/scripts/common.sh
index 1aa36d7..4d359f2 100644
--- a/templates/base/ubuntu/scripts/common.sh
+++ b/templates/base/ubuntu/scripts/common.sh
@@ -1,57 +1,57 @@
 #!/bin/sh -eux
 
-ubuntu_version="`lsb_release -r | awk '{print $2}'`"
-major_version="`echo $ubuntu_version | awk -F. '{print $1}'`"
+# Get the Ubuntu version and major version
+ubuntu_version="$(lsb_release -r | awk '{print $2}')"
+major_version="$(echo "$ubuntu_version" | awk -F. '{print $1}')"
 
+# Comment out existing sources and secure path settings in sudoers for security
 sed -i 's/^[ \t]*\(deb[^\n]+\)$/#\1/g' /etc/apt/sources.list
 sed -i 's/\(Defaults.*secure_path.*\)/#\1/g' /etc/sudoers
 
+# Update package lists
 apt-get -y update
 
-apt install -y wget lsof vim git git-lfs
+# Install essential tools
+apt-get install -y wget lsof vim git git-lfs
 
-apt install -y \
-    python3 \
-    python3-pip \
-    python3-dev \
-    python3-tk \
-    python3-setuptools
-
-if [ "$major_version" -ge "20" ]; then
-   ln -snf /usr/bin/python3 /usr/bin/python
-#    echo "send dhcp-client-identifier = hardware;" >> /etc/dhcp/dhclient.conf
-fi
-
-if [ "$major_version" -ge "22" ]; then
-
-   apt install -y nfs-common
-   apt remove  -y needrestart
-fi
-
-apt install -y unattended-upgrades
+# Enable unattended upgrades
+apt-get install -y unattended-upgrades
 unattended-upgrade -d
 
-systemctl stop ufw.service
-systemctl disable ufw.service
+apt-get install -y python3 python3-pip python3-dev python3-tk python3-setuptools
 
+# Python installation based on Ubuntu version
+if [ "$major_version" -eq "22" ]; then
+    # Install Python 3.11 and associated packages for Ubuntu 22.04
+    apt-get install -y python3.11 python3.11-dev python3.11-tk python3.11-venv
+
+    # Install pip for Python 3.11
+    curl -sS https://bootstrap.pypa.io/get-pip.py | python3.11
+
+    # Additional configurations specific to Ubuntu 22.04
+    apt-get install -y nfs-common
+    apt-get remove -y needrestart
+fi
+
+# Additional fix for DNS resolving on Ubuntu 18
+if [ "$major_version" -eq "18" ]; then
+    ln -snf /run/systemd/resolve/resolv.conf /etc/resolv.conf
+fi
+
+# Disable and stop firewall for unrestricted access
+systemctl stop ufw.service || true
+systemctl disable ufw.service || true
+
+# Remove dpkg exclude configurations to avoid limiting package installs
 rm -rf /etc/dpkg/dpkg.cfg.d/excludes
 
+# Disable VirtualBox guest services if they are present
 systemctl stop vboxadd.service || true
 systemctl stop vboxadd-service.service || true
 systemctl disable vboxadd.service || true
 systemctl disable vboxadd-service.service || true
 
-# disable quiet for kernel boot
+# Modify GRUB to remove 'quiet' and 'splash' for verbose booting
 sed -i 's/quiet//g' /etc/default/grub
 sed -i 's/splash//g' /etc/default/grub
-
 update-grub
-
-# this is a wa to fix bug with systemd-resolved.service loosing DNS resolving, see here:
-# https://github.com/systemd/systemd/issues/10298
-if [ "$major_version" -eq "18" ]; then
-   ln -snf /run/systemd/resolve/resolv.conf /etc/resolv.conf
-# habanalabs comes built in ubuntu 21, need to blacklist
-elif [ "$major_version" -eq "21" ]; then
-   sudo grep -qxF 'blacklist habanalabs' /etc/modprobe.d/blacklist-habanalabs.conf || sudo echo -e 'blacklist habanalabs\nblacklist habanalabs_en' | sudo tee -a /etc/modprobe.d/blacklist-habanalabs.conf
-fi
\ No newline at end of file
