From c48913bead73083f0586a2d5bcd1d3d103d45cad Mon Sep 17 00:00:00 2001
From: Sergei Shalnov <sergey.shalnov@habana.ai>
Date: Tue, 05 Nov 2024 17:28:23 +0200
Subject: [PATCH] [SW-201534] Fix license header part 5

Change-Id: I1b5b05136814bf028a80e616a501fff997e750c5
---

diff --git a/habana_lazy/aten_lazy_bridge.h b/habana_lazy/aten_lazy_bridge.h
index 5383da4..1aee00f 100644
--- a/habana_lazy/aten_lazy_bridge.h
+++ b/habana_lazy/aten_lazy_bridge.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include <c10/core/TensorImpl.h>
 #include "backend/helpers/get_n_bytes.h"
 #include "hpu_lazy_tensors.h"
diff --git a/habana_lazy/debug_utils.h b/habana_lazy/debug_utils.h
index a151693..8b8b39d 100644
--- a/habana_lazy/debug_utils.h
+++ b/habana_lazy/debug_utils.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "aten_lazy_bridge.h"
 #include "backend/synapse_helpers/env_flags.h"
diff --git a/habana_lazy/hlexec.h b/habana_lazy/hlexec.h
index 35c41e7..1481f84 100644
--- a/habana_lazy/hlexec.h
+++ b/habana_lazy/hlexec.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 #include <iostream>
diff --git a/habana_lazy/hpu_lazy_tensors.h b/habana_lazy/hpu_lazy_tensors.h
index a0f04fa..fbb244f 100644
--- a/habana_lazy/hpu_lazy_tensors.h
+++ b/habana_lazy/hpu_lazy_tensors.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 #include <c10/core/Device.h>
diff --git a/habana_lazy/hpu_stage_submission.h b/habana_lazy/hpu_stage_submission.h
index 4364ba6..1ab74db 100644
--- a/habana_lazy/hpu_stage_submission.h
+++ b/habana_lazy/hpu_stage_submission.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <atomic>
diff --git a/habana_lazy/ir.h b/habana_lazy/ir.h
index 9141fac..06b4306 100644
--- a/habana_lazy/ir.h
+++ b/habana_lazy/ir.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 #include <absl/container/flat_hash_set.h>
diff --git a/habana_lazy/ir_utils.h b/habana_lazy/ir_utils.h
index 6f09b44..f82ffbc 100644
--- a/habana_lazy/ir_utils.h
+++ b/habana_lazy/ir_utils.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include <unordered_map>
 #include <vector>
diff --git a/habana_lazy/lazy_arg_spec.h b/habana_lazy/lazy_arg_spec.h
index 03496a3..abbcdbe 100644
--- a/habana_lazy/lazy_arg_spec.h
+++ b/habana_lazy/lazy_arg_spec.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include "backend/habana_operator.h"
diff --git a/habana_lazy/lazy_executor.h b/habana_lazy/lazy_executor.h
index 9519cdb..5e36bec 100644
--- a/habana_lazy/lazy_executor.h
+++ b/habana_lazy/lazy_executor.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include <thread>
 
diff --git a/habana_lazy/lazy_graph_hash_builder.h b/habana_lazy/lazy_graph_hash_builder.h
index 4d43b36..af0aeb0 100644
--- a/habana_lazy/lazy_graph_hash_builder.h
+++ b/habana_lazy/lazy_graph_hash_builder.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <limits>
diff --git a/habana_lazy/lazy_graph_hash_disabler.h b/habana_lazy/lazy_graph_hash_disabler.h
index 674d5bb..e66dffc 100644
--- a/habana_lazy/lazy_graph_hash_disabler.h
+++ b/habana_lazy/lazy_graph_hash_disabler.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/lazy_storage.h b/habana_lazy/lazy_storage.h
index 39583a4..5181ea8 100644
--- a/habana_lazy/lazy_storage.h
+++ b/habana_lazy/lazy_storage.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <ATen/Tensor.h>
diff --git a/habana_lazy/memlog.h b/habana_lazy/memlog.h
index 8cd2a3c..c71dba4 100644
--- a/habana_lazy/memlog.h
+++ b/habana_lazy/memlog.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include <string_view>
 
diff --git a/habana_lazy/permute_tensors.h b/habana_lazy/permute_tensors.h
index 47084a8..097ce4f 100644
--- a/habana_lazy/permute_tensors.h
+++ b/habana_lazy/permute_tensors.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 #include <torch/torch.h>
diff --git a/habana_lazy/sbs_debug.h b/habana_lazy/sbs_debug.h
index 514e85c..25f242d 100644
--- a/habana_lazy/sbs_debug.h
+++ b/habana_lazy/sbs_debug.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "hpu_lazy_tensors.h"
 
diff --git a/habana_lazy/sbs_runner.h b/habana_lazy/sbs_runner.h
index b59c5d3..71a5289 100644
--- a/habana_lazy/sbs_runner.h
+++ b/habana_lazy/sbs_runner.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 #include "habana_lazy/ir.h"
 
diff --git a/habana_lazy/tensor_impl.h b/habana_lazy/tensor_impl.h
index a3526bc..a12bc8b 100644
--- a/habana_lazy/tensor_impl.h
+++ b/habana_lazy/tensor_impl.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <ATen/Tensor.h>
diff --git a/habana_lazy/view.h b/habana_lazy/view.h
index c71c289..3a8292d 100644
--- a/habana_lazy/view.h
+++ b/habana_lazy/view.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 #include <iostream>
diff --git a/habana_lazy/view_utils.h b/habana_lazy/view_utils.h
index d4d7a7d..ddf20aa 100644
--- a/habana_lazy/view_utils.h
+++ b/habana_lazy/view_utils.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2021-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #pragma once
 
 #include <tuple>
