From 08ed76d3643946f8ea92ac190e14bea371ff0872 Mon Sep 17 00:00:00 2001
From: Sergei Shalnov <sergey.shalnov@habana.ai>
Date: Tue, 05 Nov 2024 17:29:03 +0200
Subject: [PATCH] [SW-201534] Fix license header part 6

Change-Id: Ib39bcd9bf0828f6aa759637d24326142c4ccf0c3
---

diff --git a/habana_lazy/aten_lazy_bridge.cpp b/habana_lazy/aten_lazy_bridge.cpp
index 9072f50..ce444ed 100644
--- a/habana_lazy/aten_lazy_bridge.cpp
+++ b/habana_lazy/aten_lazy_bridge.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "aten_lazy_bridge.h"
 #include "backend/backend_meta.h"
 #include "habana_helpers/misc_utils.h"
diff --git a/habana_lazy/debug_utils.cpp b/habana_lazy/debug_utils.cpp
index ddd7982..eb7a774 100644
--- a/habana_lazy/debug_utils.cpp
+++ b/habana_lazy/debug_utils.cpp
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include <regex>
 #include <sstream>
 #include <unordered_map>
diff --git a/habana_lazy/habana_custom_op.cpp b/habana_lazy/habana_custom_op.cpp
index 25a0b01..655869b 100644
--- a/habana_lazy/habana_custom_op.cpp
+++ b/habana_lazy/habana_custom_op.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2021-2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "habana_kernels/binary_kernels.h"
 #include "habana_kernels/custom_op_kernel.h"
 #include "habana_kernels/lazy_kernels.h"
diff --git a/habana_lazy/hlexec.cpp b/habana_lazy/hlexec.cpp
index 0b785be..e3d4bcb 100644
--- a/habana_lazy/hlexec.cpp
+++ b/habana_lazy/hlexec.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include <torch/csrc/jit/ir/irparser.h>
 #include <torch/csrc/jit/passes/common_subexpression_elimination.h>
 #include <torch/csrc/jit/passes/constant_pooling.h>
diff --git a/habana_lazy/hpu_lazy_tensors.cpp b/habana_lazy/hpu_lazy_tensors.cpp
index 130435a..014d127 100644
--- a/habana_lazy/hpu_lazy_tensors.cpp
+++ b/habana_lazy/hpu_lazy_tensors.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "hpu_lazy_tensors.h"
 #include <torch/csrc/jit/ir/ir.h>
 #include "backend/habana_device/HPUDevice.h"
diff --git a/habana_lazy/hpu_stage_submission.cpp b/habana_lazy/hpu_stage_submission.cpp
index 5649490..d938f25 100644
--- a/habana_lazy/hpu_stage_submission.cpp
+++ b/habana_lazy/hpu_stage_submission.cpp
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2021 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "habana_lazy/hpu_stage_submission.h"
 #include "lazy_executor.h"
 #include "pytorch_helpers/habana_helpers/kernels_accumulation.h"
diff --git a/habana_lazy/ir.cpp b/habana_lazy/ir.cpp
index 558ad6a..894b5b7 100644
--- a/habana_lazy/ir.cpp
+++ b/habana_lazy/ir.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #include "ir.h"
 #include <absl/strings/str_format.h>
diff --git a/habana_lazy/ir_utils.cpp b/habana_lazy/ir_utils.cpp
index 64de2e9..f3857b1 100644
--- a/habana_lazy/ir_utils.cpp
+++ b/habana_lazy/ir_utils.cpp
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "ir_utils.h"
 #include "habana_helpers/logging.h"
 #include "ir.h"
diff --git a/habana_lazy/lazy_arg_spec.cpp b/habana_lazy/lazy_arg_spec.cpp
index a6bc714..ae4d780 100644
--- a/habana_lazy/lazy_arg_spec.cpp
+++ b/habana_lazy/lazy_arg_spec.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "habana_lazy/lazy_arg_spec.h"
 #include "habana_lazy/aten_lazy_bridge.h"
 #include "habana_lazy/passes/pass_utils.h"
diff --git a/habana_lazy/lazy_executor.cpp b/habana_lazy/lazy_executor.cpp
index 40cad50..dbd0479 100644
--- a/habana_lazy/lazy_executor.cpp
+++ b/habana_lazy/lazy_executor.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "lazy_executor.h"
 #include "backend/habana_device/hpu_cached_devices.h"
 #include "habana_helpers/python_utils.h"
diff --git a/habana_lazy/lazy_graph_hash_builder.cpp b/habana_lazy/lazy_graph_hash_builder.cpp
index 99dacbd..776eeb6 100644
--- a/habana_lazy/lazy_graph_hash_builder.cpp
+++ b/habana_lazy/lazy_graph_hash_builder.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #include "habana_lazy/lazy_graph_hash_builder.h"
 #include "habana_lazy/lazy_executor.h"
diff --git a/habana_lazy/lazy_graph_hash_disabler.cpp b/habana_lazy/lazy_graph_hash_disabler.cpp
index 7687421..862cb88 100644
--- a/habana_lazy/lazy_graph_hash_disabler.cpp
+++ b/habana_lazy/lazy_graph_hash_disabler.cpp
@@ -1,15 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #include "habana_lazy/lazy_graph_hash_disabler.h"
 
diff --git a/habana_lazy/lazy_storage.cpp b/habana_lazy/lazy_storage.cpp
index d288787..f7c900d 100644
--- a/habana_lazy/lazy_storage.cpp
+++ b/habana_lazy/lazy_storage.cpp
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #include "lazy_storage.h"
 #include <ATen/Tensor.h>
diff --git a/habana_lazy/memlog.cpp b/habana_lazy/memlog.cpp
index cf7a805..3fdd706 100644
--- a/habana_lazy/memlog.cpp
+++ b/habana_lazy/memlog.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "memlog.h"
 #include <sstream>
 #include <utility>
diff --git a/habana_lazy/permute_tensors.cpp b/habana_lazy/permute_tensors.cpp
index be07982..c787a50 100644
--- a/habana_lazy/permute_tensors.cpp
+++ b/habana_lazy/permute_tensors.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #include "habana_lazy/permute_tensors.h"
 #include <c10/core/Storage.h>
diff --git a/habana_lazy/sbs_debug.cpp b/habana_lazy/sbs_debug.cpp
index d925d0f..79d786a 100644
--- a/habana_lazy/sbs_debug.cpp
+++ b/habana_lazy/sbs_debug.cpp
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2022 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "sbs_debug.h"
 #include "aten_lazy_bridge.h"
 #include "debug_utils.h"
diff --git a/habana_lazy/sbs_runner.cpp b/habana_lazy/sbs_runner.cpp
index a9e3ab8..48ec459 100644
--- a/habana_lazy/sbs_runner.cpp
+++ b/habana_lazy/sbs_runner.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2022-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "sbs_runner.h"
 #include <exception>
 #include <memory>
diff --git a/habana_lazy/tensor_impl.cpp b/habana_lazy/tensor_impl.cpp
index 36c23c1..d6d574a 100644
--- a/habana_lazy/tensor_impl.cpp
+++ b/habana_lazy/tensor_impl.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2023 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "tensor_impl.h"
 #include <c10/core/Device.h>
 #include <c10/core/ScalarType.h>
diff --git a/habana_lazy/user_custom_op_frontend.cpp b/habana_lazy/user_custom_op_frontend.cpp
index 41d6074..6682b68 100644
--- a/habana_lazy/user_custom_op_frontend.cpp
+++ b/habana_lazy/user_custom_op_frontend.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 #include "habana_kernels/lazy_kernels.h"
 #include "include/habanalabs/hpu_custom_op_pt2.h"
 
diff --git a/habana_lazy/view_utils.cpp b/habana_lazy/view_utils.cpp
index fe33fcd..61ca9ec 100644
--- a/habana_lazy/view_utils.cpp
+++ b/habana_lazy/view_utils.cpp
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2021-2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #include "habana_lazy/view_utils.h"
 #include "habana_kernels/basic_kernels.h"
