From d82f5e5cb9eeec93b33afd68882a97b2dc58b423 Mon Sep 17 00:00:00 2001
From: Adrian Shvay <ashvay@habana.ai>
Date: Tue, 05 Nov 2024 17:35:43 +0200
Subject: [PATCH] [SW-0] Lazy execution context must be deleted after pipeline is destroyed.  HabanaLaunchOpPT::run accesses to it context

Change-Id: I7d437199d0af8af8805807f130209f4828e6a81f
---

diff --git a/backend/habana_device/HPUDevice.cpp b/backend/habana_device/HPUDevice.cpp
index d9ba3b7..9076a55 100644
--- a/backend/habana_device/HPUDevice.cpp
+++ b/backend/habana_device/HPUDevice.cpp
@@ -45,6 +45,7 @@
   void JoinPipelineThreads();
   void CreateDevice();
   void Finish();
+  void ThreadsRelease();
 } device_context;
 
 void HPUDeviceContextImpl::JoinAllThreads() {
@@ -96,17 +97,20 @@
   constant_information_ = ConstantInformationPtr();
   scalar_cache_ = std::make_unique<backend::ScalarCache>();
 
+  HPURegistrar::get_hpu_registrar().register_thread_deleter(
+      []() { device_context.ThreadsRelease(); });
   HPURegistrar::get_hpu_registrar().register_device_deleter(
       []() { device_context.Finish(); });
 }
 
+void HPUDeviceContextImpl::ThreadsRelease() {
+  habana_helpers::AutoNoGIL gil_release;
+  device_context.lowering_thread_.reset();
+  device_context.compile_thread_.reset();
+  device_context.execute_thread_.reset();
+}
+
 void HPUDeviceContextImpl::Finish() {
-  {
-    habana_helpers::AutoNoGIL gil_release;
-    device_context.lowering_thread_.reset();
-    device_context.compile_thread_.reset();
-    device_context.execute_thread_.reset();
-  }
   device_context.recipe_cache_.reset();
   device_context.scalar_cache_.reset();
 
diff --git a/backend/habana_device/hpu_cached_devices.h b/backend/habana_device/hpu_cached_devices.h
index 79c264b..533787c 100644
--- a/backend/habana_device/hpu_cached_devices.h
+++ b/backend/habana_device/hpu_cached_devices.h
@@ -125,6 +125,11 @@
         std::move(lazy_execution_arena_cleanup));
   }
 
+  void register_thread_deleter(CallFinally::FinalFunc&& thread_deleter) {
+    TORCH_CHECK(!thread_deleter_);
+    thread_deleter_.reset(std::move(thread_deleter));
+  }
+
   void register_device_deleter(CallFinally::FinalFunc&& device_deleter) {
     TORCH_CHECK(!device_deleter_);
     device_deleter_.reset(std::move(device_deleter));
@@ -144,6 +149,7 @@
 
   CallFinally device_deleter_;
   CallFinally lazy_execution_arena_cleanup_{};
+  CallFinally thread_deleter_;
   CallFinally lazy_exec_thread_pool_cleanup_{};
   CallFinally process_group_finalizer_;
   CallFinally accumulation_thread_cleanup_{};
