From bb2d77268148737308027bfd412eb790d079640e Mon Sep 17 00:00:00 2001
From: Sergei Shalnov <sergey.shalnov@habana.ai>
Date: Tue, 05 Nov 2024 17:11:07 +0200
Subject: [PATCH] [SW-201534] Fix license header part 2

Change-Id: Ied4a5fde2ab779abeb7f20fa7217b37b6534b15b
---

diff --git a/habana_lazy/passes/fold_conv_batchnorm.h b/habana_lazy/passes/fold_conv_batchnorm.h
index cdd3535..75d3feb 100644
--- a/habana_lazy/passes/fold_conv_batchnorm.h
+++ b/habana_lazy/passes/fold_conv_batchnorm.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/fuse_bn_relu_residual_add.h b/habana_lazy/passes/fuse_bn_relu_residual_add.h
index fdce3b1..fc02e57 100644
--- a/habana_lazy/passes/fuse_bn_relu_residual_add.h
+++ b/habana_lazy/passes/fuse_bn_relu_residual_add.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/fuse_mm_transpose.h b/habana_lazy/passes/fuse_mm_transpose.h
index 0372cd1..a8bb17e 100644
--- a/habana_lazy/passes/fuse_mm_transpose.h
+++ b/habana_lazy/passes/fuse_mm_transpose.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/fuse_strided_views.h b/habana_lazy/passes/fuse_strided_views.h
index 6beb73d..34c99f9 100644
--- a/habana_lazy/passes/fuse_strided_views.h
+++ b/habana_lazy/passes/fuse_strided_views.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2024 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/pass_utils.h b/habana_lazy/passes/pass_utils.h
index bc1dac0..f51f7dd 100644
--- a/habana_lazy/passes/pass_utils.h
+++ b/habana_lazy/passes/pass_utils.h
@@ -1,15 +1,17 @@
-/*******************************************************************************
- * Copyright (C) 2020-2024 Habana Labs, Ltd. an Intel Company
- * All Rights Reserved.
- *
- * Unauthorized copying of this file or any element(s) within it, via any medium
- * is strictly prohibited.
- * This file contains Habana Labs, Ltd. proprietary and confidential information
- * and is subject to the confidentiality and license agreements under which it
- * was provided.
- *
- *******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/recalculate_batchnorm_params.h b/habana_lazy/passes/recalculate_batchnorm_params.h
index bbee383..ad16372 100644
--- a/habana_lazy/passes/recalculate_batchnorm_params.h
+++ b/habana_lazy/passes/recalculate_batchnorm_params.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/replace_inplace_ops.h b/habana_lazy/passes/replace_inplace_ops.h
index e4ddfe0..0f2cacd 100644
--- a/habana_lazy/passes/replace_inplace_ops.h
+++ b/habana_lazy/passes/replace_inplace_ops.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/replace_views_with_reshapes.h b/habana_lazy/passes/replace_views_with_reshapes.h
index 2777238..458c9c1 100644
--- a/habana_lazy/passes/replace_views_with_reshapes.h
+++ b/habana_lazy/passes/replace_views_with_reshapes.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
diff --git a/habana_lazy/passes/transform_graph.h b/habana_lazy/passes/transform_graph.h
index 8a2e050..1c4d2f9 100644
--- a/habana_lazy/passes/transform_graph.h
+++ b/habana_lazy/passes/transform_graph.h
@@ -1,12 +1,17 @@
-/******************************************************************************
- * Copyright (C) 2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- * Unauthorized copying of this file, via any medium is strictly prohibited.
- * Proprietary and confidential.
- *
- ******************************************************************************
- */
+/**
+* Copyright (c) 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
 
 #pragma once
 
