From edb122d8df6bfa4dfb620c0e18ecf4f08b7d905c Mon Sep 17 00:00:00 2001
From: grabinovich <grabinovich@habana.ai>
Date: Thu, 07 Nov 2024 16:46:54 +0200
Subject: [PATCH] add random flags

Change-Id: I8c39d1e25646cacc3aa87473bce765eabce194e0
---

diff --git a/h15b/h15b_job_params_olilo.groovy b/h15b/h15b_job_params_olilo.groovy
index 92e43c4..4167fee 100644
--- a/h15b/h15b_job_params_olilo.groovy
+++ b/h15b/h15b_job_params_olilo.groovy
@@ -8,7 +8,7 @@
                 'BRANCH_NAME': 'h15b_mesh',
                 'BLOCK_NAME': 'fc_logic',
                 'PATH_NAME': 'mesh_ic_verif',
-                'ADDITIONAL_RUN_PARAMS': '-stub_defs "+PROJ_H15B+SVT_CXL_DISABLE_LPIF+SVT_CXL_SUBSYSTEM_DISABLE_PCIE_VIP" -af "+define+HBW_DW_SEL_1+HBW_DW_SEL_0+DISABLE_MSTR_IF_COVERAGE+MSTR_IF_BLOCK_LEVEL+LBW_MSG_EN+PROJ_H15B+DSCNCT_SCBD"',
+                'ADDITIONAL_RUN_PARAMS': '-stub_defs "+PROJ_H15B+SVT_CXL_DISABLE_LPIF+SVT_CXL_SUBSYSTEM_DISABLE_PCIE_VIP" -af "+define+HBW_DW_SEL_1+HBW_DW_SEL_0+DISABLE_MSTR_IF_COVERAGE+MSTR_IF_BLOCK_LEVEL+LBW_MSG_EN+PROJ_H15B+DSCNCT_SCBD" -af_random "+define "define+AXI_HBW_TERMINATE+define+ASYNC_ISO_EN"',
                 'MAILING_LIST': 'olilo@habana.ai grabinovich@habana.ai',
                 'CSV_FILE': 'mesh_ic_verif/fc_logic_verif/tb/tests/fc_logic_regression.csv',
                 'BUILD_PERIODICALLY': 'H 23 * * *',
