From bbeb6f8a81da93b7c663dc16b988c7b2eecf96f3 Mon Sep 17 00:00:00 2001
From: grabinovich <grabinovich@habana.ai>
Date: Thu, 07 Nov 2024 16:02:04 +0200
Subject: [PATCH] new groovy grabinovich

Change-Id: Id81e5dc2dcbc248f21f5fdc38421cdd2ce5cf291
---

diff --git a/h15b/h15b_job_params_grabinovich.groovy b/h15b/h15b_job_params_grabinovich.groovy
new file mode 100644
index 0000000..9e7948f
--- /dev/null
+++ b/h15b/h15b_job_params_grabinovich.groovy
@@ -0,0 +1,26 @@
+def JobsConfig() {
+    def jobsConfig = [
+            'h15b_fc_logic_regression': [
+            'jobDescription': 'Regression environment for  fc_logic - random',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15b',
+                'BRANCH_NAME': 'h15b_mesh',
+                'BLOCK_NAME': 'fc_logic',
+                'PATH_NAME': 'mesh_ic_verif',
+                'ADDITIONAL_RUN_PARAMS': '-stub_defs "+PROJ_H15B+SVT_CXL_DISABLE_LPIF+SVT_CXL_SUBSYSTEM_DISABLE_PCIE_VIP" -af "+define+HBW_DW_SEL_1+HBW_DW_SEL_0+DISABLE_MSTR_IF_COVERAGE+MSTR_IF_BLOCK_LEVEL+LBW_MSG_EN+PROJ_H15B+DSCNCT_SCBD+AXI_HBW_TERMINATE"',
+                'MAILING_LIST': 'olilo@habana.ai grabinovich@habana.ai',
+                'CSV_FILE': 'mesh_ic_verif/fc_logic_verif/tb/tests/fc_logic_regression.csv',
+                'BUILD_PERIODICALLY': 'H 23 * * *',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_LOW -verdi -lte 2 ',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ]
+
+
+
+    ]
+    return jobsConfig
+}
