From a1e344081d11d878ea40196fb79d22c4a88d44ce Mon Sep 17 00:00:00 2001
From: grabinovich <grabinovich@habana.ai>
Date: Thu, 07 Nov 2024 17:14:32 +0200
Subject: [PATCH] add a random flag, edited

Change-Id: I8266d5bea356d848d387a847db2f2410b522ea3a
---

diff --git a/h15b/h15b_job_params_olilo.groovy b/h15b/h15b_job_params_olilo.groovy
index 4167fee..7e82c84 100644
--- a/h15b/h15b_job_params_olilo.groovy
+++ b/h15b/h15b_job_params_olilo.groovy
@@ -8,7 +8,7 @@
                 'BRANCH_NAME': 'h15b_mesh',
                 'BLOCK_NAME': 'fc_logic',
                 'PATH_NAME': 'mesh_ic_verif',
-                'ADDITIONAL_RUN_PARAMS': '-stub_defs "+PROJ_H15B+SVT_CXL_DISABLE_LPIF+SVT_CXL_SUBSYSTEM_DISABLE_PCIE_VIP" -af "+define+HBW_DW_SEL_1+HBW_DW_SEL_0+DISABLE_MSTR_IF_COVERAGE+MSTR_IF_BLOCK_LEVEL+LBW_MSG_EN+PROJ_H15B+DSCNCT_SCBD" -af_random "+define "define+AXI_HBW_TERMINATE+define+ASYNC_ISO_EN"',
+                'ADDITIONAL_RUN_PARAMS': '-stub_defs "+PROJ_H15B+SVT_CXL_DISABLE_LPIF+SVT_CXL_SUBSYSTEM_DISABLE_PCIE_VIP" -af "+define+HBW_DW_SEL_1+HBW_DW_SEL_0+DISABLE_MSTR_IF_COVERAGE+MSTR_IF_BLOCK_LEVEL+LBW_MSG_EN+PROJ_H15B+DSCNCT_SCBD" -af_random "define+AXI_HBW_TERMINATE"',
                 'MAILING_LIST': 'olilo@habana.ai grabinovich@habana.ai',
                 'CSV_FILE': 'mesh_ic_verif/fc_logic_verif/tb/tests/fc_logic_regression.csv',
                 'BUILD_PERIODICALLY': 'H 23 * * *',
