From 67681b77058eadb3ef2490a7b0b58d6138fafc09 Mon Sep 17 00:00:00 2001
From: Bartlomiej Stemborowski <bstemborowskix@habana.ai>
Date: Fri, 08 Nov 2024 16:48:35 +0200
Subject: [PATCH] [SW-203886][SW-203887][SW-206351] Unskip pt2e tests blocked due to upgrade

Change-Id: I18c9c037c3325146b6c25a980ff333d649faf831
---

diff --git a/tests/pytest_working/compile/test_pt2e_quant_flow.py b/tests/pytest_working/compile/test_pt2e_quant_flow.py
index 06be343..f4a108a 100644
--- a/tests/pytest_working/compile/test_pt2e_quant_flow.py
+++ b/tests/pytest_working/compile/test_pt2e_quant_flow.py
@@ -25,7 +25,6 @@
     habana_quantizer,
 )
 from habana_frameworks.torch.utils.debug.dynamo_utils import FxGraphAnalyzer
-from packaging.version import Version
 from test_utils import fga_assert_helper, is_gaudi1
 from torch.ao.quantization.observer import MinMaxObserver
 from torch.ao.quantization.qconfig import _ObserverOrFakeQuantizeConstructor
@@ -200,9 +199,6 @@
 @pytest.mark.parametrize("use_graph_break", [False, True])
 @pytest.mark.parametrize("pass_input_during_export", [False, True])
 def test_pt2e_quant_float(set_env_variable, test_case, quant_dtype, use_graph_break, pass_input_during_export):
-    if Version(Version(torch.__version__).base_version) >= Version("2.5"):
-        if not (use_graph_break == False and pass_input_during_export == True):
-            pytest.xfail("SW-203886")
 
     quantizer = habana_quantizer()
     quant_config = habana_quant_config_symmetric(quant_dtype)
@@ -236,9 +232,6 @@
 @pytest.mark.parametrize("use_graph_break", [False, True])
 @pytest.mark.parametrize("pass_input_during_export", [False, True])
 def test_pt2e_quant_int(test_case, quant_dtype, use_graph_break, pass_input_during_export):
-    if Version(Version(torch.__version__).base_version) >= Version("2.5"):
-        if not (use_graph_break == False and pass_input_during_export == True):
-            pytest.xfail("SW-203887")
 
     class custom_quantizer(Quantizer):
 
