From a5e24f2d13d6966847d42b68f792c9e1d0db1e18 Mon Sep 17 00:00:00 2001
From: Tobiasz Domagala <tdomagala@habana.ai>
Date: Fri, 08 Nov 2024 12:41:46 +0100
Subject: [PATCH] [SW-0] fix formatting causing review issues

Change-Id: Ice384e1427c175a1fc1d74616773142016c3ac4e
---

diff --git a/python_packages/habana_frameworks/torch/jit/csrc/jit_fork/ir/ir.cpp b/python_packages/habana_frameworks/torch/jit/csrc/jit_fork/ir/ir.cpp
index ae91150..2548695 100644
--- a/python_packages/habana_frameworks/torch/jit/csrc/jit_fork/ir/ir.cpp
+++ b/python_packages/habana_frameworks/torch/jit/csrc/jit_fork/ir/ir.cpp
@@ -227,6 +227,13 @@
       // TODO(@anjali411): fix this
       HABANA_ASSERT(false);
       break;
+    // todo SW-199903
+    // case AttributeKind::b:
+    //   printAttribute(out, b(name));
+    //   break;
+    // case AttributeKind::bs:
+    //   printAttribute(out, bs(name));
+    //   break;
     case AttributeKind::f:
       printAttribute(out, f(name));
       break;
@@ -666,1582 +673,1600 @@
       for (auto kv : anticipated_uses) {
         HABANA_ASSERT(kv.second == -1);
       }
-        HABANA_ASSERT(std::includes(ALL_OF(sum_set), ALL_OF(all_nodes_set)));
-      }
-    };
-    LintImpl(*this).check_graph();
+      HABANA_ASSERT(std::includes(ALL_OF(sum_set), ALL_OF(all_nodes_set)));
+    }
+  };
+  LintImpl(*this).check_graph();
+}
+
+void Graph::dump() const {
+  std::cout << *this << "\n";
+}
+
+void Graph::push_scope(const std::string& scope_name) {
+  current_scope_ = current_scope_->push(Symbol::scope(scope_name));
+  Node* block_node = insertNode(create(prim::TracedModuleForward, 0));
+  block_node->s_(attr::scope, scope_name);
+  Block* b = block_node->addBlock();
+  setInsertPoint(b);
+}
+void Graph::pop_scope() {
+  current_scope_ = current_scope_->parent();
+  if (insertPoint()->owningBlock()->owningNode()->kind() ==
+      prim::TracedModuleForward) {
+    setInsertPoint(insertPoint()->owningBlock()->owningNode()->next());
+  }
+}
+
+void LintGraph(const std::shared_ptr<Graph>& graph) {
+  graph->lint();
+}
+
+Block::Block(Graph* graph_, Node* node_)
+    : graph_(graph_),
+      output_(graph_->create(prim::Return, 0)),
+      input_(graph_->create(prim::Param, 0)),
+      owning_node_(node_) {
+  input_->next() = output_;
+  input_->prev() = output_;
+  output_->next() = input_;
+  output_->prev() = input_;
+
+  graph_->all_blocks.emplace(this);
+  output_->owning_block_ = this;
+  output_->topo_position_ = kUpperBound;
+  input_->owning_block_ = this;
+  input_->topo_position_ = kLowerBound;
+}
+
+void Block::reIndexTopology() {
+  auto curPos = kLowerBound;
+  for (auto node : nodes()) {
+    HABANA_ASSERT(curPos <= (kUpperBound - kAppendInterval));
+    curPos += kAppendInterval;
+    node->topo_position_ = curPos;
+  }
+}
+
+void Block::cloneFrom(Block* src, std::function<Value*(Value*)> value_map) {
+  std::unordered_map<Value*, Value*> local_map;
+  auto env = [&](Value* v) {
+    auto it = local_map.find(v);
+    if (it != local_map.end()) {
+      return it->second;
+    }
+    return value_map(v);
+  };
+
+  auto graph = owningGraph();
+  for (auto input : src->inputs()) {
+    local_map[input] = this->addInput()->copyMetadata(input);
   }
 
-  void
-  Graph::dump() const {
-    std::cout << *this << "\n";
-  }
-
-  void Graph::push_scope(const std::string& scope_name) {
-    current_scope_ = current_scope_->push(Symbol::scope(scope_name));
-    Node* block_node = insertNode(create(prim::TracedModuleForward, 0));
-    block_node->s_(attr::scope, scope_name);
-    Block* b = block_node->addBlock();
-    setInsertPoint(b);
-  }
-  void Graph::pop_scope() {
-    current_scope_ = current_scope_->parent();
-    if (insertPoint()->owningBlock()->owningNode()->kind() ==
-        prim::TracedModuleForward) {
-      setInsertPoint(insertPoint()->owningBlock()->owningNode()->next());
+  for (auto node : src->nodes()) {
+    auto new_node = this->appendNode(graph->createClone(node, env));
+    for (size_t i = 0; i < node->outputs().size(); ++i) {
+      auto oo = node->outputs()[i];
+      auto no = new_node->outputs()[i];
+      local_map[oo] = no;
+      no->copyMetadata(oo);
     }
   }
-
-  void LintGraph(const std::shared_ptr<Graph>& graph) {
-    graph->lint();
+  for (auto output : src->outputs()) {
+    this->registerOutput(env(output));
   }
+}
 
-  Block::Block(Graph * graph_, Node * node_)
-      : graph_(graph_),
-        output_(graph_->create(prim::Return, 0)),
-        input_(graph_->create(prim::Param, 0)),
-        owning_node_(node_) {
-    input_->next() = output_;
-    input_->prev() = output_;
-    output_->next() = input_;
-    output_->prev() = input_;
-
-    graph_->all_blocks.emplace(this);
-    output_->owning_block_ = this;
-    output_->topo_position_ = kUpperBound;
-    input_->owning_block_ = this;
-    input_->topo_position_ = kLowerBound;
+void Block::destroy() {
+  // we cannot destroy the output because it is used as the sentinel
+  // for the nodes() list and has to remain valid for the loop
+  output_->removeAllInputs();
+  for (auto it = this->nodes().reverse().begin(),
+            end = this->nodes().reverse().end();
+       it != end;
+       ++it) {
+    it.destroyCurrent();
   }
+  output_->destroy();
+  input_->destroy();
+  graph_->freeBlock(this);
+}
 
-  void Block::reIndexTopology() {
-    auto curPos = kLowerBound;
-    for (auto node : nodes()) {
-      HABANA_ASSERT(curPos <= (kUpperBound - kAppendInterval));
-      curPos += kAppendInterval;
-      node->topo_position_ = curPos;
+void Graph::cloneFrom(Graph& src) {
+  auto env = [](Value* v) -> Value* {
+    AT_ERROR(
+        "Graph::copy() encountered a use of a value " + v->debugName() +
+        " not in scope. Run lint!");
+  };
+  block()->cloneFrom(src.block(), env);
+}
+
+std::shared_ptr<Graph> Graph::copy() {
+  auto new_g = std::make_shared<Graph>();
+  new_g->cloneFrom(*this);
+  return new_g;
+}
+
+std::unique_ptr<Graph> Graph::copyUnique() {
+  auto new_g = std::make_unique<Graph>();
+  new_g->cloneFrom(*this);
+  return new_g;
+}
+
+void Block::remapTypes(
+    const std::function<TypeWrapper(TypeWrapper)>& type_map) {
+  for (Value* input : inputs()) {
+    input->setType(type_map(input->typeWrapper()));
+  }
+  for (Node* node : nodes()) {
+    for (Value* output : node->outputs()) {
+      output->setType(type_map(output->typeWrapper()));
     }
-  }
-
-  void Block::cloneFrom(Block * src, std::function<Value*(Value*)> value_map) {
-    std::unordered_map<Value*, Value*> local_map;
-    auto env = [&](Value* v) {
-      auto it = local_map.find(v);
-      if (it != local_map.end()) {
-        return it->second;
-      }
-      return value_map(v);
-    };
-
-    auto graph = owningGraph();
-    for (auto input : src->inputs()) {
-      local_map[input] = this->addInput()->copyMetadata(input);
+    for (Block* sub_block : node->blocks()) {
+      sub_block->remapTypes(type_map);
     }
-
-    for (auto node : src->nodes()) {
-      auto new_node = this->appendNode(graph->createClone(node, env));
-      for (size_t i = 0; i < node->outputs().size(); ++i) {
-        auto oo = node->outputs()[i];
-        auto no = new_node->outputs()[i];
-        local_map[oo] = no;
-        no->copyMetadata(oo);
-      }
-    }
-    for (auto output : src->outputs()) {
-      this->registerOutput(env(output));
-    }
-  }
-
-  void Block::destroy() {
-    // we cannot destroy the output because it is used as the sentinel
-    // for the nodes() list and has to remain valid for the loop
-    output_->removeAllInputs();
-    for (auto it = this->nodes().reverse().begin(),
-              end = this->nodes().reverse().end();
-         it != end;
-         ++it) {
-      it.destroyCurrent();
-    }
-    output_->destroy();
-    input_->destroy();
-    graph_->freeBlock(this);
-  }
-
-  void Graph::cloneFrom(Graph & src) {
-    auto env = [](Value* v) -> Value* {
-      AT_ERROR(
-          "Graph::copy() encountered a use of a value " + v->debugName() +
-          " not in scope. Run lint!");
-    };
-    block()->cloneFrom(src.block(), env);
-  }
-
-  std::shared_ptr<Graph> Graph::copy() {
-    auto new_g = std::make_shared<Graph>();
-    new_g->cloneFrom(*this);
-    return new_g;
-  }
-
-  std::unique_ptr<Graph> Graph::copyUnique() {
-    auto new_g = std::make_unique<Graph>();
-    new_g->cloneFrom(*this);
-    return new_g;
-  }
-
-  void Block::remapTypes(
-      const std::function<TypeWrapper(TypeWrapper)>& type_map) {
-    for (Value* input : inputs()) {
-      input->setType(type_map(input->typeWrapper()));
-    }
-    for (Node* node : nodes()) {
-      for (Value* output : node->outputs()) {
-        output->setType(type_map(output->typeWrapper()));
-      }
-      for (Block* sub_block : node->blocks()) {
-        sub_block->remapTypes(type_map);
-      }
-      for (Symbol name : node->attributeNames()) {
-        if (node->kindOf(name) == AttributeKind::g) {
-          node->g(name)->remapTypes(type_map);
-        } else if (node->kindOf(name) == AttributeKind::gs) {
-          for (const auto& g : node->gs(name)) {
-            g->remapTypes(type_map);
-          }
+    for (Symbol name : node->attributeNames()) {
+      if (node->kindOf(name) == AttributeKind::g) {
+        node->g(name)->remapTypes(type_map);
+      } else if (node->kindOf(name) == AttributeKind::gs) {
+        for (const auto& g : node->gs(name)) {
+          g->remapTypes(type_map);
         }
       }
     }
   }
+}
 
-  void Graph::remapTypes(
-      const std::function<TypeWrapper(TypeWrapper)>& type_map) {
-    block()->remapTypes(type_map);
+void Graph::remapTypes(
+    const std::function<TypeWrapper(TypeWrapper)>& type_map) {
+  block()->remapTypes(type_map);
+}
+
+void Value::inferTypeFrom(const at::Tensor& output) {
+  setType(TensorType::create(output));
+}
+
+void Value::inferTypeFrom(
+    const c10::intrusive_ptr<c10::ivalue::Object>& output) {
+  setType(output->type());
+}
+
+bool Value::mustBeNone() const {
+  return type()->cast<NoneType>() || node_->mustBeNone();
+}
+bool Value::mustNotBeNone() const {
+  return node_->kind() != prim::AutogradAdd && type() != NoneType::get() &&
+      !type()->cast<OptionalType>() &&
+      !(type()->cast<UnionType>() &&
+        type()->expect<UnionType>()->canHoldType(*NoneType::get()));
+}
+
+std::string Value::debugNameBase() const {
+  std::string name = debugName();
+  std::string name_base = name;
+  auto last_dot_pos = name.find_last_of('.');
+  if (last_dot_pos != std::string::npos && last_dot_pos + 1 != name.size()) {
+    if (name.find_first_not_of("0123456789", last_dot_pos + 1) ==
+        std::string::npos) {
+      name_base = name.substr(0, last_dot_pos);
+    }
+  }
+  return name_base;
+}
+
+bool Value::isValidName(const std::string& name, bool allow_numbers) {
+  // Empty strings are legal
+  if (name.empty()) {
+    return true;
   }
 
-  void Value::inferTypeFrom(const at::Tensor& output) {
-    setType(TensorType::create(output));
+  // Numbers are not legal
+  if (isNumber(name) && !allow_numbers) {
+    return false;
   }
 
-  void Value::inferTypeFrom(
-      const c10::intrusive_ptr<c10::ivalue::Object>& output) {
-    setType(output->type());
+  return true;
+}
+
+Value* Value::setDebugName(const std::string& name, bool allow_numbers) {
+  if (!isValidName(name, allow_numbers)) {
+    throw std::runtime_error("Invalid name: '" + name + "'");
   }
 
-  bool Value::mustBeNone() const {
-    return type()->cast<NoneType>() || node_->mustBeNone();
-  }
-  bool Value::mustNotBeNone() const {
-    return node_->kind() != prim::AutogradAdd && type() != NoneType::get() &&
-        !type()->cast<OptionalType>() &&
-        !(type()->cast<UnionType>() &&
-          type()->expect<UnionType>()->canHoldType(*NoneType::get()));
+  auto& names = node()->owningGraph()->unique_names_;
+
+  // clear any old name from the map
+  if (hasDebugName()) {
+    names.erase(unique_name_);
+    unique_name_ = "";
   }
 
-  std::string Value::debugNameBase() const {
-    std::string name = debugName();
+  // allow "" to clear the uniquename
+  if (name.empty()) {
+    return this;
+  }
+
+  // if someone else has this name, then rename the other value
+  auto old_owner_of_name = names.find(name);
+  if (old_owner_of_name != names.end()) {
+    size_t suffix = 1;
     std::string name_base = name;
     auto last_dot_pos = name.find_last_of('.');
     if (last_dot_pos != std::string::npos && last_dot_pos + 1 != name.size()) {
       if (name.find_first_not_of("0123456789", last_dot_pos + 1) ==
           std::string::npos) {
+        suffix = std::stoll(name.substr(last_dot_pos + 1));
         name_base = name.substr(0, last_dot_pos);
       }
     }
-    return name_base;
+
+    auto& names_suffixes = node()->owningGraph()->name_base_suffix_;
+    auto it = names_suffixes.find(name_base);
+    if (it != names_suffixes.end()) {
+      suffix = std::max(suffix, it->second + 1);
+    }
+
+    // Verify that new name is not used and find next usable name in case
+    // suffix is used.
+    std::string replacement_name;
+    do {
+      std::stringstream ss;
+#ifndef _WIN32
+      // Protect 12345 integer from becoming "1,2345" if some other process
+      // sets global locale For more details see
+      // https://github.com/pytorch/pytorch/issues/79583#issuecomment-1161260061
+      static std::locale c_locale("C");
+      ss.imbue(c_locale);
+#endif
+      ss << name_base << "." << suffix++;
+      replacement_name = ss.str();
+    } while (names.count(replacement_name) > 0);
+
+    names_suffixes[name_base] = suffix;
+
+    old_owner_of_name->second->setDebugName(replacement_name);
   }
 
-  bool Value::isValidName(const std::string& name, bool allow_numbers) {
-    // Empty strings are legal
-    if (name.empty()) {
+  names[name] = this;
+  unique_name_ = name;
+  return this;
+}
+
+Value* Value::copyMetadata(Value* from) {
+  setType(from->typeWrapper());
+  if (from->hasDebugName()) {
+    setDebugName(from->debugName());
+  }
+  return this;
+}
+
+void Value::replaceFirstUseWith(Value* newValue) {
+  HABANA_ASSERT(owningGraph() == newValue->owningGraph());
+  auto u = uses()[0];
+  u.user->inputs_[u.offset] = newValue;
+  newValue->uses_.push_back(u);
+  uses_.erase(uses_.begin());
+}
+
+void Value::replaceAllUsesWith(Value* newValue) {
+  while (!uses().empty()) {
+    replaceFirstUseWith(newValue);
+  }
+}
+
+void Value::replaceAllUsesAfterNodeWith(const Node* node, Value* newValue) {
+  std::for_each(uses_.begin(), uses_.end(), [&node, newValue](Use& u) {
+    if (u.user->isAfter(node)) {
+      u.user->inputs_[u.offset] = newValue;
+      newValue->uses_.push_back(u);
+    }
+  });
+
+  uses_.erase(
+      std::remove_if(
+          uses_.begin(),
+          uses_.end(),
+          [&node](const Use& u) { return u.user->isAfter(node); }),
+      uses_.end());
+}
+
+void Value::replaceAllUsesDominatedByNodeWith(
+    const Node* node,
+    Value* newValue) {
+  std::for_each(uses_.begin(), uses_.end(), [&node, newValue](Use& u) {
+    if (u.user->isDominatedBy(node)) {
+      u.user->inputs_[u.offset] = newValue;
+      newValue->uses_.push_back(u);
+    }
+  });
+
+  uses_.erase(
+      std::remove_if(
+          uses_.begin(),
+          uses_.end(),
+          [&node](const Use& u) { return u.user->isDominatedBy(node); }),
+      uses_.end());
+}
+
+static size_t findArgument(
+    const FunctionSchema& the_schema,
+    const std::string& unqualName) {
+  for (const auto i : c10::irange(the_schema.arguments().size())) {
+    const Argument* arg = &the_schema.arguments()[i];
+    if (arg->name() == unqualName) {
+      return i;
+    }
+  }
+  throw std::runtime_error(
+      std::string("Couldn't find an argument called ") + unqualName);
+}
+
+static size_t findArgument(const FunctionSchema& the_schema, Symbol name) {
+  const auto unqualName = name.toUnqualString();
+  return findArgument(the_schema, unqualName);
+}
+
+c10::optional<IValue> Node::get(Symbol name) const {
+  return toIValue(namedInput(name));
+}
+
+bool Node::hasNamedInput(const std::string& name) const {
+  for (const auto& argument : schema().arguments()) {
+    if (argument.name() == name) {
       return true;
     }
-
-    // Numbers are not legal
-    if (isNumber(name) && !allow_numbers) {
-      return false;
-    }
-
-    return true;
   }
+  return false;
+}
 
-  Value* Value::setDebugName(const std::string& name, bool allow_numbers) {
-    if (!isValidName(name, allow_numbers)) {
-      throw std::runtime_error("Invalid name: '" + name + "'");
-    }
+Value* Node::namedInput(const std::string& unqualName) const {
+  return input(findArgument(schema(), unqualName));
+}
+Value* Node::namedInput(Symbol name) const {
+  return input(findArgument(schema(), name));
+}
 
-    auto& names = node()->owningGraph()->unique_names_;
-
-    // clear any old name from the map
-    if (hasDebugName()) {
-      names.erase(unique_name_);
-      unique_name_ = "";
-    }
-
-    // allow "" to clear the uniquename
-    if (name.empty()) {
-      return this;
-    }
-
-    // if someone else has this name, then rename the other value
-    auto old_owner_of_name = names.find(name);
-    if (old_owner_of_name != names.end()) {
-      size_t suffix = 1;
-      std::string name_base = name;
-      auto last_dot_pos = name.find_last_of('.');
-      if (last_dot_pos != std::string::npos &&
-          last_dot_pos + 1 != name.size()) {
-        if (name.find_first_not_of("0123456789", last_dot_pos + 1) ==
-            std::string::npos) {
-          suffix = std::stoll(name.substr(last_dot_pos + 1));
-          name_base = name.substr(0, last_dot_pos);
-        }
-      }
-
-      auto& names_suffixes = node()->owningGraph()->name_base_suffix_;
-      auto it = names_suffixes.find(name_base);
-      if (it != names_suffixes.end()) {
-        suffix = std::max(suffix, it->second + 1);
-      }
-
-      // Verify that new name is not used and find next usable name in case
-      // suffix is used.
-      std::string replacement_name;
-      do {
-        std::stringstream ss;
-#ifndef _WIN32
-        // Protect 12345 integer from becoming "1,2345" if some other process
-        // sets global locale For more details see
-        // https://github.com/pytorch/pytorch/issues/79583#issuecomment-1161260061
-        static std::locale c_locale("C");
-        ss.imbue(c_locale);
-#endif
-        ss << name_base << "." << suffix++;
-        replacement_name = ss.str();
-      } while (names.count(replacement_name) > 0);
-
-      names_suffixes[name_base] = suffix;
-
-      old_owner_of_name->second->setDebugName(replacement_name);
-    }
-
-    names[name] = this;
-    unique_name_ = name;
-    return this;
+bool Node::matches(const FunctionSchema& schema) const {
+  if (isBlockListedSchema(schema)) {
+    return false;
   }
-
-  Value* Value::copyMetadata(Value * from) {
-    setType(from->typeWrapper());
-    if (from->hasDebugName()) {
-      setDebugName(from->debugName());
-    }
-    return this;
+  // wrong name
+  if (kind().toQualString() != schema.name()) {
+    return false;
   }
+  at::ArrayRef<const Value*> actuals = inputs();
+  const auto& formals = schema.arguments();
 
-  void Value::replaceFirstUseWith(Value * newValue) {
-    HABANA_ASSERT(owningGraph() == newValue->owningGraph());
-    auto u = uses()[0];
-    u.user->inputs_[u.offset] = newValue;
-    newValue->uses_.push_back(u);
-    uses_.erase(uses_.begin());
-  }
-
-  void Value::replaceAllUsesWith(Value * newValue) {
-    while (!uses().empty()) {
-      replaceFirstUseWith(newValue);
-    }
-  }
-
-  void Value::replaceAllUsesAfterNodeWith(const Node* node, Value* newValue) {
-    std::for_each(uses_.begin(), uses_.end(), [&node, newValue](Use& u) {
-      if (u.user->isAfter(node)) {
-        u.user->inputs_[u.offset] = newValue;
-        newValue->uses_.push_back(u);
-      }
-    });
-
-    uses_.erase(
-        std::remove_if(
-            uses_.begin(),
-            uses_.end(),
-            [&node](const Use& u) { return u.user->isAfter(node); }),
-        uses_.end());
-  }
-
-  void Value::replaceAllUsesDominatedByNodeWith(
-      const Node* node, Value* newValue) {
-    std::for_each(uses_.begin(), uses_.end(), [&node, newValue](Use& u) {
-      if (u.user->isDominatedBy(node)) {
-        u.user->inputs_[u.offset] = newValue;
-        newValue->uses_.push_back(u);
-      }
-    });
-
-    uses_.erase(
-        std::remove_if(
-            uses_.begin(),
-            uses_.end(),
-            [&node](const Use& u) { return u.user->isDominatedBy(node); }),
-        uses_.end());
-  }
-
-  static size_t findArgument(
-      const FunctionSchema& the_schema, const std::string& unqualName) {
-    for (const auto i : c10::irange(the_schema.arguments().size())) {
-      const Argument* arg = &the_schema.arguments()[i];
-      if (arg->name() == unqualName) {
-        return i;
-      }
-    }
-    throw std::runtime_error(
-        std::string("Couldn't find an argument called ") + unqualName);
-  }
-
-  static size_t findArgument(const FunctionSchema& the_schema, Symbol name) {
-    const auto unqualName = name.toUnqualString();
-    return findArgument(the_schema, unqualName);
-  }
-
-  c10::optional<IValue> Node::get(Symbol name) const {
-    return toIValue(namedInput(name));
-  }
-
-  bool Node::hasNamedInput(const std::string& name) const {
-    for (const auto& argument : schema().arguments()) {
-      if (argument.name() == name) {
-        return true;
-      }
-    }
+  // not enough inputs
+  if (actuals.size() < formals.size()) {
     return false;
   }
 
-  Value* Node::namedInput(const std::string& unqualName) const {
-    return input(findArgument(schema(), unqualName));
-  }
-  Value* Node::namedInput(Symbol name) const {
-    return input(findArgument(schema(), name));
-  }
-
-  bool Node::matches(const FunctionSchema& schema) const {
-    if (isBlockListedSchema(schema)) {
-      return false;
-    }
-    // wrong name
-    if (kind().toQualString() != schema.name()) {
-      return false;
-    }
-    at::ArrayRef<const Value*> actuals = inputs();
-    const auto& formals = schema.arguments();
-
-    // not enough inputs
-    if (actuals.size() < formals.size()) {
+  TypeEnv type_env;
+  for (const auto i : c10::irange(formals.size())) {
+    auto formal = formals[i].type();
+    const MatchTypeReturn matched_type =
+        matchTypeVariables(formal, actuals[i]->type(), type_env);
+    if (!matched_type.success()) {
       return false;
     }
 
-    TypeEnv type_env;
-    for (const auto i : c10::irange(formals.size())) {
-      auto formal = formals[i].type();
-      const MatchTypeReturn matched_type =
-          matchTypeVariables(formal, actuals[i]->type(), type_env);
-      if (!matched_type.success()) {
-        return false;
-      }
-
-      TypePtr resolved = tryEvalTypeVariables(formal, type_env);
-      if (resolved) {
-        formal = resolved;
-      }
-      // note: it is possible at this point that type variable matching has
-      // not resolved all type variables, e.g. if None was matched to
-      // Optional[T] we will not succeed at matching T. However None <:
-      // Optional[T] so this check can still succeed.
-
-      if (!matchTypes(actuals[i]->type(), formal)) {
-        return false;
-      }
+    TypePtr resolved = tryEvalTypeVariables(formal, type_env);
+    if (resolved) {
+      formal = resolved;
     }
+    // note: it is possible at this point that type variable matching has
+    // not resolved all type variables, e.g. if None was matched to
+    // Optional[T] we will not succeed at matching T. However None <:
+    // Optional[T] so this check can still succeed.
 
-    // too many inputs
-    if (!schema.is_vararg() && actuals.size() != formals.size()) {
+    if (!matchTypes(actuals[i]->type(), formal)) {
       return false;
     }
-
-    return true;
   }
 
-  bool Node::matches(
-      const char* signature_literal, at::ArrayRef<Symbol> const_inputs) const {
-    if (!matches(
-            torch::jit::getOperatorForLiteral(signature_literal)->schema())) {
+  // too many inputs
+  if (!schema.is_vararg() && actuals.size() != formals.size()) {
+    return false;
+  }
+
+  return true;
+}
+
+bool Node::matches(
+    const char* signature_literal,
+    at::ArrayRef<Symbol> const_inputs) const {
+  if (!matches(
+          torch::jit::getOperatorForLiteral(signature_literal)->schema())) {
+    return false;
+  }
+  for (Symbol s : const_inputs) {
+    if (!is_constant(s)) {
       return false;
     }
-    for (Symbol s : const_inputs) {
-      if (!is_constant(s)) {
-        return false;
-      }
-    }
-    return true;
   }
+  return true;
+}
 
-  bool Node::mustBeNone() const {
-    // We can statically deduce this Node has returning None if:
-    return
-        // It's an AutogradZero node, or ...
-        kind_ == prim::AutogradZero ||
-        // It has only one output and that output is NoneType, or ...
-        (outputs().size() == 1 && output()->type() == NoneType::get()) ||
-        // It's a constant optional with no value in the attributes.
-        (kind_ == prim::Constant && !this->hasAttributes() &&
-         output()->type()->cast<OptionalType>());
+bool Node::mustBeNone() const {
+  // We can statically deduce this Node has returning None if:
+  return
+      // It's an AutogradZero node, or ...
+      kind_ == prim::AutogradZero ||
+      // It has only one output and that output is NoneType, or ...
+      (outputs().size() == 1 && output()->type() == NoneType::get()) ||
+      // It's a constant optional with no value in the attributes.
+      (kind_ == prim::Constant && !this->hasAttributes() &&
+       output()->type()->cast<OptionalType>());
+}
+
+void Node::dump() const {
+  std::cout << *this << "\n";
+}
+
+// todo SW-199903
+// void Node::setOperatorName(const c10::OperatorName& operator_name) {
+//   operator_name_ = operator_name;
+// }
+
+// const c10::OperatorName& Node::getOperatorName() const {
+//   if (!operator_name_) {
+//     return schema().operator_name();
+//   }
+
+//   return *operator_name_;
+// }
+
+const FunctionSchema& Node::schema() const {
+  if (op_) {
+    return op_->schema();
   }
+  return getOperator().schema();
+}
 
-  void Node::dump() const {
-    std::cout << *this << "\n";
+const FunctionSchema* Node::maybeSchema() const {
+  if (auto op = maybeOperator()) {
+    return &op->schema();
   }
+  return nullptr;
+}
 
-  const FunctionSchema& Node::schema() const {
-    if (op_) {
-      return op_->schema();
-    }
-    return getOperator().schema();
-  }
-
-  const FunctionSchema* Node::maybeSchema() const {
-    if (auto op = maybeOperator()) {
-      return &op->schema();
-    }
-    return nullptr;
-  }
-
-  const torch::jit::Operator* Node::maybeOperator() const {
-    if (!op_) {
-      const auto& candidates = torch::jit::getAllOperatorsFor(kind());
-      for (const auto& candidate : candidates) {
-        if (matches(candidate->schema())) {
-          op_ = candidate.get();
-          break;
-        }
-      }
-    }
-    return op_;
-  }
-
-  const torch::jit::Operator& Node::getOperator() const {
-    const torch::jit::Operator* maybe = maybeOperator();
-    if (maybe) {
-      return *maybe;
-    }
-
-    auto er = ErrorReport(sourceRange());
-    er << "Schema not found for node. File a bug report.\n";
-    er << "Node: " << *this << "\n";
-    er << "Input types:";
-    for (const auto i : c10::irange(inputs().size())) {
-      if (i > 0)
-        er << ", ";
-      er << *inputs()[i]->type();
-    }
+const torch::jit::Operator* Node::maybeOperator() const {
+  if (!op_) {
     const auto& candidates = torch::jit::getAllOperatorsFor(kind());
-    if (!candidates.empty()) {
-      er << "\ncandidates were:\n";
-      for (auto& candidate : candidates) {
-        er << "  " << candidate->schema() << "\n";
-      }
-    } else {
-      er << "\nno candidates found\n";
-    }
-    er << "within the graph:\n";
-    er << *owningGraph() << "\n";
-    throw er;
-  }
-
-  bool Node::isNondeterministic() const {
-    const auto schema = maybeSchema();
-    if (!kind().is_aten()) {
-      return false;
-    }
-    // All aten ops are expecte to have a schema. However this is left as a
-    // warning instead of an assert to ensure that previous use cases do not
-    // break.
-    if (!schema) {
-      TORCH_WARN("aten Schema not found.");
-      return false;
-    }
-    torch::utils::SchemaInfo schema_info(*schema);
-    if (hasNamedInput("train")) {
-      auto value = constant_as<bool>(namedInput("train"));
-      if (value.has_value()) {
-        schema_info.addArgumentValue("train", *value);
+    for (const auto& candidate : candidates) {
+      if (matches(candidate->schema())) {
+        op_ = candidate.get();
+        break;
       }
     }
-    return schema_info.is_nondeterministic();
+  }
+  return op_;
+}
+
+const torch::jit::Operator& Node::getOperator() const {
+  const torch::jit::Operator* maybe = maybeOperator();
+  if (maybe) {
+    return *maybe;
   }
 
-  bool Node::hasSideEffects() const {
-    switch (kind_) {
-      case prim::PythonOp:
-      case prim::IgnoredPythonOp:
-      case prim::Print:
-      case prim::RaiseException:
-      case aten::warn:
-      case aten::save:
-      case aten::manual_seed:
-      case prim::AddStatValue:
-      case prim::TimePoint:
-      case prim::CallFunction:
-      case prim::CallMethod:
-      case prim::BailoutTemplate:
-      case prim::BailOut:
-      case prim::rpc_async: // It represents RPC message sent.
-      case prim::rpc_sync: // It represents RPC message sent.
-      case prim::rpc_remote: // It represents RPC message sent.
-      case aten::wait: // It can represent RPC message received.
+  auto er = ErrorReport(sourceRange());
+  er << "Schema not found for node. File a bug report.\n";
+  er << "Node: " << *this << "\n";
+  er << "Input types:";
+  for (const auto i : c10::irange(inputs().size())) {
+    if (i > 0)
+      er << ", ";
+    er << *inputs()[i]->type();
+  }
+  const auto& candidates = torch::jit::getAllOperatorsFor(kind());
+  if (!candidates.empty()) {
+    er << "\ncandidates were:\n";
+    for (auto& candidate : candidates) {
+      er << "  " << candidate->schema() << "\n";
+    }
+  } else {
+    er << "\nno candidates found\n";
+  }
+  er << "within the graph:\n";
+  er << *owningGraph() << "\n";
+  throw er;
+}
+
+bool Node::isNondeterministic() const {
+  const auto schema = maybeSchema();
+  if (!kind().is_aten()) {
+    return false;
+  }
+  // All aten ops are expecte to have a schema. However this is left as a
+  // warning instead of an assert to ensure that previous use cases do not
+  // break.
+  if (!schema) {
+    TORCH_WARN("aten Schema not found.");
+    return false;
+  }
+  torch::utils::SchemaInfo schema_info(*schema);
+  if (hasNamedInput("train")) {
+    auto value = constant_as<bool>(namedInput("train"));
+    if (value.has_value()) {
+      schema_info.addArgumentValue("train", *value);
+    }
+  }
+  return schema_info.is_nondeterministic();
+}
+
+bool Node::hasSideEffects() const {
+  switch (kind_) {
+    case prim::PythonOp:
+    case prim::IgnoredPythonOp:
+    case prim::Print:
+    case prim::RaiseException:
+    case aten::warn:
+    case aten::save:
+    case aten::manual_seed:
+    case prim::AddStatValue:
+    case prim::TimePoint:
+    case prim::CallFunction:
+    case prim::CallMethod:
+    case prim::BailoutTemplate:
+    case prim::BailOut:
+    case prim::rpc_async: // It represents RPC message sent.
+    case prim::rpc_sync: // It represents RPC message sent.
+    case prim::rpc_remote: // It represents RPC message sent.
+    case aten::wait: // It can represent RPC message received.
 #if !defined(USE_ROCM)
-      case cuda::set_stream:
-      case cuda::_set_device:
-      case cuda::_current_device:
-      case cuda::synchronize:
+    case cuda::set_stream:
+    case cuda::_set_device:
+    case cuda::_current_device:
+    case cuda::synchronize:
 #endif
-      case prim::Enter:
-      case prim::Exit:
-        return true;
-    }
-
-    auto op = maybeOperator();
-    if (!op) {
-      HABANA_ASSERT(
-          kind_.is_prim(),
-          "Only prim ops are allowed to not have a registered operator but",
-          kind_.toDisplayString(),
-          "doesn't have one either. We don't know if this op has side effects.");
-      return false;
-    }
-
-    if (kind_.is_prim() || kind_.is_aten() || kind_.is_cuda()) {
-      // TODO There is nothing in the system that relies on aten:: and prim::
-      // ops using AliasAnalysisKind::FROM_SCHEMA,
-      // AliasAnalysisKind::INTERNAL_SPECIAL_CASE, or
-      // AliasAnalysisKind::CONSERVATIVE but this is the intended behavior for
-      // all current ops and a good error check. We can consider lifting this
-      // constraint later if we have a use case for it.
-
-      HABANA_ASSERT(
-          (op->aliasAnalysisKind() ==
-               c10::AliasAnalysisKind::INTERNAL_SPECIAL_CASE ||
-           op->aliasAnalysisKind() == c10::AliasAnalysisKind::FROM_SCHEMA ||
-           op->aliasAnalysisKind() == c10::AliasAnalysisKind::CONSERVATIVE),
-          "aten:: and prim:: ops should have c10::AliasAnalysisKind::INTERNAL_SPECIAL_CASE, c10::AliasAnalysisKind::FROM_SCHEMA or c10::AliasAnalysisKind::CONSERVATIVE but",
-          kind_.toDisplayString(),
-          " has ",
-          toString(op->aliasAnalysisKind()));
-    }
-
-    switch (op->aliasAnalysisKind()) {
-      case c10::AliasAnalysisKind::PURE_FUNCTION:
-      case c10::AliasAnalysisKind::FROM_SCHEMA:
-      case c10::AliasAnalysisKind::INTERNAL_SPECIAL_CASE:
-        return false;
-      case c10::AliasAnalysisKind::CONSERVATIVE:
-        return true;
-    }
-    HABANA_ASSERT(false, "Unhandled c10::AliasAnalysisKind case");
-    return false; // silence compiler warning
+    case prim::Enter:
+    case prim::Exit:
+      return true;
   }
 
-  // Assign this node a topological position, to facilitate fast isBefore() and
-  // isAfter() queries. Must be called right after a node is inserted into the
-  // node list.
-  //
-  // The basic scheme is: assign every node a position (uint64_t).  The common
-  // case (appending to the end of the graph) is made more efficient by
-  // advancing a fixed interval past the previous node and placing `this` there.
-  // Otherwise, assign `this` a position at the midpoint between its prev() and
-  // next() nodes.
-  //
-  // If we ever run out of space (by, e.g. inserting too much in place), we
-  // reindex by spreading out all the nodes again.
-  void Node::assignTopoPosition() {
-    bool is_first = prev() == owningBlock()->param_node();
-    bool is_last = next() == owningBlock()->return_node();
-
-    const auto prevPos = prev()->topo_position_;
-    const auto nextPos = next()->topo_position_;
-
-    // Append to the end of the graph
-    if (is_last) {
-      if (is_first) {
-        // the node list is empty, assign the first position
-        topo_position_ = kMidPoint;
-        return;
-      }
-
-      if (prevPos >= (kUpperBound - kAppendInterval)) {
-        // we're running off the edge
-        owningBlock()->reIndexTopology();
-        return;
-      }
-
-      topo_position_ = prevPos + kAppendInterval;
-
-      // Prepend to the graph
-    } else if (is_first) {
-      // next() is the first element in the block list
-      if (nextPos <= (kLowerBound + kAppendInterval)) {
-        // we're running off the edge
-        owningBlock()->reIndexTopology();
-        return;
-      }
-      topo_position_ = nextPos - kAppendInterval;
-
-      // insert between two existing nodes
-    } else {
-      int64_t remaining = nextPos - prevPos;
-      HABANA_ASSERT(remaining > 0);
-      if (remaining == 1) {
-        // There was no room
-        owningBlock()->reIndexTopology();
-        return;
-      }
-      int64_t predicted_future_insertions = 0;
-      if (next() == graph_->insertPoint()) {
-        predicted_future_insertions = graph_->predicted_insert_count_++;
-      }
-      topo_position_ = prevPos +
-          std::max(int64_t(1), remaining / (2 + predicted_future_insertions));
-      HABANA_ASSERT(prevPos < topo_position_ && topo_position_ < nextPos);
-    }
-  }
-
-  Node::Node(Graph * graph_, NodeKind kind_)
-      : kind_(kind_),
-        graph_(graph_),
-        owning_block_(nullptr),
-        scope_(graph_->current_scope_),
-        callstack_(c10::nullopt),
-        op_(nullptr),
-        topo_position_(0) {
-    graph_->all_nodes.emplace(this);
-  }
-
-  void Node::eraseOutput(size_t i) {
-    HABANA_ASSERT(i < outputs_.size());
-    HABANA_ASSERT(outputs_[i]->uses().empty());
-    op_ = nullptr;
-    Value* n = outputs_[i];
-    outputs_.erase(outputs_.begin() + i);
-    owningGraph()->freeValue(n);
-    for (const auto j : c10::irange(i, outputs_.size())) {
-      outputs_[j]->offset_--;
-    }
-  }
-
-  Block* Node::addBlock() {
-    op_ = nullptr;
-    blocks_.push_back(new Block(owningGraph(), this));
-    return blocks_.back();
-  }
-
-  void Node::eraseBlock(size_t i) {
-    HABANA_ASSERT(i < blocks_.size());
-    op_ = nullptr;
-    Block* n = blocks_[i];
-    blocks_.erase(blocks_.begin() + i);
-    n->destroy();
-  }
-
-  void Node::destroy() {
-    while (!outputs().empty()) {
-      eraseOutput(outputs().size() - 1);
-    }
-    while (!blocks().empty()) {
-      eraseBlock(blocks().size() - 1);
-    }
-    removeAllInputs();
-    if (inBlockList()) {
-      removeFromList();
-    }
-    graph_->freeNode(this);
-  }
-
-  void Node::cloneFrom(Node * s) {
-    source_range_ = s->source_range_;
-    if (s->scope_ && !s->scope_->isBlank()) {
-      scope_ = s->scope_;
-    }
-    copyAttributes(*s);
-    callstack_ = s->callstack_;
-  }
-
-  void Node::replaceAllUsesWith(Node * n) {
-    HABANA_ASSERT(outputs().size() == n->outputs().size());
-    size_t nOutputs = outputs().size();
-    for (const auto i : c10::irange(nOutputs)) {
-      outputs()[i]->replaceAllUsesWith(n->outputs()[i]);
-    }
-  }
-
-  Node* Node::replaceWithNewSymbol(Symbol new_symbol) {
-    WithInsertPoint insert_guard{this};
-    // bool had_operator = maybeOperator() != nullptr;
-    auto graph = owningGraph();
-    auto replace_node = graph->insertNode(graph->create(new_symbol, 0));
-    for (Value* v : inputs()) {
-      replace_node->addInput(v);
-    }
-    for (Value* v : outputs()) {
-      auto new_out = replace_node->addOutput()->copyMetadata(v);
-      v->replaceAllUsesWith(new_out);
-    }
-    replace_node->copyMetadata(this);
-    replace_node->copyAttributes(*this);
-    // HABANA_ASSERT(
-    //     (replace_node->maybeOperator() != nullptr) == had_operator,
-    //     "invalid symbol replacement:",
-    //     new_symbol,
-    //     kind());
-
-    return replace_node;
-  }
-
-  bool Node::isDominatedBy(const Node* dominator) const {
-    const Node* node = this;
-    while (node) {
-      if (node->owningBlock() == dominator->owningBlock()) {
-        return dominator->isBefore(node);
-      }
-      node = node->owningBlock()->owningNode();
-    }
-    return false;
-  }
-
-  Value* Node::insertInput(size_t i, Value * value) {
-    HABANA_ASSERT(graph_ == value->owningGraph());
-    op_ = nullptr;
-    // First we update the offsets for all existing inputs that will reside
-    // after the one we're inserting. Concretely, these are the inputs at
-    // indices [i, # input). Since we're inserting one input before all of
-    // these inputs, increment their use offsets for this value by 1
-    for (const auto use_itr : c10::irange(i, inputs_.size())) {
-      // See Note [User node does not uniquely identify use]
-      auto use = findUseForInput(use_itr);
-      use->offset += 1;
-    }
-    // Insert the actual input at the specified index
-    inputs_.insert(inputs_.begin() + i, value);
-    // Register the new use of the value we're inserted as an input.
-    value->uses_.emplace_back(this, i);
-    return value;
-  }
-
-  Value* Node::addInput(Value * value) {
-    HABANA_ASSERT(graph_ == value->owningGraph());
-    op_ = nullptr;
-    value->uses_.emplace_back(this, inputs_.size());
-    inputs_.push_back(value);
-    return value;
-  }
-
-  Value* Node::replaceInput(size_t i, Value * newValue) {
-    HABANA_ASSERT(newValue->owningGraph() == graph_);
-    op_ = nullptr;
-    Value* old = dropInput(i);
-    inputs_[i] = newValue;
-    newValue->uses_.emplace_back(this, i);
-    return old;
-  }
-
-  void Node::replaceInputWith(Value * from, Value * to) {
-    HABANA_ASSERT(from->owningGraph() == graph_);
-    HABANA_ASSERT(to->owningGraph() == graph_);
-    op_ = nullptr;
-    size_t i = 0;
-    for (auto input : inputs()) {
-      if (input == from) {
-        replaceInput(i, to);
-      }
-      i++;
-    }
-  }
-
-  Value* Node::addOutput() {
-    outputs_.push_back(new Value(this, outputs_.size()));
-    op_ = nullptr;
-    return outputs_.back();
-  }
-
-  Value* Node::insertOutput(size_t i) {
-    op_ = nullptr;
-    outputs_.insert(outputs_.begin() + i, new Value(this, i));
-    for (size_t itr = i + 1; itr < outputs_.size(); ++itr) {
-      outputs_[itr]->setOffset(outputs_[itr]->offset() + 1);
-    }
-    return outputs_.at(i);
-  }
-
-  bool Node::isBeforeOrAfter(const Node* n, MoveSide moveSide) const {
-    if (this->owningBlock() == n->owningBlock()) {
-      if (moveSide == MoveSide::BEFORE) {
-        return this->topo_position_ < n->topo_position_;
-      }
-
-      if (moveSide == MoveSide::AFTER) {
-        return this->topo_position_ > n->topo_position_;
-      }
-
-      HABANA_ASSERT(this == n);
-      return false;
-    }
-
-    // These nodes don't share a common block. Traverse the blockchains upward
-    // until we find the first common block.
-    auto lhs = this;
-    while (lhs) {
-      HABANA_ASSERT(lhs->owningBlock());
-
-      auto rhs = n;
-      while (rhs) {
-        if (!rhs->owningBlock()) {
-          break;
-        }
-
-        if (lhs->owningBlock() == rhs->owningBlock()) {
-          return lhs->isBeforeOrAfter(rhs, moveSide);
-        }
-        rhs = rhs->owningBlock()->owningNode();
-      }
-
-      lhs = lhs->owningBlock()->owningNode();
-    }
-    // should never reach here, since both nodes are ultimately in the same
-    // graph
-    HABANA_ASSERT(false);
-    // then remove return below, use proper assert instead of return false; for
-    // the need of compilation
-    return false;
-  }
-
-  bool Node::isBefore(const Node* n) const {
-    return isBeforeOrAfter(n, MoveSide::BEFORE);
-  }
-
-  bool Node::isAfter(const Node* n) const {
-    return isBeforeOrAfter(n, MoveSide::AFTER);
-  }
-
-  Node* Node::insertBefore(Node * n) {
-    HABANA_ASSERT(n->inBlockList());
-    insertAfter(n->prev());
-    return this;
-  }
-
-  Node* Node::insertAfter(Node * n) {
-    HABANA_ASSERT(!inBlockList() && n->inBlockList());
-    HABANA_ASSERT(n->owningBlock());
+  auto op = maybeOperator();
+  if (!op) {
     HABANA_ASSERT(
-        n->kind() != prim::Return,
-        "Attempting to insert a Node after the Return node or before the Param node.Tried to insert ",
-        *this,
-        " after ",
-        *n,
-        ".");
-    this->owning_block_ = n->owningBlock();
-    Node* next = n->next();
-    n->next() = this;
-    this->prev() = n;
-    this->next() = next;
-    next->prev() = this;
-    assignTopoPosition();
-    return this;
+        kind_.is_prim(),
+        "Only prim ops are allowed to not have a registered operator but",
+        kind_.toDisplayString(),
+        "doesn't have one either. We don't know if this op has side effects.");
+    return false;
   }
 
-  void Node::moveAfter(Node * n) {
+  if (kind_.is_prim() || kind_.is_aten() || kind_.is_cuda()) {
+    // TODO There is nothing in the system that relies on aten:: and prim::
+    // ops using AliasAnalysisKind::FROM_SCHEMA,
+    // AliasAnalysisKind::INTERNAL_SPECIAL_CASE, or
+    // AliasAnalysisKind::CONSERVATIVE but this is the intended behavior for
+    // all current ops and a good error check. We can consider lifting this
+    // constraint later if we have a use case for it.
+
+    HABANA_ASSERT(
+        (op->aliasAnalysisKind() ==
+             c10::AliasAnalysisKind::INTERNAL_SPECIAL_CASE ||
+         op->aliasAnalysisKind() == c10::AliasAnalysisKind::FROM_SCHEMA ||
+         op->aliasAnalysisKind() == c10::AliasAnalysisKind::CONSERVATIVE),
+        "aten:: and prim:: ops should have c10::AliasAnalysisKind::INTERNAL_SPECIAL_CASE, c10::AliasAnalysisKind::FROM_SCHEMA or c10::AliasAnalysisKind::CONSERVATIVE but",
+        kind_.toDisplayString(),
+        " has ",
+        toString(op->aliasAnalysisKind()));
+  }
+
+  switch (op->aliasAnalysisKind()) {
+    case c10::AliasAnalysisKind::PURE_FUNCTION:
+    case c10::AliasAnalysisKind::FROM_SCHEMA:
+    case c10::AliasAnalysisKind::INTERNAL_SPECIAL_CASE:
+      return false;
+    case c10::AliasAnalysisKind::CONSERVATIVE:
+      return true;
+  }
+  HABANA_ASSERT(false, "Unhandled c10::AliasAnalysisKind case");
+  return false; // silence compiler warning
+}
+
+// Assign this node a topological position, to facilitate fast isBefore() and
+// isAfter() queries. Must be called right after a node is inserted into the
+// node list.
+//
+// The basic scheme is: assign every node a position (uint64_t).  The common
+// case (appending to the end of the graph) is made more efficient by
+// advancing a fixed interval past the previous node and placing `this` there.
+// Otherwise, assign `this` a position at the midpoint between its prev() and
+// next() nodes.
+//
+// If we ever run out of space (by, e.g. inserting too much in place), we
+// reindex by spreading out all the nodes again.
+void Node::assignTopoPosition() {
+  bool is_first = prev() == owningBlock()->param_node();
+  bool is_last = next() == owningBlock()->return_node();
+
+  const auto prevPos = prev()->topo_position_;
+  const auto nextPos = next()->topo_position_;
+
+  // Append to the end of the graph
+  if (is_last) {
+    if (is_first) {
+      // the node list is empty, assign the first position
+      topo_position_ = kMidPoint;
+      return;
+    }
+
+    if (prevPos >= (kUpperBound - kAppendInterval)) {
+      // we're running off the edge
+      owningBlock()->reIndexTopology();
+      return;
+    }
+
+    topo_position_ = prevPos + kAppendInterval;
+
+    // Prepend to the graph
+  } else if (is_first) {
+    // next() is the first element in the block list
+    if (nextPos <= (kLowerBound + kAppendInterval)) {
+      // we're running off the edge
+      owningBlock()->reIndexTopology();
+      return;
+    }
+    topo_position_ = nextPos - kAppendInterval;
+
+    // insert between two existing nodes
+  } else {
+    int64_t remaining = nextPos - prevPos;
+    HABANA_ASSERT(remaining > 0);
+    if (remaining == 1) {
+      // There was no room
+      owningBlock()->reIndexTopology();
+      return;
+    }
+    int64_t predicted_future_insertions = 0;
+    if (next() == graph_->insertPoint()) {
+      predicted_future_insertions = graph_->predicted_insert_count_++;
+    }
+    topo_position_ = prevPos +
+        std::max(int64_t(1), remaining / (2 + predicted_future_insertions));
+    HABANA_ASSERT(prevPos < topo_position_ && topo_position_ < nextPos);
+  }
+}
+
+Node::Node(Graph* graph_, NodeKind kind_)
+    : kind_(kind_),
+      graph_(graph_),
+      owning_block_(nullptr),
+      scope_(graph_->current_scope_),
+      callstack_(c10::nullopt),
+      op_(nullptr),
+      topo_position_(0) {
+  graph_->all_nodes.emplace(this);
+}
+
+void Node::eraseOutput(size_t i) {
+  HABANA_ASSERT(i < outputs_.size());
+  HABANA_ASSERT(outputs_[i]->uses().empty());
+  op_ = nullptr;
+  Value* n = outputs_[i];
+  outputs_.erase(outputs_.begin() + i);
+  owningGraph()->freeValue(n);
+  for (const auto j : c10::irange(i, outputs_.size())) {
+    outputs_[j]->offset_--;
+  }
+}
+
+Block* Node::addBlock() {
+  op_ = nullptr;
+  blocks_.push_back(new Block(owningGraph(), this));
+  return blocks_.back();
+}
+
+void Node::eraseBlock(size_t i) {
+  HABANA_ASSERT(i < blocks_.size());
+  op_ = nullptr;
+  Block* n = blocks_[i];
+  blocks_.erase(blocks_.begin() + i);
+  n->destroy();
+}
+
+void Node::destroy() {
+  while (!outputs().empty()) {
+    eraseOutput(outputs().size() - 1);
+  }
+  while (!blocks().empty()) {
+    eraseBlock(blocks().size() - 1);
+  }
+  removeAllInputs();
+  if (inBlockList()) {
     removeFromList();
-    insertAfter(n);
+  }
+  graph_->freeNode(this);
+}
+
+void Node::cloneFrom(Node* s) {
+  source_range_ = s->source_range_;
+  if (s->scope_ && !s->scope_->isBlank()) {
+    scope_ = s->scope_;
+  }
+  copyAttributes(*s);
+  callstack_ = s->callstack_;
+}
+
+void Node::replaceAllUsesWith(Node* n) {
+  HABANA_ASSERT(outputs().size() == n->outputs().size());
+  size_t nOutputs = outputs().size();
+  for (const auto i : c10::irange(nOutputs)) {
+    outputs()[i]->replaceAllUsesWith(n->outputs()[i]);
+  }
+}
+
+Node* Node::replaceWithNewSymbol(Symbol new_symbol) {
+  WithInsertPoint insert_guard{this};
+  // bool had_operator = maybeOperator() != nullptr;
+  auto graph = owningGraph();
+  auto replace_node = graph->insertNode(graph->create(new_symbol, 0));
+  for (Value* v : inputs()) {
+    replace_node->addInput(v);
+  }
+  for (Value* v : outputs()) {
+    auto new_out = replace_node->addOutput()->copyMetadata(v);
+    v->replaceAllUsesWith(new_out);
+  }
+  replace_node->copyMetadata(this);
+  replace_node->copyAttributes(*this);
+  // HABANA_ASSERT(
+  //     (replace_node->maybeOperator() != nullptr) == had_operator,
+  //     "invalid symbol replacement:",
+  //     new_symbol,
+  //     kind());
+
+  return replace_node;
+}
+
+bool Node::isDominatedBy(const Node* dominator) const {
+  const Node* node = this;
+  while (node) {
+    if (node->owningBlock() == dominator->owningBlock()) {
+      return dominator->isBefore(node);
+    }
+    node = node->owningBlock()->owningNode();
+  }
+  return false;
+}
+
+Value* Node::insertInput(size_t i, Value* value) {
+  HABANA_ASSERT(graph_ == value->owningGraph());
+  op_ = nullptr;
+  // First we update the offsets for all existing inputs that will reside
+  // after the one we're inserting. Concretely, these are the inputs at
+  // indices [i, # input). Since we're inserting one input before all of
+  // these inputs, increment their use offsets for this value by 1
+  for (const auto use_itr : c10::irange(i, inputs_.size())) {
+    // See Note [User node does not uniquely identify use]
+    auto use = findUseForInput(use_itr);
+    use->offset += 1;
+  }
+  // Insert the actual input at the specified index
+  inputs_.insert(inputs_.begin() + i, value);
+  // Register the new use of the value we're inserted as an input.
+  value->uses_.emplace_back(this, i);
+  return value;
+}
+
+Value* Node::addInput(Value* value) {
+  HABANA_ASSERT(graph_ == value->owningGraph());
+  op_ = nullptr;
+  value->uses_.emplace_back(this, inputs_.size());
+  inputs_.push_back(value);
+  return value;
+}
+
+Value* Node::replaceInput(size_t i, Value* newValue) {
+  HABANA_ASSERT(newValue->owningGraph() == graph_);
+  op_ = nullptr;
+  Value* old = dropInput(i);
+  inputs_[i] = newValue;
+  newValue->uses_.emplace_back(this, i);
+  return old;
+}
+
+void Node::replaceInputWith(Value* from, Value* to) {
+  HABANA_ASSERT(from->owningGraph() == graph_);
+  HABANA_ASSERT(to->owningGraph() == graph_);
+  op_ = nullptr;
+  size_t i = 0;
+  for (auto input : inputs()) {
+    if (input == from) {
+      replaceInput(i, to);
+    }
+    i++;
+  }
+}
+
+Value* Node::addOutput() {
+  outputs_.push_back(new Value(this, outputs_.size()));
+  op_ = nullptr;
+  return outputs_.back();
+}
+
+Value* Node::insertOutput(size_t i) {
+  op_ = nullptr;
+  outputs_.insert(outputs_.begin() + i, new Value(this, i));
+  for (size_t itr = i + 1; itr < outputs_.size(); ++itr) {
+    outputs_[itr]->setOffset(outputs_[itr]->offset() + 1);
+  }
+  return outputs_.at(i);
+}
+
+bool Node::isBeforeOrAfter(const Node* n, MoveSide moveSide) const {
+  if (this->owningBlock() == n->owningBlock()) {
+    if (moveSide == MoveSide::BEFORE) {
+      return this->topo_position_ < n->topo_position_;
+    }
+
+    if (moveSide == MoveSide::AFTER) {
+      return this->topo_position_ > n->topo_position_;
+    }
+
+    HABANA_ASSERT(this == n);
+    return false;
   }
 
-  void Node::moveBefore(Node * n) {
-    removeFromList();
-    insertBefore(n);
-  }
+  // These nodes don't share a common block. Traverse the blockchains upward
+  // until we find the first common block.
+  auto lhs = this;
+  while (lhs) {
+    HABANA_ASSERT(lhs->owningBlock());
 
-  void Node::removeInput(size_t i) {
-    op_ = nullptr;
+    auto rhs = n;
+    while (rhs) {
+      if (!rhs->owningBlock()) {
+        break;
+      }
+
+      if (lhs->owningBlock() == rhs->owningBlock()) {
+        return lhs->isBeforeOrAfter(rhs, moveSide);
+      }
+      rhs = rhs->owningBlock()->owningNode();
+    }
+
+    lhs = lhs->owningBlock()->owningNode();
+  }
+  // should never reach here, since both nodes are ultimately in the same
+  // graph
+  HABANA_ASSERT(false);
+}
+
+bool Node::isBefore(const Node* n) const {
+  return isBeforeOrAfter(n, MoveSide::BEFORE);
+}
+
+bool Node::isAfter(const Node* n) const {
+  return isBeforeOrAfter(n, MoveSide::AFTER);
+}
+
+Node* Node::insertBefore(Node* n) {
+  HABANA_ASSERT(n->inBlockList());
+  insertAfter(n->prev());
+  return this;
+}
+
+Node* Node::insertAfter(Node* n) {
+  HABANA_ASSERT(!inBlockList() && n->inBlockList());
+  HABANA_ASSERT(n->owningBlock());
+  HABANA_ASSERT(
+      n->kind() != prim::Return,
+      "Attempting to insert a Node after the Return node or before the Param node.Tried to insert ",
+      *this,
+      " after ",
+      *n,
+      ".");
+  this->owning_block_ = n->owningBlock();
+  Node* next = n->next();
+  n->next() = this;
+  this->prev() = n;
+  this->next() = next;
+  next->prev() = this;
+  assignTopoPosition();
+  return this;
+}
+
+void Node::moveAfter(Node* n) {
+  removeFromList();
+  insertAfter(n);
+}
+
+void Node::moveBefore(Node* n) {
+  removeFromList();
+  insertBefore(n);
+}
+
+void Node::removeInput(size_t i) {
+  op_ = nullptr;
+  dropInput(i);
+  // everything after this input shifts left,
+  // so we need to update their use offsets to match
+  for (size_t j = i + 1; j < inputs_.size(); j++) {
+    auto it = findUseForInput(j);
+    it->offset--;
+  }
+  inputs_.erase(inputs_.begin() + i);
+}
+
+void Node::removeAllInputs() {
+  op_ = nullptr;
+  for (const auto i : c10::irange(inputs().size())) {
     dropInput(i);
-    // everything after this input shifts left,
-    // so we need to update their use offsets to match
-    for (size_t j = i + 1; j < inputs_.size(); j++) {
-      auto it = findUseForInput(j);
-      it->offset--;
-    }
-    inputs_.erase(inputs_.begin() + i);
+  }
+  inputs_.clear();
+}
+
+void Node::removeAllOutputs() {
+  op_ = nullptr;
+  size_t init_osize = outputs_.size();
+  for (auto i : c10::irange(init_osize)) {
+    eraseOutput(init_osize - i - 1);
+  }
+}
+void Node::permuteInputs(const std::vector<size_t>& new_order) {
+  op_ = nullptr;
+  HABANA_ASSERT(new_order.size() == inputs_.size());
+  std::vector<Value*> new_inputs;
+  new_inputs.reserve(new_order.size());
+  for (const auto i : c10::irange(new_order.size())) {
+    HABANA_ASSERT(inputs_.at(new_order[i]) != nullptr, "Repeated index");
+    new_inputs.push_back(inputs_.at(new_order[i]));
+    auto it = findUseForInput(new_order[i]);
+    it->offset = i;
+    inputs_.at(new_order[i]) = nullptr;
+  }
+  inputs_ = std::move(new_inputs);
+}
+void Node::permuteOutputs(const std::vector<size_t>& new_order) {
+  op_ = nullptr;
+  HABANA_ASSERT(new_order.size() == outputs_.size());
+  std::vector<Value*> new_outputs;
+  new_outputs.reserve(new_order.size());
+  for (const auto i : c10::irange(new_order.size())) {
+    HABANA_ASSERT(outputs_.at(new_order[i]) != nullptr, "Repeated index");
+    new_outputs.push_back(outputs_.at(new_order[i]));
+    outputs_.at(new_order[i])->setOffset(i);
+    outputs_.at(new_order[i]) = nullptr;
+  }
+  outputs_ = std::move(new_outputs);
+}
+
+use_list::iterator Node::findUseForInput(size_t i) {
+  auto& input_uses = inputs_[i]->uses_;
+  // O(N) on the use list, but unless we get nodes with +100 uses
+  // vector traversal still is probably faster than linked list
+  auto use_it = std::find(input_uses.begin(), input_uses.end(), Use(this, i));
+  HABANA_ASSERT(use_it != input_uses.end());
+  return use_it;
+}
+Value* Node::dropInput(size_t i) {
+  HABANA_ASSERT(i < inputs_.size());
+  auto input_node = inputs_[i];
+  auto use_it = findUseForInput(i);
+  input_node->uses_.erase(use_it);
+  inputs_[i] = nullptr;
+  return input_node;
+}
+void Node::removeFromList() {
+  HABANA_ASSERT(inBlockList());
+  this->owning_block_ = nullptr;
+  Node* next = this->next();
+  Node* prev = this->prev();
+  prev->next() = next;
+  next->prev() = prev;
+  this->next() = nullptr;
+  this->prev() = nullptr;
+}
+
+Block* Node::findCommonAncestorBlockWith(Node* n) {
+  if (n->owningBlock() == owningBlock()) {
+    return owningBlock();
   }
 
-  void Node::removeAllInputs() {
-    op_ = nullptr;
-    for (const auto i : c10::irange(inputs().size())) {
-      dropInput(i);
-    }
-    inputs_.clear();
+  Node* n1 = this;
+  Node* n2 = n;
+
+  size_t d_1 = n1->blocksFromGraphBlock();
+  size_t d_2 = n2->blocksFromGraphBlock();
+
+  for (; d_1 > d_2; --d_1) {
+    n1 = n1->owningBlock()->owningNode();
+    // n2 contains n1
   }
 
-  void Node::removeAllOutputs() {
-    op_ = nullptr;
-    size_t init_osize = outputs_.size();
-    for (auto i : c10::irange(init_osize)) {
-      eraseOutput(init_osize - i - 1);
-    }
-  }
-  void Node::permuteInputs(const std::vector<size_t>& new_order) {
-    op_ = nullptr;
-    HABANA_ASSERT(new_order.size() == inputs_.size());
-    std::vector<Value*> new_inputs;
-    new_inputs.reserve(new_order.size());
-    for (const auto i : c10::irange(new_order.size())) {
-      HABANA_ASSERT(inputs_.at(new_order[i]) != nullptr, "Repeated index");
-      new_inputs.push_back(inputs_.at(new_order[i]));
-      auto it = findUseForInput(new_order[i]);
-      it->offset = i;
-      inputs_.at(new_order[i]) = nullptr;
-    }
-    inputs_ = std::move(new_inputs);
-  }
-  void Node::permuteOutputs(const std::vector<size_t>& new_order) {
-    op_ = nullptr;
-    HABANA_ASSERT(new_order.size() == outputs_.size());
-    std::vector<Value*> new_outputs;
-    new_outputs.reserve(new_order.size());
-    for (const auto i : c10::irange(new_order.size())) {
-      HABANA_ASSERT(outputs_.at(new_order[i]) != nullptr, "Repeated index");
-      new_outputs.push_back(outputs_.at(new_order[i]));
-      outputs_.at(new_order[i])->setOffset(i);
-      outputs_.at(new_order[i]) = nullptr;
-    }
-    outputs_ = std::move(new_outputs);
+  for (; d_2 > d_1; --d_2) {
+    n2 = n2->owningBlock()->owningNode();
   }
 
-  use_list::iterator Node::findUseForInput(size_t i) {
-    auto& input_uses = inputs_[i]->uses_;
-    // O(N) on the use list, but unless we get nodes with +100 uses
-    // vector traversal still is probably faster than linked list
-    auto use_it = std::find(input_uses.begin(), input_uses.end(), Use(this, i));
-    HABANA_ASSERT(use_it != input_uses.end());
-    return use_it;
-  }
-  Value* Node::dropInput(size_t i) {
-    HABANA_ASSERT(i < inputs_.size());
-    auto input_node = inputs_[i];
-    auto use_it = findUseForInput(i);
-    input_node->uses_.erase(use_it);
-    inputs_[i] = nullptr;
-    return input_node;
-  }
-  void Node::removeFromList() {
-    HABANA_ASSERT(inBlockList());
-    this->owning_block_ = nullptr;
-    Node* next = this->next();
-    Node* prev = this->prev();
-    prev->next() = next;
-    next->prev() = prev;
-    this->next() = nullptr;
-    this->prev() = nullptr;
-  }
-
-  Block* Node::findCommonAncestorBlockWith(Node * n) {
-    if (n->owningBlock() == owningBlock()) {
-      return owningBlock();
+  // Now they are the same numer of blocks from the graph block,
+  // recurse upwards, checking if they are on the same block
+  while (true) {
+    if (n1->owningBlock() == n2->owningBlock()) {
+      return n1->owningBlock();
     }
 
-    Node* n1 = this;
-    Node* n2 = n;
+    n1 = n1->owningBlock()->owningNode();
+    n2 = n2->owningBlock()->owningNode();
+    HABANA_ASSERT(n1 != nullptr);
+    HABANA_ASSERT(n2 != nullptr);
+  }
+}
 
-    size_t d_1 = n1->blocksFromGraphBlock();
-    size_t d_2 = n2->blocksFromGraphBlock();
+size_t Node::blocksFromGraphBlock() {
+  Node* n = this;
+  size_t dist = 0;
+  while (n->owningBlock()->owningNode()) {
+    n = n->owningBlock()->owningNode();
+    ++dist;
+  }
+  return dist;
+}
 
-    for (; d_1 > d_2; --d_1) {
-      n1 = n1->owningBlock()->owningNode();
-      // n2 contains n1
-    }
+inline const SourceRange& fakeRange() {
+  static SourceRange range(std::make_shared<Source>(std::string("")), 0, 1);
+  return range;
+}
 
-    for (; d_2 > d_1; --d_2) {
-      n2 = n2->owningBlock()->owningNode();
-    }
+Value* Graph::insert(
+    Symbol opname,
+    at::ArrayRef<NamedValue> args,
+    at::ArrayRef<NamedValue> kwargs,
+    const c10::optional<SourceRange>& range) {
+  return emitBuiltinCall(
+      range.value_or(fakeRange()), *this, opname, args, kwargs);
+}
 
-    // Now they are the same numer of blocks from the graph block,
-    // recurse upwards, checking if they are on the same block
-    while (true) {
-      if (n1->owningBlock() == n2->owningBlock()) {
-        return n1->owningBlock();
-      }
+Node* Graph::create(NodeKind kind, size_t num_outputs) {
+  // NB: Node constructor adds node to all_nodes
+  auto n = new Node(this, kind);
+  for (const auto i : c10::irange(num_outputs)) {
+    (void)i;
+    n->addOutput();
+  }
+  return n;
+}
 
-      n1 = n1->owningBlock()->owningNode();
-      n2 = n2->owningBlock()->owningNode();
-      HABANA_ASSERT(n1 != nullptr);
-      HABANA_ASSERT(n2 != nullptr);
+Node* Graph::create(
+    NodeKind kind,
+    ArrayRef<Value*> inputs,
+    size_t num_outputs) {
+  auto n = create(kind, num_outputs);
+  for (auto i : inputs) {
+    n->addInput(i);
+  }
+  return n;
+}
+
+Node* Graph::createAutogradZero() {
+  return create(prim::AutogradZero);
+}
+
+Node* Graph::createNone() {
+  Node* n = create(prim::Constant);
+  n->output()->setType(NoneType::get());
+  return n;
+}
+
+Node* Graph::createUninitialized(TypePtr type) {
+  Node* n = create(prim::Uninitialized);
+  n->output()->setType(std::move(type));
+  return n;
+}
+
+Node* Graph::createWithSubgraph(Symbol kind) {
+  auto n = create(kind, 0);
+  n->g_(attr::Subgraph, std::make_shared<Graph>(current_scope()));
+  return n;
+}
+
+Node* Graph::createTuple(at::ArrayRef<Value*> values, TupleTypePtr tuple_type) {
+  if (!tuple_type) {
+    auto types = fmap(values, [](Value* v) { return v->type(); });
+    tuple_type = TupleType::create(std::move(types));
+  }
+  auto n = create(prim::TupleConstruct, values);
+
+  n->output()->setType(tuple_type);
+  return n;
+}
+
+Node* Graph::createTupleUnpack(Value* v) {
+  TupleTypePtr tt = v->type()->expect<TupleType>();
+  auto n = create(prim::TupleUnpack, {v}, 0);
+  for (auto& element : tt->elements()) {
+    n->addOutput()->setType(element);
+  }
+  return n;
+}
+
+Node* Graph::createTupleIndex(
+    Value* tup,
+    Value* idx,
+    const TypePtr& output_type) {
+  auto n = create(prim::TupleIndex, {tup, idx});
+  n->output()->setType(output_type);
+  return n;
+}
+
+Node* Graph::createTupleSlice(
+    Value* tup,
+    int64_t beg,
+    int64_t step_size,
+    int64_t num_values) {
+  std::vector<Value*> new_vals;
+  TupleTypePtr tt = tup->type()->expect<TupleType>();
+  new_vals.reserve(num_values);
+
+  int64_t i = beg;
+  for (const auto j : c10::irange(num_values)) {
+    (void)j; // Suppress unused variable warning
+    auto idx = insertConstant(IValue(static_cast<int64_t>(i)));
+    auto tupleIndex = insertNode(createTupleIndex(tup, idx, tt->elements()[i]));
+
+    new_vals.push_back(tupleIndex->output());
+    i += step_size;
+  }
+
+  auto n = createTuple(new_vals);
+  return n;
+}
+
+Node* Graph::createEnumName(Value* e) {
+  e->type()->expect<EnumType>();
+  assert(e->type()->cast<EnumType>());
+  auto n = create(prim::EnumName, {e});
+  n->output()->setType(StringType::get());
+  return n;
+}
+
+Node* Graph::createEnumValue(Value* e) {
+  auto enum_type = e->type()->expect<EnumType>();
+  auto n = create(prim::EnumValue, {e});
+  n->output()->setType(enum_type->getValueType());
+  return n;
+}
+
+Node* Graph::createList(
+    const TypePtr& contained_type,
+    at::ArrayRef<Value*> values) {
+  auto n = create(prim::ListConstruct, values);
+  for (const auto& v : values) {
+    TORCH_CHECK(
+        matchTypes(v->type(), contained_type),
+        "Expected a list element that subtypes '",
+        contained_type->repr_str(),
+        "' but got an element of type '",
+        v->type()->repr_str(),
+        "'");
+  }
+  n->output()->setType(ListType::create(contained_type));
+  return n;
+}
+
+Node* Graph::createListUnpack(Value* v, size_t size) {
+  ListTypePtr list_type = v->type()->expect<ListType>();
+  TypePtr elem_type = list_type->getElementType();
+  auto n = create(prim::ListUnpack, {v}, 0);
+  for (const auto i : c10::irange(size)) {
+    (void)i; // Suppress unused variable warning
+    n->addOutput()->setType(elem_type);
+  }
+  return n;
+}
+
+Value* Graph::packValues(
+    at::ArrayRef<Value*> values,
+    c10::OptNameList field_names) {
+  if (values.size() == 1) {
+    return values[0];
+  }
+  std::shared_ptr<FunctionSchema> schema;
+  TupleTypePtr named_tuple = nullptr;
+  if (field_names) {
+    auto types = fmap(values, [](Value* v) { return v->type(); });
+    named_tuple =
+        TupleType::createNamed(c10::nullopt, field_names.value(), types);
+  }
+  return insertNode(createTuple(values, named_tuple))->output();
+}
+Node* Graph::createDict(
+    const TypePtr& key_type,
+    const TypePtr& value_type,
+    at::ArrayRef<Value*> keys,
+    at::ArrayRef<Value*> values) {
+  HABANA_ASSERT(keys.size() == values.size());
+  auto n = create(prim::DictConstruct, 1);
+  for (const auto i : c10::irange(keys.size())) {
+    HABANA_ASSERT(matchTypes(keys[i]->type(), key_type));
+    HABANA_ASSERT(matchTypes(values[i]->type(), value_type));
+
+    n->addInput(keys[i]);
+    n->addInput(values[i]);
+  }
+  n->output()->setType(DictType::create(key_type, value_type));
+  return n;
+}
+
+Node* Graph::createNumToTensor(Value* value) {
+  Node* result = create(prim::NumToTensor, {value});
+  result->output()->setType(TensorType::fromNumberType(*value->type()));
+  return result;
+}
+
+Node* Graph::createObject(const ClassTypePtr& type) {
+  auto result = create(prim::CreateObject);
+  result->output()->setType(type);
+  return result;
+}
+
+Node* Graph::createSetAttr(
+    Value* obj,
+    const std::string& field,
+    Value* newValue) {
+  auto n = create(prim::SetAttr, {obj, newValue}, /*num_outputs=*/0);
+  n->s_(attr::name, field);
+  return n;
+}
+
+Node* Graph::createGetAttr(Value* obj, const std::string& field) {
+  const auto classType = obj->type()->expect<ClassType>();
+
+  auto n = create(prim::GetAttr, {obj}, /*num_outputs=*/1);
+  n->s_(attr::name, field);
+
+  const auto outputType = classType->getAttribute(field);
+  n->output()->setType(outputType);
+  n->output()->setDebugName(normalizeAttrName(field));
+  return n;
+}
+
+Node* Graph::createStore(const std::string& name, Value* v) {
+  auto n = create(prim::Store, {v}, /*num_outputs*/ 0);
+  n->s_(attr::name, name);
+  return n;
+}
+
+Node* Graph::createLoad(const std::string& name, const TypeWrapper& type) {
+  auto n = create(prim::Load, {}, /*num_outputs*/ 1);
+  n->s_(attr::name, name);
+  n->output()->setType(type);
+  return n;
+}
+
+Node* Graph::createIsInstance(Value* v, at::ArrayRef<TypeWrapper> types) {
+  auto n = create(prim::isinstance, {v}, /*num_outputs*/ 1);
+  n->tys_(attr::types, types.vec());
+  n->output()->setType(BoolType::get());
+  return n;
+}
+Value* Graph::insertUncheckedCast(Value* v, const TypeWrapper& type) {
+  Node* n = insertNode(create(prim::unchecked_cast, {v}));
+  n->output()->setType(type);
+  return n->output();
+}
+
+Value* Graph::insertToList(Value* v, TypePtr type) {
+  int dim = 0;
+  TypePtr ptr = type;
+
+  // Unwrap the type to determine the number of dimensions.
+  while (auto list_type = ptr->cast<ListType>()) {
+    ptr = list_type->getElementType();
+    ++dim;
+  }
+
+  // Encode the base element type as an integer.
+  int elem_ty = 0;
+  if (ptr == IntType::get()) {
+    elem_ty = 0;
+  } else if (ptr == FloatType::get()) {
+    elem_ty = 1;
+  } else if (ptr == BoolType::get()) {
+    elem_ty = 2;
+  } else if (ptr == ComplexType::get()) {
+    elem_ty = 3;
+  } else {
+    TORCH_CHECK(
+        false,
+        ptr->repr_str(),
+        " is not one of the supported element types for tolist: int, float, complex, bool");
+  }
+
+  // Pass in the number of dimensions and base element type as arguments
+  // to the op.
+  Value* dim_val = insertConstant(IValue(dim));
+  Value* elem_ty_val = insertConstant(IValue(elem_ty));
+  Node* n = insertNode(create(prim::tolist, {v, dim_val, elem_ty_val}));
+  n->output()->setType(std::move(type));
+  return n->output();
+}
+
+Value* Graph::insertFunctionCall(
+    torch::jit::Function* callee,
+    const MatchedSchema& matched) {
+  std::string func_name = callee->name();
+  Value* fn_constant = insertNode(create(prim::Constant))
+                           ->s_(attr::name, func_name)
+                           ->output()
+                           ->setType(FunctionType::create(callee));
+  std::vector<Value*> inputs = {fn_constant};
+  inputs.insert(inputs.end(), matched.inputs.begin(), matched.inputs.end());
+  Value* result = insertNode(create(prim::CallFunction, inputs))
+                      ->output()
+                      ->setType(matched.return_types.at(0));
+  return result;
+}
+
+Value* Graph::insertMethodCall(
+    std::string method_name,
+    const MatchedSchema& matched) {
+  Value* result = insertNode(create(prim::CallMethod, matched.inputs))
+                      ->s_(attr::name, std::move(method_name))
+                      ->output()
+                      ->setType(matched.return_types.at(0));
+  return result;
+}
+
+Node* Graph::createClone(
+    Node* n,
+    const std::function<Value*(Value*)>& value_map,
+    bool copy_blocks) {
+  // n can be from a different graph
+  Node* r = n->allocNewInstance(this);
+  for (auto o : n->outputs()) {
+    r->addOutput()->copyMetadata(o);
+  }
+  r->cloneFrom(n);
+  for (auto i : n->inputs()) {
+    r->addInput(value_map(i));
+  }
+  if (copy_blocks) {
+    for (auto b : n->blocks()) {
+      r->addBlock()->cloneFrom(b, value_map);
     }
   }
+  return r;
+}
 
-  size_t Node::blocksFromGraphBlock() {
-    Node* n = this;
-    size_t dist = 0;
-    while (n->owningBlock()->owningNode()) {
-      n = n->owningBlock()->owningNode();
-      ++dist;
-    }
-    return dist;
+Value* Graph::insertConstant(
+    const IValue& val,
+    c10::optional<SourceRange> loc,
+    c10::optional<ScopePtr> scope) {
+  // GAUDI_JIT_TRACE;
+  return jit::insertConstant(*this, val, std::move(loc), std::move(scope));
+}
+
+std::string Graph::toString(bool print_source_info) const {
+  // GAUDI_JIT_TRACE;
+  std::ostringstream oss;
+  print(oss, print_source_info);
+  return oss.str();
+}
+
+Graph::~Graph() {
+  for (const Node* n : all_nodes) {
+    delete n;
   }
-
-  inline const SourceRange& fakeRange() {
-    static SourceRange range(std::make_shared<Source>(std::string("")), 0, 1);
-    return range;
+  for (const Value* v : all_values) {
+    delete v;
   }
-
-  Value* Graph::insert(
-      Symbol opname,
-      at::ArrayRef<NamedValue> args,
-      at::ArrayRef<NamedValue> kwargs,
-      const c10::optional<SourceRange>& range) {
-    return emitBuiltinCall(
-        range.value_or(fakeRange()), *this, opname, args, kwargs);
+  for (const Block* b : all_blocks) {
+    delete b;
   }
+}
 
-  Node* Graph::create(NodeKind kind, size_t num_outputs) {
-    // NB: Node constructor adds node to all_nodes
-    auto n = new Node(this, kind);
-    for (const auto i : c10::irange(num_outputs)) {
-      (void)i;
-      n->addOutput();
-    }
-    return n;
+void Graph::freeNode(Node* n) {
+  auto it = all_nodes.find(n);
+  HABANA_ASSERT(it != all_nodes.end());
+  delete *it;
+  all_nodes.erase(it);
+}
+void Graph::freeValue(Value* v) {
+  v->setDebugName("");
+  auto it = all_values.find(v);
+  HABANA_ASSERT(it != all_values.end());
+  delete *it;
+  all_values.erase(it);
+}
+void Graph::freeBlock(Block* b) {
+  auto it = all_blocks.find(b);
+  HABANA_ASSERT(it != all_blocks.end());
+  delete *it;
+  all_blocks.erase(it);
+}
+
+at::ArrayRef<Value*> createTupleUnpack(Value* v) {
+  // small peephole optimization to ensure IntArrayRef attributes can still
+  // turn into constants e.g. in x.expand([3, 4])
+  if (v->node()->kind() == prim::TupleConstruct) {
+    return v->node()->inputs();
   }
+  auto& g = *v->owningGraph();
+  return g.insertNode(g.createTupleUnpack(v))->outputs();
+}
 
-  Node* Graph::create(
-      NodeKind kind, ArrayRef<Value*> inputs, size_t num_outputs) {
-    auto n = create(kind, num_outputs);
-    for (auto i : inputs) {
-      n->addInput(i);
-    }
-    return n;
+void inlineCallStackOfNode(
+    Node* n,
+    std::unordered_map<InlinedCallStack*, InlinedCallStackPtr>& new_cs_entries,
+    torch::jit::Function* callee,
+    Node* to_replace,
+    c10::optional<ModuleInstanceInfo> m_info);
+
+static void inlineCallStackOfBlock(
+    Block* b,
+    std::unordered_map<InlinedCallStack*, InlinedCallStackPtr>& new_cs_entries,
+    torch::jit::Function* callee,
+    Node* to_replace,
+    c10::optional<ModuleInstanceInfo> m_info) {
+  for (auto n : b->nodes()) {
+    inlineCallStackOfNode(n, new_cs_entries, callee, to_replace, m_info);
   }
+}
 
-  Node* Graph::createAutogradZero() {
-    return create(prim::AutogradZero);
-  }
+void inlineCallStackOfNode(
+    Node* new_node,
+    std::unordered_map<InlinedCallStack*, InlinedCallStackPtr>& new_cs_entries,
+    torch::jit::Function* callee,
+    Node* to_replace,
+    c10::optional<ModuleInstanceInfo> m_info) {
+  auto new_node_cs = new_node->callstack();
 
-  Node* Graph::createNone() {
-    Node* n = create(prim::Constant);
-    n->output()->setType(NoneType::get());
-    return n;
-  }
+  InlinedCallStack* raw_callstack_ptr =
+      new_node_cs ? new_node_cs->get() : nullptr;
 
-  Node* Graph::createUninitialized(TypePtr type) {
-    Node* n = create(prim::Uninitialized);
-    n->output()->setType(std::move(type));
-    return n;
-  }
-
-  Node* Graph::createWithSubgraph(Symbol kind) {
-    auto n = create(kind, 0);
-    n->g_(attr::Subgraph, std::make_shared<Graph>(current_scope()));
-    return n;
-  }
-
-  Node* Graph::createTuple(
-      at::ArrayRef<Value*> values, TupleTypePtr tuple_type) {
-    if (!tuple_type) {
-      auto types = fmap(values, [](Value* v) { return v->type(); });
-      tuple_type = TupleType::create(std::move(types));
-    }
-    auto n = create(prim::TupleConstruct, values);
-
-    n->output()->setType(tuple_type);
-    return n;
-  }
-
-  Node* Graph::createTupleUnpack(Value * v) {
-    TupleTypePtr tt = v->type()->expect<TupleType>();
-    auto n = create(prim::TupleUnpack, {v}, 0);
-    for (auto& element : tt->elements()) {
-      n->addOutput()->setType(element);
-    }
-    return n;
-  }
-
-  Node* Graph::createTupleIndex(
-      Value * tup, Value * idx, const TypePtr& output_type) {
-    auto n = create(prim::TupleIndex, {tup, idx});
-    n->output()->setType(output_type);
-    return n;
-  }
-
-  Node* Graph::createTupleSlice(
-      Value * tup, int64_t beg, int64_t step_size, int64_t num_values) {
-    std::vector<Value*> new_vals;
-    TupleTypePtr tt = tup->type()->expect<TupleType>();
-    new_vals.reserve(num_values);
-
-    int64_t i = beg;
-    for (const auto j : c10::irange(num_values)) {
-      (void)j; // Suppress unused variable warning
-      auto idx = insertConstant(IValue(static_cast<int64_t>(i)));
-      auto tupleIndex =
-          insertNode(createTupleIndex(tup, idx, tt->elements()[i]));
-
-      new_vals.push_back(tupleIndex->output());
-      i += step_size;
-    }
-
-    auto n = createTuple(new_vals);
-    return n;
-  }
-
-  Node* Graph::createEnumName(Value * e) {
-    e->type()->expect<EnumType>();
-    assert(e->type()->cast<EnumType>());
-    auto n = create(prim::EnumName, {e});
-    n->output()->setType(StringType::get());
-    return n;
-  }
-
-  Node* Graph::createEnumValue(Value * e) {
-    auto enum_type = e->type()->expect<EnumType>();
-    auto n = create(prim::EnumValue, {e});
-    n->output()->setType(enum_type->getValueType());
-    return n;
-  }
-
-  Node* Graph::createList(
-      const TypePtr& contained_type, at::ArrayRef<Value*> values) {
-    auto n = create(prim::ListConstruct, values);
-    for (const auto& v : values) {
-      TORCH_CHECK(
-          matchTypes(v->type(), contained_type),
-          "Expected a list element that subtypes '",
-          contained_type->repr_str(),
-          "' but got an element of type '",
-          v->type()->repr_str(),
-          "'");
-    }
-    n->output()->setType(ListType::create(contained_type));
-    return n;
-  }
-
-  Node* Graph::createListUnpack(Value * v, size_t size) {
-    ListTypePtr list_type = v->type()->expect<ListType>();
-    TypePtr elem_type = list_type->getElementType();
-    auto n = create(prim::ListUnpack, {v}, 0);
-    for (const auto i : c10::irange(size)) {
-      (void)i; // Suppress unused variable warning
-      n->addOutput()->setType(elem_type);
-    }
-    return n;
-  }
-
-  Value* Graph::packValues(
-      at::ArrayRef<Value*> values, c10::OptNameList field_names) {
-    if (values.size() == 1) {
-      return values[0];
-    }
-    std::shared_ptr<FunctionSchema> schema;
-    TupleTypePtr named_tuple = nullptr;
-    if (field_names) {
-      auto types = fmap(values, [](Value* v) { return v->type(); });
-      named_tuple =
-          TupleType::createNamed(c10::nullopt, field_names.value(), types);
-    }
-    return insertNode(createTuple(values, named_tuple))->output();
-  }
-  Node* Graph::createDict(
-      const TypePtr& key_type,
-      const TypePtr& value_type,
-      at::ArrayRef<Value*> keys,
-      at::ArrayRef<Value*> values) {
-    HABANA_ASSERT(keys.size() == values.size());
-    auto n = create(prim::DictConstruct, 1);
-    for (const auto i : c10::irange(keys.size())) {
-      HABANA_ASSERT(matchTypes(keys[i]->type(), key_type));
-      HABANA_ASSERT(matchTypes(values[i]->type(), value_type));
-
-      n->addInput(keys[i]);
-      n->addInput(values[i]);
-    }
-    n->output()->setType(DictType::create(key_type, value_type));
-    return n;
-  }
-
-  Node* Graph::createNumToTensor(Value * value) {
-    Node* result = create(prim::NumToTensor, {value});
-    result->output()->setType(TensorType::fromNumberType(*value->type()));
-    return result;
-  }
-
-  Node* Graph::createObject(const ClassTypePtr& type) {
-    auto result = create(prim::CreateObject);
-    result->output()->setType(type);
-    return result;
-  }
-
-  Node* Graph::createSetAttr(
-      Value * obj, const std::string& field, Value* newValue) {
-    auto n = create(prim::SetAttr, {obj, newValue}, /*num_outputs=*/0);
-    n->s_(attr::name, field);
-    return n;
-  }
-
-  Node* Graph::createGetAttr(Value * obj, const std::string& field) {
-    const auto classType = obj->type()->expect<ClassType>();
-
-    auto n = create(prim::GetAttr, {obj}, /*num_outputs=*/1);
-    n->s_(attr::name, field);
-
-    const auto outputType = classType->getAttribute(field);
-    n->output()->setType(outputType);
-    n->output()->setDebugName(normalizeAttrName(field));
-    return n;
-  }
-
-  Node* Graph::createStore(const std::string& name, Value* v) {
-    auto n = create(prim::Store, {v}, /*num_outputs*/ 0);
-    n->s_(attr::name, name);
-    return n;
-  }
-
-  Node* Graph::createLoad(const std::string& name, const TypeWrapper& type) {
-    auto n = create(prim::Load, {}, /*num_outputs*/ 1);
-    n->s_(attr::name, name);
-    n->output()->setType(type);
-    return n;
-  }
-
-  Node* Graph::createIsInstance(Value * v, at::ArrayRef<TypeWrapper> types) {
-    auto n = create(prim::isinstance, {v}, /*num_outputs*/ 1);
-    n->tys_(attr::types, types.vec());
-    n->output()->setType(BoolType::get());
-    return n;
-  }
-  Value* Graph::insertUncheckedCast(Value * v, const TypeWrapper& type) {
-    Node* n = insertNode(create(prim::unchecked_cast, {v}));
-    n->output()->setType(type);
-    return n->output();
-  }
-
-  Value* Graph::insertToList(Value * v, TypePtr type) {
-    int dim = 0;
-    TypePtr ptr = type;
-
-    // Unwrap the type to determine the number of dimensions.
-    while (auto list_type = ptr->cast<ListType>()) {
-      ptr = list_type->getElementType();
-      ++dim;
-    }
-
-    // Encode the base element type as an integer.
-    int elem_ty = 0;
-    if (ptr == IntType::get()) {
-      elem_ty = 0;
-    } else if (ptr == FloatType::get()) {
-      elem_ty = 1;
-    } else if (ptr == BoolType::get()) {
-      elem_ty = 2;
-    } else if (ptr == ComplexType::get()) {
-      elem_ty = 3;
+  if (!new_cs_entries.count(raw_callstack_ptr)) {
+    if (new_node_cs) {
+      new_cs_entries[raw_callstack_ptr] = c10::make_intrusive<InlinedCallStack>(
+          *new_node_cs, callee, to_replace->sourceRange(), m_info);
     } else {
-      TORCH_CHECK(
-          false,
-          ptr->repr_str(),
-          " is not one of the supported element types for tolist: int, float, complex, bool");
-    }
-
-    // Pass in the number of dimensions and base element type as arguments
-    // to the op.
-    Value* dim_val = insertConstant(IValue(dim));
-    Value* elem_ty_val = insertConstant(IValue(elem_ty));
-    Node* n = insertNode(create(prim::tolist, {v, dim_val, elem_ty_val}));
-    n->output()->setType(std::move(type));
-    return n->output();
-  }
-
-  Value* Graph::insertFunctionCall(
-      torch::jit::Function * callee, const MatchedSchema& matched) {
-    std::string func_name = callee->name();
-    Value* fn_constant = insertNode(create(prim::Constant))
-                             ->s_(attr::name, func_name)
-                             ->output()
-                             ->setType(FunctionType::create(callee));
-    std::vector<Value*> inputs = {fn_constant};
-    inputs.insert(inputs.end(), matched.inputs.begin(), matched.inputs.end());
-    Value* result = insertNode(create(prim::CallFunction, inputs))
-                        ->output()
-                        ->setType(matched.return_types.at(0));
-    return result;
-  }
-
-  Value* Graph::insertMethodCall(
-      std::string method_name, const MatchedSchema& matched) {
-    Value* result = insertNode(create(prim::CallMethod, matched.inputs))
-                        ->s_(attr::name, std::move(method_name))
-                        ->output()
-                        ->setType(matched.return_types.at(0));
-    return result;
-  }
-
-  Node* Graph::createClone(
-      Node * n,
-      const std::function<Value*(Value*)>& value_map,
-      bool copy_blocks) {
-    // n can be from a different graph
-    Node* r = n->allocNewInstance(this);
-    for (auto o : n->outputs()) {
-      r->addOutput()->copyMetadata(o);
-    }
-    r->cloneFrom(n);
-    for (auto i : n->inputs()) {
-      r->addInput(value_map(i));
-    }
-    if (copy_blocks) {
-      for (auto b : n->blocks()) {
-        r->addBlock()->cloneFrom(b, value_map);
-      }
-    }
-    return r;
-  }
-
-  Value* Graph::insertConstant(
-      const IValue& val,
-      c10::optional<SourceRange> loc,
-      c10::optional<ScopePtr> scope) {
-    // GAUDI_JIT_TRACE;
-    return jit::insertConstant(*this, val, std::move(loc), std::move(scope));
-  }
-
-  std::string Graph::toString(bool print_source_info) const {
-    // GAUDI_JIT_TRACE;
-    std::ostringstream oss;
-    print(oss, print_source_info);
-    return oss.str();
-  }
-
-  Graph::~Graph() {
-    for (const Node* n : all_nodes) {
-      delete n;
-    }
-    for (const Value* v : all_values) {
-      delete v;
-    }
-    for (const Block* b : all_blocks) {
-      delete b;
+      new_cs_entries[raw_callstack_ptr] = c10::make_intrusive<InlinedCallStack>(
+          callee, to_replace->sourceRange(), m_info);
     }
   }
-
-  void Graph::freeNode(Node * n) {
-    auto it = all_nodes.find(n);
-    HABANA_ASSERT(it != all_nodes.end());
-    delete *it;
-    all_nodes.erase(it);
+  new_node->setCallStack(new_cs_entries.at(raw_callstack_ptr));
+  // We updated the inlined callstack of new_node.
+  // Same must be done for the nodes of the blocks of new_node.
+  // For example If node's block otherwise is not annotated appropriately.
+  for (auto block : new_node->blocks()) {
+    inlineCallStackOfBlock(block, new_cs_entries, callee, to_replace, m_info);
   }
-  void Graph::freeValue(Value * v) {
-    v->setDebugName("");
-    auto it = all_values.find(v);
-    HABANA_ASSERT(it != all_values.end());
-    delete *it;
-    all_values.erase(it);
-  }
-  void Graph::freeBlock(Block * b) {
-    auto it = all_blocks.find(b);
-    HABANA_ASSERT(it != all_blocks.end());
-    delete *it;
-    all_blocks.erase(it);
-  }
+}
 
-  at::ArrayRef<Value*> createTupleUnpack(Value * v) {
-    // small peephole optimization to ensure IntArrayRef attributes can still
-    // turn into constants e.g. in x.expand([3, 4])
-    if (v->node()->kind() == prim::TupleConstruct) {
-      return v->node()->inputs();
-    }
-    auto& g = *v->owningGraph();
-    return g.insertNode(g.createTupleUnpack(v))->outputs();
-  }
-
-  void inlineCallStackOfNode(
-      Node * n,
-      std::unordered_map<InlinedCallStack*, InlinedCallStackPtr> &
-          new_cs_entries,
-      torch::jit::Function * callee,
-      Node * to_replace,
-      c10::optional<ModuleInstanceInfo> m_info);
-
-  static void inlineCallStackOfBlock(
-      Block * b,
-      std::unordered_map<InlinedCallStack*, InlinedCallStackPtr> &
-          new_cs_entries,
-      torch::jit::Function * callee,
-      Node * to_replace,
-      c10::optional<ModuleInstanceInfo> m_info) {
-    for (auto n : b->nodes()) {
-      inlineCallStackOfNode(n, new_cs_entries, callee, to_replace, m_info);
-    }
-  }
-
-  void inlineCallStackOfNode(
-      Node * new_node,
-      std::unordered_map<InlinedCallStack*, InlinedCallStackPtr> &
-          new_cs_entries,
-      torch::jit::Function * callee,
-      Node * to_replace,
-      c10::optional<ModuleInstanceInfo> m_info) {
-    auto new_node_cs = new_node->callstack();
-
-    InlinedCallStack* raw_callstack_ptr =
-        new_node_cs ? new_node_cs->get() : nullptr;
-
-    if (!new_cs_entries.count(raw_callstack_ptr)) {
-      if (new_node_cs) {
-        new_cs_entries[raw_callstack_ptr] =
-            c10::make_intrusive<InlinedCallStack>(
-                *new_node_cs, callee, to_replace->sourceRange(), m_info);
-      } else {
-        new_cs_entries[raw_callstack_ptr] =
-            c10::make_intrusive<InlinedCallStack>(
-                callee, to_replace->sourceRange(), m_info);
-      }
-    }
-    new_node->setCallStack(new_cs_entries.at(raw_callstack_ptr));
-    // We updated the inlined callstack of new_node.
-    // Same must be done for the nodes of the blocks of new_node.
-    // For example If node's block otherwise is not annotated appropriately.
-    for (auto block : new_node->blocks()) {
-      inlineCallStackOfBlock(block, new_cs_entries, callee, to_replace, m_info);
-    }
-  }
-
-  std::vector<Value*> unpackOutputs(const std::vector<Value*>& outputs) {
-    std::vector<Value*> new_outputs;
-    if (outputs.size() != 1 ||
-        outputs.at(0)->type()->kind() != TupleType::Kind) {
-      return outputs;
-    }
-
-    auto tup = outputs[0];
-    for (Value* v : createTupleUnpack(tup)) {
-      new_outputs.emplace_back(v);
-    }
-    // if this was a peephole tuple unpack we can just get rid of
-    // the tuple construct here and prevent needing DCE
-    if (tup->node()->kind() == prim::TupleConstruct &&
-        !tup->node()->hasUses()) {
-      tup->node()->destroy();
-    }
-    return new_outputs;
-  }
-
-  std::vector<Node*> findAllNodes(
-      at::ArrayRef<Block*> array, Symbol kind, bool recurse) {
-    std::vector<Node*> ret;
-    for (auto block : array) {
-      findAllNodes(*block, kind, recurse, ret);
-    }
-    return ret;
-  }
-
-  std::vector<Node*> findAllNodes(Block & block, Symbol kind, bool recurse) {
-    return findAllNodes({&block}, kind, recurse);
-  }
-
-  std::vector<Node*> findAllNodes(Graph & g, Symbol kind, bool recurse) {
-    return findAllNodes(*g.block(), kind, recurse);
-  }
-
-  std::vector<Value*> insertGraph(
-      Graph & g,
-      Graph & callee,
-      ArrayRef<Value*> inputs,
-      std::unordered_map<Value*, Value*> & value_map) {
-    auto value_map_func = [&](Value* v) { return value_map.at(v); };
-    HABANA_ASSERT(callee.inputs().size() == inputs.size());
-    for (const auto i : c10::irange(inputs.size())) {
-      value_map[callee.inputs()[i]] = inputs[i];
-    }
-    for (auto* node : callee.nodes()) {
-      auto* new_node = g.insertNode(g.createClone(node, value_map_func));
-      for (size_t i = 0; i < node->outputs().size(); ++i) {
-        value_map[node->outputs()[i]] = new_node->outputs()[i];
-      }
-    }
-
-    std::vector<Value*> outputs;
-    for (auto* output : callee.outputs()) {
-      outputs.push_back(value_map_func(output));
-    }
-
+std::vector<Value*> unpackOutputs(const std::vector<Value*>& outputs) {
+  std::vector<Value*> new_outputs;
+  if (outputs.size() != 1 || outputs.at(0)->type()->kind() != TupleType::Kind) {
     return outputs;
   }
 
-  std::vector<Value*> insertGraph(
-      Graph & g, Graph & callee, ArrayRef<Value*> inputs) {
-    std::unordered_map<Value*, Value*> value_map;
-    return insertGraph(g, callee, inputs, value_map);
+  auto tup = outputs[0];
+  for (Value* v : createTupleUnpack(tup)) {
+    new_outputs.emplace_back(v);
   }
-
-  void ProfileOp::cloneFrom(Node * other_) {
-    Node::cloneFrom(other_);
-    auto other = other_->cast<ProfileOp>();
-    this->callback_ = other->getCallback();
+  // if this was a peephole tuple unpack we can just get rid of
+  // the tuple construct here and prevent needing DCE
+  if (tup->node()->kind() == prim::TupleConstruct && !tup->node()->hasUses()) {
+    tup->node()->destroy();
   }
+  return new_outputs;
+}
 
-  Node* ProfileOp::allocNewInstance(Graph * g) {
-    return new ProfileOp(g, {nullptr});
+std::vector<Node*> findAllNodes(
+    at::ArrayRef<Block*> array,
+    Symbol kind,
+    bool recurse) {
+  std::vector<Node*> ret;
+  for (auto block : array) {
+    findAllNodes(*block, kind, recurse, ret);
   }
+  return ret;
+}
 
-  void ProfileIValueOp::cloneFrom(Node * other_) {
-    Node::cloneFrom(other_);
-    auto other = other_->cast<ProfileIValueOp>();
-    this->callback_ = other->getCallback();
+std::vector<Node*> findAllNodes(Block& block, Symbol kind, bool recurse) {
+  return findAllNodes({&block}, kind, recurse);
+}
+
+std::vector<Node*> findAllNodes(Graph& g, Symbol kind, bool recurse) {
+  return findAllNodes(*g.block(), kind, recurse);
+}
+
+std::vector<Value*> insertGraph(
+    Graph& g,
+    Graph& callee,
+    ArrayRef<Value*> inputs,
+    std::unordered_map<Value*, Value*>& value_map) {
+  auto value_map_func = [&](Value* v) { return value_map.at(v); };
+  HABANA_ASSERT(callee.inputs().size() == inputs.size());
+  for (const auto i : c10::irange(inputs.size())) {
+    value_map[callee.inputs()[i]] = inputs[i];
   }
-
-  Node* ProfileIValueOp::allocNewInstance(Graph * g) {
-    return new ProfileIValueOp(g, {nullptr});
-  }
-
-  TypePtr NamedValue::type() const {
-    if (value_) {
-      return value_->type();
-    } else {
-      return ivalue_.type();
+  for (auto* node : callee.nodes()) {
+    auto* new_node = g.insertNode(g.createClone(node, value_map_func));
+    for (size_t i = 0; i < node->outputs().size(); ++i) {
+      value_map[node->outputs()[i]] = new_node->outputs()[i];
     }
   }
 
-  const Symbol ProfileOp::Kind = ::c10::prim::profile;
-  const Symbol ProfileIValueOp::Kind = ::c10::prim::profile_ivalue;
-
-  OperatorSet::OperatorSet(std::initializer_list<const char*> sig_literals) {
-    insert(sig_literals);
+  std::vector<Value*> outputs;
+  for (auto* output : callee.outputs()) {
+    outputs.push_back(value_map_func(output));
   }
 
-  std::vector<std::shared_ptr<torch::jit::Operator>> OperatorSet::getOps()
-      const {
-    std::vector<std::shared_ptr<torch::jit::Operator>> result;
-    for (const auto& kv : ops) {
-      auto ops_for_symbol = kv.second;
-      result.insert(result.end(), ops_for_symbol.begin(), ops_for_symbol.end());
-    }
-    return result;
-  }
+  return outputs;
+}
 
-  void OperatorSet::insert(std::initializer_list<const char*> sig_literals) {
-    for (const char* sig : sig_literals) {
-      auto op = torch::jit::getOperatorForLiteral(sig);
-      ops[Symbol::fromQualString(op->schema().name())].push_back(op);
-    }
-  }
+std::vector<Value*> insertGraph(
+    Graph& g,
+    Graph& callee,
+    ArrayRef<Value*> inputs) {
+  std::unordered_map<Value*, Value*> value_map;
+  return insertGraph(g, callee, inputs, value_map);
+}
 
-  bool Node::isMemberOf(const OperatorSet& os) const {
-    auto it = os.ops.find(kind());
-    if (it == os.ops.end()) {
-      return false;
-    }
-    for (auto& op : it->second) {
-      if (matches(op->schema())) {
-        return true;
-      }
-    }
+void ProfileOp::cloneFrom(Node* other_) {
+  Node::cloneFrom(other_);
+  auto other = other_->cast<ProfileOp>();
+  this->callback_ = other->getCallback();
+}
+
+Node* ProfileOp::allocNewInstance(Graph* g) {
+  return new ProfileOp(g, {nullptr});
+}
+
+void ProfileIValueOp::cloneFrom(Node* other_) {
+  Node::cloneFrom(other_);
+  auto other = other_->cast<ProfileIValueOp>();
+  this->callback_ = other->getCallback();
+}
+
+Node* ProfileIValueOp::allocNewInstance(Graph* g) {
+  return new ProfileIValueOp(g, {nullptr});
+}
+
+TypePtr NamedValue::type() const {
+  if (value_) {
+    return value_->type();
+  } else {
+    return ivalue_.type();
+  }
+}
+
+const Symbol ProfileOp::Kind = ::c10::prim::profile;
+const Symbol ProfileIValueOp::Kind = ::c10::prim::profile_ivalue;
+
+OperatorSet::OperatorSet(std::initializer_list<const char*> sig_literals) {
+  insert(sig_literals);
+}
+
+std::vector<std::shared_ptr<torch::jit::Operator>> OperatorSet::getOps() const {
+  std::vector<std::shared_ptr<torch::jit::Operator>> result;
+  for (const auto& kv : ops) {
+    auto ops_for_symbol = kv.second;
+    result.insert(result.end(), ops_for_symbol.begin(), ops_for_symbol.end());
+  }
+  return result;
+}
+
+void OperatorSet::insert(std::initializer_list<const char*> sig_literals) {
+  for (const char* sig : sig_literals) {
+    auto op = torch::jit::getOperatorForLiteral(sig);
+    ops[Symbol::fromQualString(op->schema().name())].push_back(op);
+  }
+}
+
+bool Node::isMemberOf(const OperatorSet& os) const {
+  auto it = os.ops.find(kind());
+  if (it == os.ops.end()) {
     return false;
   }
+  for (auto& op : it->second) {
+    if (matches(op->schema())) {
+      return true;
+    }
+  }
+  return false;
+}
 
 } // namespace habana_torch::jit
