From 2327ea57c7937cd366d3f44535dc64e9532df9e8 Mon Sep 17 00:00:00 2001
From: dkumbar <dkumbar@habana.ai>
Date: Thu, 14 Nov 2024 10:01:26 +0200
Subject: [PATCH] added hd2d_vd2d_sd regression in phytop

Change-Id: I73c3c353cfc98cc01b78b9b1f72d4aa12c0f1f18
---

diff --git a/h15a/h15a_job_params_dkumbar.groovy b/h15a/h15a_job_params_dkumbar.groovy
index a20730c..5a7a2b6 100644
--- a/h15a/h15a_job_params_dkumbar.groovy
+++ b/h15a/h15a_job_params_dkumbar.groovy
@@ -1,6 +1,6 @@
 def JobsConfig() {
     def jobsConfig = [
-         'h15a_vd2d_phy_top_regression': [
+         'h15a_n3n5_phy_top_regression': [
             'jobDescription': 'Regression environment for H15A VD2D PHYTOP',
             'jenkinsView': 'VLSI-H15A',
             'jobParameters': [
@@ -13,13 +13,13 @@
                 'CSV_FILE': 'die2die_verif/d2d_phy_verif/tb/tests/n3n5_phy_regression.csv',
 		        'BUILD_PERIODICALLY': 'H 22 * * * ',
                 'RERUN_FAILED': false,
-                'RERUN_FAILED_PARAMS': ' -lte 10 ',
+                'RERUN_FAILED_PARAMS': ' -lte 12 ',
                 'ADDITIONAL_COVERAGE_PARAMS': ' -db_name h15a_mactop',
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
 
-          'h15a_vd2d_phy_top_complex_mem_regression': [
+          'h15a_n3n5_phy_top_complex_mem_regression': [
                       'jobDescription': 'Regression environment for H15A VD2D PHYTOP',
                       'jenkinsView': 'VLSI-H15A',
                       'jobParameters': [
@@ -32,13 +32,13 @@
                           'CSV_FILE': 'die2die_verif/d2d_phy_verif/tb/tests/n3n5_phy_regression.csv',
           		        'BUILD_PERIODICALLY': 'H 22 * * * ',
                           'RERUN_FAILED': false,
-                          'RERUN_FAILED_PARAMS': ' -lte 10 ',
+                          'RERUN_FAILED_PARAMS': ' -lte 12 ',
                           'ADDITIONAL_COVERAGE_PARAMS': ' -db_name h15a_n3n5_complex_mem',
                           'RUN_ON_REPO_UPDATE': false
                       ]
                   ],
-          'h15a_vd2d_phy_top_sva_regression': [
-                               'jobDescription': 'Regression environment for H15A VD2D PHYTOP',
+          'h15a_n3n5_phy_top_sva_regression': [
+                               'jobDescription': 'Regression environment for H15A N3N5 PHYTOP',
                                'jenkinsView': 'VLSI-H15A',
                                'jobParameters': [
                                    'PROJECT_NAME': 'H15A',
@@ -50,11 +50,30 @@
                                    'CSV_FILE': 'die2die_verif/d2d_phy_verif/tb/tests/n3n5_phy_regression.csv',
                    		        'BUILD_PERIODICALLY': 'H 22 * * * ',
                                    'RERUN_FAILED': false,
-                                   'RERUN_FAILED_PARAMS': ' -lte 10 ',
+                                   'RERUN_FAILED_PARAMS': ' -lte 12 ',
                                    'ADDITIONAL_COVERAGE_PARAMS': ' -db_name h15a_n3n5_sva',
                                    'RUN_ON_REPO_UPDATE': false
                                ]
                            ],
+          'h15a_hd2d_vd2d_phytop_regression': [
+                               'jobDescription': 'Regression environment for H15A VD2D_HD2D_SD PHYTOP',
+                               'jenkinsView': 'VLSI-H15A',
+                               'jobParameters': [
+                                   'PROJECT_NAME': 'H15A',
+                                   'BRANCH_NAME': 'h15a_die2die',
+                                   'BLOCK_NAME': 'd2d_phy',
+                                   'PATH_NAME': 'die2die_verif',
+                                   'ADDITIONAL_RUN_PARAMS': ' +define+INCLUDE_RAL_OBJUTILS+SVT_AXI_MAX_NUM_MASTERS_2+H15A_REPO+H15A_D2D_PHY',
+                                   'MAILING_LIST': 'dkumbar@habana.ai cchandrasekaran@habana.ai  rprakash@habana.ai ',
+                                   'CSV_FILE': 'die2die_verif/d2d_phy_verif/tb/tests/hd2d_vd2d_sd_regression.csv',
+                   		        'BUILD_PERIODICALLY': 'H 22 * * * ',
+                                   'RERUN_FAILED': false,
+                                   'RERUN_FAILED_PARAMS': ' -lte 12 ',
+                                   'ADDITIONAL_COVERAGE_PARAMS': ' -db_name h15a_hd2d_vd2d_sva',
+                                   'RUN_ON_REPO_UPDATE': false
+                               ]
+                           ],
+
           'h15a_c2i_e2e_phy_regression': [
             'jobDescription': 'Regression environment for IO2COMPUTE E2E PHY',
             'jenkinsView': 'VLSI-h15A',
@@ -68,13 +87,13 @@
                 'CSV_FILE': 'die2die_verif/d2d_e2e_verif/tb/tests/c2id2d_e2e_with_phy_regression.csv',
 		        'BUILD_PERIODICALLY': 'H 22 * * * ',
                 'RERUN_FAILED': false,
-                'RERUN_FAILED_PARAMS': ' -lte 10 ',
+                'RERUN_FAILED_PARAMS': ' -lte 12 ',
                 'ADDITIONAL_COVERAGE_PARAMS': ' -db_name h15a_c2i_e2e',
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
 
-         'h15a_vd2d_phy_top_coverage_regression': [
+         'h15a_n3n5_phy_top_coverage_regression': [
             'jobDescription': 'Regression environment for H15A VD2D PHYTOP with code coverage enabled',
             'jenkinsView': 'VLSI-H15A',
             'jobParameters': [
@@ -87,7 +106,7 @@
                 'CSV_FILE': 'die2die_verif/d2d_phy_verif/tb/tests/n3n5_phy_regression.csv',
 		        'BUILD_PERIODICALLY': 'H 22 * * * ',
                 'RERUN_FAILED': false,
-                'RERUN_FAILED_PARAMS': ' -lte 10 ',
+                'RERUN_FAILED_PARAMS': ' -lte 12 ',
                 'ADDITIONAL_COVERAGE_PARAMS': ' -db_name h15a_phytop_coverage',
                 'RUN_ON_REPO_UPDATE': false
             ]
