From fe23b63bcdb85123256d7cc6a804ba5b329586c1 Mon Sep 17 00:00:00 2001
From: vkkodix <vkkodix@habana.ai>
Date: Sun, 17 Nov 2024 15:18:39 +0200
Subject: [PATCH] regressions of nic_cls and port

Change-Id: Ia545102ea51a37e211124b2e9c187b53e897bd02
---

diff --git a/h15b/h15b_job_params_vkkodix.groovy b/h15b/h15b_job_params_vkkodix.groovy
index e646a5f..f984961 100644
--- a/h15b/h15b_job_params_vkkodix.groovy
+++ b/h15b/h15b_job_params_vkkodix.groovy
@@ -38,7 +38,7 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
-		 'h15b_nic_cls_nic2_regression': [
+		 'h15b_nic_cls_nic12_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
             'jobParameters': [
@@ -46,26 +46,7 @@
                 'BRANCH_NAME': 'h15b_nic',
                 'BLOCK_NAME': 'full_nic',
                 'PATH_NAME': 'nic_verif',
-                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_2X200G+NIC3+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC" -af_cpus 4 -lte 6',
-                'MAILING_LIST': 'vkkodix@habana.ai ',
-                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
-                //'BUILD_PERIODICALLY': 'H H * * H',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
-                'RERUN_FAILED': false,
-                'RERUN_FAILED_PARAMS': '-lte 10',
-                'ADDITIONAL_COVERAGE_PARAMS': '',
-                'RUN_ON_REPO_UPDATE': false
-            ]
-        ],
-		 'h15b_nic_cls_nic3_regression': [
-            'jobDescription': 'Regression environment for nic cluster with default test',
-            'jenkinsView': 'VLSI-H15B',
-            'jobParameters': [
-                'PROJECT_NAME': 'H15B',
-                'BRANCH_NAME': 'h15b_nic',
-                'BLOCK_NAME': 'full_nic',
-                'PATH_NAME': 'nic_verif',
-                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_2X200G+NIC4+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC" -af_cpus 4 -lte 6',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_2X200G+NIC12+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC" -af_cpus 4 -lte 6',
                 'MAILING_LIST': 'vkkodix@habana.ai ',
                 'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
                 //'BUILD_PERIODICALLY': 'H H * * H',
@@ -115,7 +96,7 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
-		 'h15b_nic_cls_400G_nic2_regression': [
+		 'h15b_nic_cls_400G_nic12_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
             'jobParameters': [
@@ -123,7 +104,7 @@
                 'BRANCH_NAME': 'h15b_nic',
                 'BLOCK_NAME': 'full_nic',
                 'PATH_NAME': 'nic_verif',
-                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_1X400G+NIC3+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC" -af_cpus 4 -lte 6',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_1X400G+NIC12+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC" -af_cpus 4 -lte 6',
                 'MAILING_LIST': 'vkkodix@habana.ai ',
                 'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
                 //'BUILD_PERIODICALLY': 'H H * * H',
@@ -173,26 +154,6 @@
             ]
         ],
 
-		'h15b_nic_cls_rn_nic2_regression': [
-            'jobDescription': 'Regression environment for nic cluster with default test',
-            'jenkinsView': 'VLSI-H15B',
-            'jobParameters': [
-                'PROJECT_NAME': 'H15B',
-                'BRANCH_NAME': 'h15b_nic',
-                'BLOCK_NAME': 'full_nic',
-                'PATH_NAME': 'nic_verif',
-                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_2X200G+NIC3+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=3f" -af_cpus 4 -lte 6',
-                'MAILING_LIST': 'vkkodix@habana.ai ',
-                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
-                //'BUILD_PERIODICALLY': 'H H * * H',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
-                'RERUN_FAILED': false,
-                'RERUN_FAILED_PARAMS': '-lte 10',
-                'ADDITIONAL_COVERAGE_PARAMS': '',
-                'RUN_ON_REPO_UPDATE': false
-            ]
-        ],
-
 	 'h15b_nic_cls_rn_400G_nic0_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
@@ -201,7 +162,7 @@
                 'BRANCH_NAME': 'h15b_nic',
                 'BLOCK_NAME': 'full_nic',
                 'PATH_NAME': 'nic_verif',
-                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_2X200G+NIC1+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=f" -af_cpus 4 -lte 6',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_1X400G+NIC1+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=f" -af_cpus 4 -lte 6',
                 'MAILING_LIST': 'vkkodix@habana.ai ',
                 'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
                 //'BUILD_PERIODICALLY': 'H H * * H',
@@ -220,7 +181,7 @@
                 'BRANCH_NAME': 'h15b_nic',
                 'BLOCK_NAME': 'full_nic',
                 'PATH_NAME': 'nic_verif',
-                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_2X200G+NIC2+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=f" -af_cpus 4 -lte 6',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_1X400G+NIC2+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=f" -af_cpus 4 -lte 6',
                 'MAILING_LIST': 'vkkodix@habana.ai ',
                 'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
                 //'BUILD_PERIODICALLY': 'H H * * H',
@@ -231,7 +192,149 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
-		
+
+
+	 'h15b_nic_cls_rn_400G_nic0_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+PHY_LOOPBACK+define+SNPS_FAST_SIM_FFV+FAST_SIM+PORT_1X400G+NIC3+AWAVE_NO_POWER_PORTS+AWAVE_BUMPS_AS_LOGIC+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=3" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+ 'h15b_nic_cls_nic_0_combo_1_rn_lft_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+MAC_LOOPBACK+define+PHY_STUB+SNPS_FAST_SIM_FFV+PORT_2X200G+NIC3+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=3" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+ 'h15b_nic_cls_nic_0_combo_1_rn_lft_400G_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+MAC_LOOPBACK+define+PHY_STUB+SNPS_FAST_SIM_FFV+PORT_1X400G+NIC3+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=3" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+ 'h15b_nic_cls_nic_0_combo_1_rn_rgt_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+MAC_LOOPBACK+define+PHY_STUB+SNPS_FAST_SIM_FFV+PORT_2X200G+NIC4"  --rf "+REDUNDANT_NIC=3" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+ 'h15b_nic_cls_nic_0_combo_1_rn_rgt_400G_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+MAC_LOOPBACK+define+PHY_STUB+SNPS_FAST_SIM_FFV+PORT_1X400G+NIC4"  --rf "+REDUNDANT_NIC=3" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+ 'h15b_nic_cls_nic_1_combo_5_rn_lft_400G_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+MAC_LOOPBACK+define+PHY_STUB+SNPS_FAST_SIM_FFV+PORT_1X400G+NIC12+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=C" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+ 'h15b_nic_cls_nic_1_combo_5_rn_lft_200G_regression': [
+            'jobDescription': 'Regression environment for nic cluster with default test',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_nic',
+                'BLOCK_NAME': 'full_nic',
+                'PATH_NAME': 'nic_verif',
+                'ADDITIONAL_RUN_PARAMS': '--af "+define+MAC_LOOPBACK+define+PHY_STUB+SNPS_FAST_SIM_FFV+PORT_2X200G+NIC12+REDUNDENCY_ENABLE"  --rf "+REDUNDANT_NIC=C" -af_cpus 4 -lte 6',
+                'MAILING_LIST': 'vkkodix@habana.ai ',
+                'CSV_FILE': 'nic_verif/full_nic_verif/tb/tests/nic_cls_regression.csv',
+                //'BUILD_PERIODICALLY': 'H H * * H',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 6',// MINUTE(0-59) HOUR(0-23) DOM(1-31) MONTH(1-12) DOW(0-7);  H->ANY , *->EVERY
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': '-lte 10',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+
+
 		'h15b_port_env_200G_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
