From f2b0f81f2d177635003ecbb10eb8b87ebe27fb92 Mon Sep 17 00:00:00 2001
From: vkkodix <vkkodix@habana.ai>
Date: Sun, 17 Nov 2024 15:26:28 +0200
Subject: [PATCH] nic_cls and port regressions

Change-Id: I2e92f75da8d22f01d4a579e958111301ffff0871
---

diff --git a/h15b/h15b_job_params_vkkodix.groovy b/h15b/h15b_job_params_vkkodix.groovy
index f984961..113aaae 100644
--- a/h15b/h15b_job_params_vkkodix.groovy
+++ b/h15b/h15b_job_params_vkkodix.groovy
@@ -38,6 +38,7 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
+
 		 'h15b_nic_cls_nic12_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
@@ -77,6 +78,7 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
+
 		 'h15b_nic_cls_400G_nic1_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
@@ -96,7 +98,8 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
-		 'h15b_nic_cls_400G_nic12_regression': [
+
+		'h15b_nic_cls_400G_nic12_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
             'jobParameters': [
@@ -115,6 +118,8 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
+
+
 		 'h15b_nic_cls_rn_nic0_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
@@ -134,6 +139,7 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
+
 		 'h15b_nic_cls_rn_nic1_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
@@ -194,7 +200,7 @@
         ],
 
 
-	 'h15b_nic_cls_rn_400G_nic0_regression': [
+	 'h15b_nic_cls_rn_400G_nic0_combo1_lft_regression': [
             'jobDescription': 'Regression environment for nic cluster with default test',
             'jenkinsView': 'VLSI-H15B',
             'jobParameters': [
