From 38ac19b9876e249f3d9a4bca4d0b51ed606ff962 Mon Sep 17 00:00:00 2001
From: dkumbar <dkumbar@habana.ai>
Date: Wed, 20 Nov 2024 10:38:28 +0200
Subject: [PATCH] updates pushe related to build failure

Change-Id: I290705bfe48fffa846f3dcd49e17175c44f66bd9
---

diff --git a/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy b/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy
index 4ee4897..db0d892 100644
--- a/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy
+++ b/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy
@@ -1,7 +1,7 @@
  
 def JobsConfig() {
     def jobsConfig = [
-            'h15b_sharedlib_d2d_phy_regression': [
+            'sharedlib_h15b_d2d_phy_top_regression': [
             'jobDescription': 'Create external database for N3N5_D2D_PHY setup',
             'jenkinsView': 'SHAREDLIB-H15',
             'jobParameters': [
@@ -14,7 +14,7 @@
                 'RUN_ON_REPO_UPDATE': true                           
             ]
          ],
-            'h15b_sharedlib_d2d_e2e_with_phy': [
+            'sharedlib_h15b_d2d_e2e_with_phy_regression': [
             'jobDescription': 'Create external database for C2I_D2D_E2E WITH PHY setup',
             'jenkinsView': 'SHAREDLIB-H15',
             'jobParameters': [
@@ -28,7 +28,7 @@
             ]
          ],
 
-            'h15b_sharedlib_d2d_e2e_macb2b': [
+            'sharedlib_h15b_d2d_e2e_macb2b_regression': [
             'jobDescription': 'Create external database for C2I_D2D_E2E MAC B2B setup',
             'jenkinsView': 'SHAREDLIB-H15',
             'jobParameters': [
@@ -42,7 +42,7 @@
             ]
          ],
 
-            'h15b_sharedlib_d2d_e2e_pldm': [
+            'sharedlib_h15b_d2d_e2e_pldm_regression': [
             'jobDescription': 'Create external database for C2I_D2D_E2E PLDM setup',
             'jenkinsView': 'SHAREDLIB-H15',
             'jobParameters': [
