From c5e690fdee85b0914e901667cb114267f921959a Mon Sep 17 00:00:00 2001
From: dkumbar <dkumbar@habana.ai>
Date: Wed, 20 Nov 2024 09:42:54 +0200
Subject: [PATCH] Added sharedlib generation jobs in H15B

Change-Id: Icd8dd537f16122efb4b1916ab07ff40993df34a5
---

diff --git a/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy b/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy
new file mode 100644
index 0000000..4ee4897
--- /dev/null
+++ b/sharedlib_h15/sharedlib_h15_jobs_params_dkumbar.groovy
@@ -0,0 +1,61 @@
+ 
+def JobsConfig() {
+    def jobsConfig = [
+            'h15b_sharedlib_d2d_phy_regression': [
+            'jobDescription': 'Create external database for N3N5_D2D_PHY setup',
+            'jenkinsView': 'SHAREDLIB-H15',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_die2die',
+                'BLOCK_NAME': 'd2d_phy',
+                'PATH_NAME': 'die2die_verif',
+                'ADDITIONAL_RUN_PARAMS': '-af " +define+INCLUDE_RAL_OBJUTILS+SVT_AXI_MAX_NUM_MASTERS_1" -xprop  -lte 1 -create_rtl_ext -wd H15B_D2D_PHY_DB -co ',
+                'MAILING_LIST': 'fbathurudeen@habana.ai cchandrasekaran@habana.ai rprakash@habana.ai  vdegalahal@habana.ai',
+                'RUN_ON_REPO_UPDATE': true                           
+            ]
+         ],
+            'h15b_sharedlib_d2d_e2e_with_phy': [
+            'jobDescription': 'Create external database for C2I_D2D_E2E WITH PHY setup',
+            'jenkinsView': 'SHAREDLIB-H15',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_die2die',
+                'BLOCK_NAME': 'd2d_e2e',
+                'PATH_NAME': 'die2die_verif',
+                'ADDITIONAL_RUN_PARAMS': '-af "   +define+D2D_ULT+D2D_0" -xprop  -lte 7 -create_rtl_ext -wd H15B_D2D_E2E_WITH_PHY -co',
+                'MAILING_LIST': 'fbathurudeen@habana.ai cchandrasekaran@habana.ai rprakash@habana.ai  vdegalahal@habana.ai',
+                'RUN_ON_REPO_UPDATE': true                           
+            ]
+         ],
+
+            'h15b_sharedlib_d2d_e2e_macb2b': [
+            'jobDescription': 'Create external database for C2I_D2D_E2E MAC B2B setup',
+            'jenkinsView': 'SHAREDLIB-H15',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_die2die',
+                'BLOCK_NAME': 'd2d_e2e',
+                'PATH_NAME': 'die2die_verif',
+                'ADDITIONAL_RUN_PARAMS': '-af " +define+MACTOP_B2B_TB+D2D_ULT+D2D_0" -xprop  -lte 7 -create_rtl_ext -wd H15B_D2D_E2E_MAC_B2B -co ',
+                'MAILING_LIST': 'fbathurudeen@habana.ai cchandrasekaran@habana.ai rprakash@habana.ai  vdegalahal@habana.ai',
+                'RUN_ON_REPO_UPDATE': true                           
+            ]
+         ],
+
+            'h15b_sharedlib_d2d_e2e_pldm': [
+            'jobDescription': 'Create external database for C2I_D2D_E2E PLDM setup',
+            'jenkinsView': 'SHAREDLIB-H15',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_die2die',
+                'BLOCK_NAME': 'd2d_e2e',
+                'PATH_NAME': 'die2die_verif',
+                'ADDITIONAL_RUN_PARAMS': '-af " +define+INCLUDE_RAL_OBJUTILS+SVT_AXI_MAX_NUM_MASTERS_1+HL_DFT_H15A_REMOVE_INTERFACE_IDD+D2D_ULT+H15A_C2ID2D+D2D_E2E_PHY_ENV+IPTD2DPHY_BYPASS_PMA+EMULATION+HL_PLDM_TOP_TB=top_tb+D2D_FASTSIM " -xprop -lte 7 -create_rtl_ext -wd H15B_D2D_E2E_PLDM -co ',
+                'MAILING_LIST': 'fbathurudeen@habana.ai cchandrasekaran@habana.ai rprakash@habana.ai  vdegalahal@habana.ai',
+                'RUN_ON_REPO_UPDATE': true                           
+            ]
+         ]
+    ]
+    return jobsConfig
+}
+
