From ec0fdca407df7c74dbbb5c72bcee0237fe4760bc Mon Sep 17 00:00:00 2001
From: Sailaja <sbandarupalli@habana.ai>
Date: Sun, 24 Nov 2024 08:17:51 +0200
Subject: [PATCH] [SW-0] enable boot info sync config at preboot

Change-Id: I618361f58b972b5da6bdbcc3c89e0ae154e24706
Signed-off-by: Sailaja <sbandarupalli@habana.ai>
---

diff --git a/dts/arc/hl_gaudi3_preboot.dtsi b/dts/arc/hl_gaudi3_preboot.dtsi
index 4dd329f..21c8ee7 100644
--- a/dts/arc/hl_gaudi3_preboot.dtsi
+++ b/dts/arc/hl_gaudi3_preboot.dtsi
@@ -6,10 +6,14 @@
  */
 
 #include "hl_gaudi3.dtsi"
+#include <hl_gaudi3_sem.dtsi>
+#include <hl_sem_cpuid.dtsi>
+
 #include <hl_general_oper.h>
 
 / {
 	chosen {
+		semaphore,cpuid = &mgmt_cpuid;
 		zephyr,console = &uart0;
 		zephyr,shell-uart = &uart0;
 	};
