From 024c6dcecdc0db79f50be360dcc5d2bf9e9ba083 Mon Sep 17 00:00:00 2001
From: fbathurudeen <fbathurudeen@habana.ai>
Date: Fri, 22 Nov 2024 03:19:02 +0200
Subject: [PATCH] changed to h15b

Change-Id: Ia850ec7e4f5c8ed2df159084c0e5932f8e65cacd
---

diff --git a/sharedlib_h15/sharedlib_h15_jobs_params_fbathurudeen.groovy b/sharedlib_h15/sharedlib_h15_jobs_params_fbathurudeen.groovy
index 550c317..c507eeb 100644
--- a/sharedlib_h15/sharedlib_h15_jobs_params_fbathurudeen.groovy
+++ b/sharedlib_h15/sharedlib_h15_jobs_params_fbathurudeen.groovy
@@ -4,8 +4,8 @@
             'jobDescription': 'Sharedlib test for h15b',
             'jenkinsView': 'SHAREDLIB-H15',
             'jobParameters': [
-                'PROJECT_NAME': 'H15',
-                'BRANCH_NAME': 'h15_fullchip',
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
                 'BLOCK_NAME': 'io_die_direct',
                 'PATH_NAME': 'fullchip_verif',
                 'ADDITIONAL_RUN_PARAMS': '-t io_die_direct_base_test -ct main_test -s 1  -args_defs " USE_PSOC_AGENT +REAL_PLL +PSOC +ALL_RTL +NO_PARTCOMP" -wd h15b_compiled_db -rmem_256g  -af_rmem_64g  -create_rtl_ext -co',
