From ffc1f93e22810a89e072f5c1931c7075b069a812 Mon Sep 17 00:00:00 2001
From: Or Cohav Lilo <olilo@habana.ai>
Date: Sun, 24 Nov 2024 20:36:20 +0200
Subject: [PATCH] Stress 6 added

Change-Id: Ie4c22b1a348900d17593a8633e27cc31961a525e
---

diff --git a/h15a/h15a_job_params_olilo.groovy b/h15a/h15a_job_params_olilo.groovy
index f6bfb4e..7087b5a 100644
--- a/h15a/h15a_job_params_olilo.groovy
+++ b/h15a/h15a_job_params_olilo.groovy
@@ -18,26 +18,8 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
-	'h15a_rtr_ctrl_stress_regression': [
-            'jobDescription': ' Regression environment fo  rtr_ctrl - stress random',
-            'jenkinsView': 'VLSI-H15A',
-            'jobParameters': [
-                'PROJECT_NAME': 'H15a',
-                'BRANCH_NAME': 'h15a_mesh',
-                'BLOCK_NAME': 'rtr_ctrl',
-                'PATH_NAME': 'mesh_ic_verif',
-                'ADDITIONAL_RUN_PARAMS': '-stub_defs "+NO_RTR_CTRL_SCBD+RTR_CTRL_BLOCK_LEVEL+INCLUDE_RAL_OBJUTILS+NO_RTR_CTRL_SCBD+RTR_CTRL_BLOCK_LEVEL+INCLUDE_RAL_OBJUTILS+SRAMBLE_EN+PLOY_RAND+WR_ONLY+RD_ONLY+STRESS_TEST"  -af "+H15A_PROJECT"',
-                'MAILING_LIST': 'olilo@habana.ai ',
-                'CSV_FILE': 'mesh_ic_verif/rtr_ctrl_verif/tb/tests/rtr_ctrl_regression.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * 3,6',
-                'RERUN_FAILED': true,
-                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi -lte 3',
-                'ADDITIONAL_COVERAGE_PARAMS': '',
-                'RUN_ON_REPO_UPDATE': false
-            ]
-        ],
         'h15a_rtr_ctrl_noscram_regression': [
-            'jobDescription': 'Regression environment for  rtr_ctrl - random',
+            'jobDescription': 'Regression environment for  rtr_ctrl  - random',
             'jenkinsView': 'VLSI-H15A',
             'jobParameters': [
                 'PROJECT_NAME': 'H15a',
