From d14a0a1db1caf1704426619a61e38aa7f5b4165f Mon Sep 17 00:00:00 2001
From: Or Cohav Lilo <olilo@habana.ai>
Date: Sun, 24 Nov 2024 20:26:08 +0200
Subject: [PATCH] added stress2

Change-Id: I1f0e4786259a240d2fd32259049b3a16d25e15d7
---

diff --git a/h15a/h15a_job_params_olilo.groovy b/h15a/h15a_job_params_olilo.groovy
index 5ce4f76..f6bfb4e 100644
--- a/h15a/h15a_job_params_olilo.groovy
+++ b/h15a/h15a_job_params_olilo.groovy
@@ -18,8 +18,8 @@
                 'RUN_ON_REPO_UPDATE': false
             ]
         ],
-	'h15a_rtr_ctrl_regression_stress': [
-            'jobDescription': ' Regression environment for  rtr_ctrl - stress random',
+	'h15a_rtr_ctrl_stress_regression': [
+            'jobDescription': ' Regression environment fo  rtr_ctrl - stress random',
             'jenkinsView': 'VLSI-H15A',
             'jobParameters': [
                 'PROJECT_NAME': 'H15a',
