From e6f24c3f1a05a0cea4b658e53afe5c36887a6813 Mon Sep 17 00:00:00 2001
From: Erez Shaul <eshaul@habana.ai>
Date: Sun, 01 Dec 2024 11:44:37 +0200
Subject: [PATCH] [SW-207386] hl-smi-async: async: remove redundant comment

remove redundant comment

Change-Id: I55f1536ba64d295fc1597e9cc4c8658f010fb707
Signed-off-by: Erez Shaul <eshaul@habana.ai>
---

diff --git a/app/hl-smi-async/hl-smi-async.cpp b/app/hl-smi-async/hl-smi-async.cpp
index 7474319..a2eb8e2 100644
--- a/app/hl-smi-async/hl-smi-async.cpp
+++ b/app/hl-smi-async/hl-smi-async.cpp
@@ -207,7 +207,6 @@
 	if (ret || !present)
 		return HLML_ERROR_INT_PCI_ERROR;
 
-	/* Read VENDOR_ID through elbi */
 	if (is_pci_lib_initialized() && present) {
 		/* map pci memory bar */
 		bar4_addr = hl_smi_map_pci_bar(opts.device, SRAM_MEMORY_BAR_ID);
