From bb2c452a57e6fbfb6b77b4116436a350fcecc96f Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Sun, 03 Nov 2024 16:41:46 +0200
Subject: [PATCH] WIP

Change-Id: Ie76d5809613b9bbde650d414ee84876947523e3c
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/scripts/make/hl-common/Makefile b/scripts/make/hl-common/Makefile
index 171be88..60541ce 100644
--- a/scripts/make/hl-common/Makefile
+++ b/scripts/make/hl-common/Makefile
@@ -127,6 +127,8 @@
 		$(MAKE) $(patsubst %-build,%-prepare,$@))
 	@$(if $(filter $(HL_MULTI_ARCH), $@), \
 		$(MAKE) $(call hl-device, $@)-post-image)
+	@$(if $(filter $(HL_TARGET_ARCH), $@), \
+		$(MAKE) $(patsubst %-build,%-prepare,$@))
 
 HL_TARGET_ARCH_CLEAN := $(addsuffix -%,$(HL_SINGLE_ARCH_RELEASE)) \
 			$(addsuffix _%,$(HL_MULTI_ARCH_RELEASE))
