From d501bb61915db24da11c78c3b802b7704bfe0a9d Mon Sep 17 00:00:00 2001
From: amohrilx <amohrilx@habana.ai>
Date: Mon, 09 Dec 2024 09:20:40 +0200
Subject: [PATCH] .

Change-Id: I19cf19b49640087a93925d6109fe46770af046f1
---

diff --git a/h15b/h15b_job_params_amohrilx.groovy b/h15b/h15b_job_params_amohrilx.groovy
index 8f143f2..8c10035 100644
--- a/h15b/h15b_job_params_amohrilx.groovy
+++ b/h15b/h15b_job_params_amohrilx.groovy
@@ -8,7 +8,7 @@
                 'BRANCH_NAME': 'h15b_video',
                 'BLOCK_NAME': 'video_top',
                 'PATH_NAME': 'video_verif',
-                'ADDITIONAL_RUN_PARAMS': '-stub_defs VDEC_TOP -af "+define+VDEC_TOP+INCLUDE_RAL_OBJUTILS" -lte 9 -xprop -cov -code_cov',
+                'ADDITIONAL_RUN_PARAMS': '-stub_defs VDEC_TOP -af "+define+VDEC_TOP+INCLUDE_RAL_OBJUTILS" -lte 72 -xprop -cov -code_cov -sva',
                 'MAILING_LIST': 'amohrilx@habana.ai vkkodix@habana.ai',
                 'CSV_FILE': 'video_verif/video_top_verif/tb/tests/video_top_dec_regression.csv',
   		        'BUILD_PERIODICALLY': 'H H(0-23) * * 1,4',
@@ -26,7 +26,7 @@
                 'BRANCH_NAME': 'h15b_video',
                 'BLOCK_NAME': 'video_top',
                 'PATH_NAME': 'video_verif',
-                'ADDITIONAL_RUN_PARAMS': '-stub_defs VENC_TOP -af "+define+VENC_TOP+INCLUDE_RAL_OBJUTILS" -lte 9 -xprop -cov -code_cov',
+                'ADDITIONAL_RUN_PARAMS': '-stub_defs VENC_TOP -af "+define+VENC_TOP+INCLUDE_RAL_OBJUTILS" -lte 72 -sva -xprop -cov -code_cov',
                 'MAILING_LIST': 'amohrilx@habana.ai vkkodix@habana.ai',
                 'CSV_FILE': 'video_verif/video_top_verif/tb/tests/video_top_enc_regression.csv',
   		        'BUILD_PERIODICALLY': 'H H(0-23) * * 1,4',
@@ -44,7 +44,7 @@
                 'BRANCH_NAME': 'h15b_video',
                 'BLOCK_NAME': 'video_top',
                 'PATH_NAME': 'video_verif',
-                'ADDITIONAL_RUN_PARAMS': '-stub_defs "VENC_TOP +VENC_BRDG +HL_MACRO0_VC9000E_IS_STUB +HL_VENC1_MACRO_IS_STUB +HL_VENC2_MACRO_IS_STUB +HL_VENC3_MACRO_IS_STUB" -af "+define+VENC_TOP+VENC_BRDG+INCLUDE_RAL_OBJUTILS" -lte 9 -xprop -cov -code_cov',
+                'ADDITIONAL_RUN_PARAMS': '-stub_defs "VENC_TOP +VENC_BRDG +HL_MACRO0_VC9000E_IS_STUB +HL_VENC1_MACRO_IS_STUB +HL_VENC2_MACRO_IS_STUB +HL_VENC3_MACRO_IS_STUB" -af "+define+VENC_TOP+VENC_BRDG+INCLUDE_RAL_OBJUTILS" -lte 20 -xprop -cov -code_cov -sva',
                 'MAILING_LIST': 'amohrilx@habana.ai vkkodix@habana.ai',
                 'CSV_FILE': 'video_verif/video_top_verif/tb/tests/bridge_regression.csv',
   		        'BUILD_PERIODICALLY': 'H H(0-23) * * 1,6',
@@ -62,7 +62,7 @@
                 'BRANCH_NAME': 'h15b_fullchip',
                 'BLOCK_NAME': 'io_die_direct',
                 'PATH_NAME': 'fullchip_verif',
-                'ADDITIONAL_RUN_PARAMS': '-args_defs " USE_PSOC_AGENT +PSOC +DTSTDVM +MEDIA" -af "+define+IO_DIE0+MEM_MANAGER_IS_INCLUDED+IO_AXI_AGENT+INDP_BRIDGE" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +DTSTDVM +INDP_BRIDGE" -lte 30 -rmem_256g -af_rmem_64g -lte 9 -xprop -cov -code_cov',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs " USE_PSOC_AGENT +PSOC +DTSTDVM +MEDIA" -af "+define+IO_DIE0+MEM_MANAGER_IS_INCLUDED+IO_AXI_AGENT+INDP_BRIDGE" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +DTSTDVM +INDP_BRIDGE" -lte 30 -rmem_256g -af_rmem_64g -lte 9 -xprop -cov -code_cov -sva',
                 'MAILING_LIST': 'amohrilx@habana.ai vkkodix@habana.ai',
                 'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/media_regression.csv',
   		        'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
@@ -80,7 +80,7 @@
                 'BRANCH_NAME': 'h15b_fullchip',
                 'BLOCK_NAME': 'io_die_direct',
                 'PATH_NAME': 'fullchip_verif',
-                'ADDITIONAL_RUN_PARAMS': '-args_defs " USE_PSOC_AGENT +PSOC +DTSTDVM +MEDIA" -af "+define+IO_DIE0+MEM_MANAGER_IS_INCLUDED+IO_AXI_AGENT+INDP_BRIDGE" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +DTSTDVM +INDP_BRIDGE" -lte 30 -rmem_256g -af_rmem_64g -lte 9 -xprop -cov -code_cov ',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs " USE_PSOC_AGENT +PSOC +DTSTDVM +MEDIA" -af "+define+IO_DIE0+MEM_MANAGER_IS_INCLUDED+IO_AXI_AGENT+INDP_BRIDGE" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +DTSTDVM +INDP_BRIDGE" -lte 30 -rmem_256g -af_rmem_64g -lte 9 -xprop -cov -code_cov -sva',
                 'MAILING_LIST': 'amohrilx@habana.ai vkkodix@habana.ai',
                 'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/media_enc_regression.csv',
   		        'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
