From 67805043c4358fbcc0cd7d66f3ddb7ca9fdf9da4 Mon Sep 17 00:00:00 2001
From: Alexey Nikandrov <anikandrov@habana.ai>
Date: Sun, 15 Dec 2024 11:52:58 +0200
Subject: [PATCH] [SW-213020] cd: docker - U22 update libfabric version

Change-Id: Id77466f2fef043c17c6b443f6c188c55977993d8
Signed-off-by: Alexey Nikandrov <anikandrov@habana.ai>
---

diff --git a/dockerfiles/base/Dockerfile.ubuntu22.04 b/dockerfiles/base/Dockerfile.ubuntu22.04
index 283bbf0..48085d6 100644
--- a/dockerfiles/base/Dockerfile.ubuntu22.04
+++ b/dockerfiles/base/Dockerfile.ubuntu22.04
@@ -65,7 +65,7 @@
 COPY install_efa.sh .
 RUN ./install_efa.sh && rm install_efa.sh && rm -rf /etc/ld.so.conf.d/efa.conf /etc/profile.d/efa.sh
 
-ENV LIBFABRIC_VERSION="1.22.0"
+ENV LIBFABRIC_VERSION="2.0.0"
 ENV LIBFABRIC_ROOT="/opt/habanalabs/libfabric-${LIBFABRIC_VERSION}"
 ENV MPI_ROOT=/opt/amazon/openmpi
 ENV LD_LIBRARY_PATH=$LIBFABRIC_ROOT/lib:${MPI_ROOT}/lib:/usr/lib/habanalabs:$LD_LIBRARY_PATH
