From 4fc95786220119a82dd6bb61dfff2f14837ad5d2 Mon Sep 17 00:00:00 2001
From: Rajesh Sanga <rsanga@habana.ai>
Date: Wed, 18 Dec 2024 10:57:31 +0200
Subject: [PATCH] [SW-213381] Change the user for kubeconfig

Change-Id: I475902971619b6ee47039410008581cf752fc654
Signed-off-by: Rajesh Sanga <rsanga@habana.ai>
---

diff --git a/Jenkinsfile b/Jenkinsfile
index 8bc1a16..c1fd1fb 100644
--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -81,7 +81,7 @@
                     def rc = 0
                     try {
                         if (!isStaging) { sendMail("Experiment session - started") }
-                        withCredentials([usernamePassword(credentialsId: constants.SWJENKINS_CREDENTIALS_ID,
+                        withCredentials([usernamePassword(credentialsId: '8a42a110-b7db-4939-afe8-ee46a7c9f81f',
                             passwordVariable: 'USER_PASSWORD', usernameVariable: 'USER_NAME')]) {
                             def cmd = '''#!/bin/bash
                             ./scripts/init_k8s.sh --exp-file output/experiment.json &
diff --git a/scripts/init_k8s.sh b/scripts/init_k8s.sh
index 3d89234..b7304e3 100755
--- a/scripts/init_k8s.sh
+++ b/scripts/init_k8s.sh
@@ -69,7 +69,9 @@
     export BUILD_URL=${BUILD_URL:-'some_url'}
     curl -sS https://artifactory-kfs.habana-labs.com:443/artifactory/k8s-infra-generic-dev-local/hlctl/download-latest.sh | bash
     set +x
-    ./hlctl kubeconfig --username $USER_NAME --password "$USER_PASSWORD"
+    echo "$USER_NAME"
+    echo "$USER_PASSWORD"
+    ./hlctl kubeconfig --username "qauser" --password "jMEdmg!u67n8V%*qa"
     [ -n "$DEBUG" ] && set -x
 }
 
