From ce3335ac9ac92f660a7a339a64589c7c77733333 Mon Sep 17 00:00:00 2001
From: lnagabhushan <lnagabhushan@habana.ai>
Date: Thu, 19 Dec 2024 17:50:04 +0200
Subject: [PATCH] h15a_job_params_lnagabhushan.groovy

Change-Id: If5ed53ef3aedc7733945edb284afdaac01349e1e
---

diff --git a/h15a/h15a_job_params_lnagabhushan.groovy b/h15a/h15a_job_params_lnagabhushan.groovy
new file mode 100644
index 0000000..f0b7d2b
--- /dev/null
+++ b/h15a/h15a_job_params_lnagabhushan.groovy
@@ -0,0 +1,41 @@
+def JobsConfig() {
+    def jobsConfig = [
+        'h15a_single_cache_coral_regression': [
+            'jobDescription': 'CACHE CORAL Regression',
+            'jenkinsView': 'VLSI-H15A',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15A',
+                'BRANCH_NAME': 'h15a_cache',
+                'BLOCK_NAME': 'cache',
+                'PATH_NAME': 'cache_verif',
+                'ADDITIONAL_RUN_PARAMS': ' -args_defs "CACHE_TOP" -coral_sim cache -lte 10 -rmem_8g -cov ',
+                'MAILING_LIST': 'yzarko@habana.ai rkehat@habana.ai lnagabhushan@habana.ai',
+                'CSV_FILE': 'cache_verif/cache_verif/tb/csv/cache_coral_regression.csv',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 1,5',
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': ' -args_defs "CACHE_TOP" -coral_sim cache -lte 10 -rmem_8g -verdi -v UVM_MEDIUM ',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ],
+        'h15a_single_cache_simple_regression': [
+            'jobDescription': 'CACHE Simple Regression',
+            'jenkinsView': 'VLSI-H15A',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15A',
+                'BRANCH_NAME': 'h15a_cache',
+                'BLOCK_NAME': 'cache',
+                'PATH_NAME': 'cache_verif',
+                'ADDITIONAL_RUN_PARAMS': ' -args_defs "CACHE_TOP" -ignore_c_comp -lte 10 -rmem_8g -cov ',
+                'MAILING_LIST': 'yzarko@habana.ai rkehat@habana.ai lnagabhushan@habana.ai',
+                'CSV_FILE': 'cache_verif/cache_verif/tb/csv/cache_simple_regression.csv',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 1,5',
+                'RERUN_FAILED': false,
+                'RERUN_FAILED_PARAMS': ' -args_defs "CACHE_TOP" -ignore_c_comp -lte 10 -rmem_8g -verdi -v UVM_MEDIUM ',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': false
+            ]
+        ]
+    ]
+    return jobsConfig
+}
