From ce4c4950d3c2ae487d236037ae026ba7f8912c95 Mon Sep 17 00:00:00 2001
From: Avi <amarom@habana.ai>
Date: Mon, 30 Dec 2024 15:12:05 +0200
Subject: [PATCH] [HLB][Tests] Add mark HLB_margin

Change-Id: Ic308add2c0e614e4f3bf82a5956e6efc6bdde7da
Signed-off-by: Avi <amarom@habana.ai>
---

diff --git a/tests/HLS3/test_information.py b/tests/HLS3/test_information.py
index 6fb3b34..31df4d8 100644
--- a/tests/HLS3/test_information.py
+++ b/tests/HLS3/test_information.py
@@ -27,7 +27,7 @@
 
 logger: Logger = logging.getLogger(__name__)
 
-
+@pytest.mark.HLB_margin
 @pytest.mark.HLS3_pt_bmc
 @pytest.mark.HLS3_pt_bmc_exp
 @pytest.mark.HLS3_pt_i2c
