From 4cc836ab182baa0a64d5ad867ff61116d77629f5 Mon Sep 17 00:00:00 2001
From: Gennady Davidov <gdavidov@habana.ai>
Date: Sun, 05 Jan 2025 17:28:23 +0200
Subject: [PATCH] [SW-214603] tests: pt-run-docker-container.sh - add rocky

Change-Id: I17d8d5e4832893ab1fdb3558d3893cf20077c98e
Signed-off-by: Gennady Davidov <gdavidov@habana.ai>
---

diff --git a/tests/asic/PT-Suite/pt-run-docker-container.sh b/tests/asic/PT-Suite/pt-run-docker-container.sh
index 22fd326..1a71d3d 100755
--- a/tests/asic/PT-Suite/pt-run-docker-container.sh
+++ b/tests/asic/PT-Suite/pt-run-docker-container.sh
@@ -38,6 +38,10 @@
     ubuntu24.04)
         JOB_SCENARIO="pt_cd_u24"
     ;;
+    rocky9.3)
+        JOB_SCENARIO="pt_cd_rocky93"
+        HL_TEST_OS="ubuntu22.04"
+    ;;
     *)
         echo -n "unsupported OS ${HL_TEST_OS}"
         exit 1
