From f6f97c9311ce29373d842c2a2bd874a29250e933 Mon Sep 17 00:00:00 2001
From: Gennady Davidov <gdavidov@habana.ai>
Date: Wed, 22 Jan 2025 19:22:39 +0200
Subject: [PATCH] [SW-215815] tests:  hl-test-eager-performance - add FORCE_EAGER

Change-Id: I6a937c31340709f18ff41356d83c3d4faf4ca5fc
Signed-off-by: Gennady Davidov <gdavidov@habana.ai>
---

diff --git a/tests/asic/performance/hl-test-eager-performance.sh b/tests/asic/performance/hl-test-eager-performance.sh
index af9da77..1a33d33 100755
--- a/tests/asic/performance/hl-test-eager-performance.sh
+++ b/tests/asic/performance/hl-test-eager-performance.sh
@@ -12,6 +12,7 @@
 export HABANA_NPU_STACK_PATH=${HOME}
 export SYNAPSE_RELEASE_BUILD=/opt/habanalabs
 export SYNAPSE_ROOT=${HOME}/synapse
+export FORCE_EAGER=1
 
 TEST_NAME="${RELEASE_VERSION}-${RELEASE_BUILD_ID}"
 TEST_TYPE='perf_eager'
