From 1bbaa16aae1794de7b3fba347ee97cd5dfd5ef21 Mon Sep 17 00:00:00 2001
From: Swapnali Vakil <svakil@vlsi-int03.kfs.habana-labs.com>
Date: Fri, 24 Jan 2025 05:23:22 +0200
Subject: [PATCH] INTR Regression Commit

Change-Id: If683f83846c38994a8c258542d351c297a7edc5d
---

diff --git a/h15b/h15b_job_params_svakil.groovy b/h15b/h15b_job_params_svakil.groovy
index 72d4ad3..5c61381 100644
--- a/h15b/h15b_job_params_svakil.groovy
+++ b/h15b/h15b_job_params_svakil.groovy
@@ -35,7 +35,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
 
        'h15b_d2d_mac_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_D2D_MAC ',
@@ -54,7 +54,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
 
 	'h15b_d2d_phy_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_D2D_PHY ',
@@ -73,7 +73,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
 
 	'h15b_dtstdvm_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_DTSTDVM ',
@@ -92,7 +92,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
 
         'h15b_muc_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_MUC ',
@@ -111,7 +111,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
        
   	'h15b_nic_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_NIC ',
@@ -130,7 +130,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
 
 	'h15b_nms_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_NMS ',
@@ -149,7 +149,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
+         ],
 
 	'h15b_noc_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_NOC ',
@@ -168,11 +168,11 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ]
-
+         ],
+ 
 	'h15b_pll_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_PLL ',
-            'jenkinsView': 'VLSI-H15B',
+           'jenkinsView': 'VLSI-H15B',
             'jobParameters': [
                 'PROJECT_NAME': 'H15B',
                 'BRANCH_NAME': 'h15b_fullchip',
@@ -188,7 +188,7 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ]
-	]
+    ]
   
  
     return jobsConfig
