From 09ae9e01ae89c3f281842f377669790da06fb429 Mon Sep 17 00:00:00 2001
From: Swapnali Vakil <svakil@vlsi-int03.kfs.habana-labs.com>
Date: Fri, 24 Jan 2025 05:17:22 +0200
Subject: [PATCH] INTR Regression Update

Change-Id: I9caf1eaada69835412627be1ef9cb1257f20289d
---

diff --git a/h15b/h15b_job_params_svakil.groovy b/h15b/h15b_job_params_svakil.groovy
index 631c7fa..72d4ad3 100644
--- a/h15b/h15b_job_params_svakil.groovy
+++ b/h15b/h15b_job_params_svakil.groovy
@@ -27,7 +27,7 @@
                 'BLOCK_NAME': 'io_die_random_nsf',
                 'PATH_NAME': 'fullchip_verif',
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "+USE_IO_MESH_NSF_AGENT+USE_AXI_FC_TRACKER_AGENT" -v UVM_MEDIUM -lte 10',
-                'MAILING_LIST': 'bgunasekaran@habana.ai ',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai  ',
                 'CSV_FILE': 'fullchip_verif/io_die_random_nsf_verif/tb/regressions/regression_random.csv',
 		        'BUILD_PERIODICALLY': '',
                 'RERUN_FAILED': true,
@@ -36,7 +36,159 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ]
-    ]
+
+       'h15b_d2d_mac_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_D2D_MAC ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D" -v UVM_MEDIUM -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/d2d_mac_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+
+	'h15b_d2d_phy_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_D2D_PHY ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D" -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/d2d_phy_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+
+	'h15b_dtstdvm_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_DTSTDVM ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +DTSTDVM" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +DTSTDVM"  -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/d2d_phy_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+
+        'h15b_muc_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_MUC ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +NIC_NMS +MUC_UMR" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +NIC_NMS +MUC_UMR"  -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/muc_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+       
+  	'h15b_nic_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_NIC ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +NIC_NMS" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +NIC_NMS"  -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/nic_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+
+	'h15b_nms_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_NMS ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +NIC_NMS" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +NIC_NMS"  -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/nms_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+
+	'h15b_noc_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_NOC ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D +CPU" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D +CPU"  -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/noc_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+
+	'h15b_pll_intr_fc_regression': [
+             'jobDescription': 'INTR_FC_IO_DIE_PLL ',
+            'jenkinsView': 'VLSI-H15B',
+            'jobParameters': [
+                'PROJECT_NAME': 'H15B',
+                'BRANCH_NAME': 'h15b_fullchip',
+                'BLOCK_NAME': 'io_die_direct',
+                'PATH_NAME': 'fullchip_verif',
+                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D" -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM" -af "+define+REAL_BOOT" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC +SOC_0_D2D"  -lte 10',
+                'MAILING_LIST': 'bgunasekaran@habana.ai svakil@habana.ai abshaik@habana.ai  ',
+                'CSV_FILE': 'fullchip_verif/io_die_direct_verif/tb/regressions/pll_intr_regressions.csv',
+		        'BUILD_PERIODICALLY': '',
+                'RERUN_FAILED': true,
+                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -verdi',
+                'ADDITIONAL_COVERAGE_PARAMS': '',
+                'RUN_ON_REPO_UPDATE': true
+            ]
+         ]
+	]
   
  
     return jobsConfig
