From bebb1a1c98096694ca4464c7641597adcb51e8f9 Mon Sep 17 00:00:00 2001
From: Swapnali Vakil <svakil@vlsi-int03.kfs.habana-labs.com>
Date: Fri, 24 Jan 2025 05:36:33 +0200
Subject: [PATCH] INTR Regression list Changes

Change-Id: Ibf4e0f38cdca070923c0243a65e051a4024fe31a
---

diff --git a/h15b/h15b_job_params_svakil.groovy b/h15b/h15b_job_params_svakil.groovy
index 5c61381..206e655 100644
--- a/h15b/h15b_job_params_svakil.groovy
+++ b/h15b/h15b_job_params_svakil.groovy
@@ -36,7 +36,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
-
        'h15b_d2d_mac_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_D2D_MAC ',
             'jenkinsView': 'VLSI-H15B',
@@ -55,7 +54,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
-
 	'h15b_d2d_phy_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_D2D_PHY ',
             'jenkinsView': 'VLSI-H15B',
@@ -74,7 +72,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
-
 	'h15b_dtstdvm_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_DTSTDVM ',
             'jenkinsView': 'VLSI-H15B',
@@ -93,7 +90,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
-
         'h15b_muc_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_MUC ',
             'jenkinsView': 'VLSI-H15B',
@@ -111,8 +107,7 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-         ],
-       
+         ],    
   	'h15b_nic_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_NIC ',
             'jenkinsView': 'VLSI-H15B',
@@ -131,7 +126,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
-
 	'h15b_nms_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_NMS ',
             'jenkinsView': 'VLSI-H15B',
@@ -150,7 +144,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
-
 	'h15b_noc_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_NOC ',
             'jenkinsView': 'VLSI-H15B',
@@ -169,7 +162,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ],
- 
 	'h15b_pll_intr_fc_regression': [
              'jobDescription': 'INTR_FC_IO_DIE_PLL ',
            'jenkinsView': 'VLSI-H15B',
@@ -188,8 +180,6 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
          ]
-    ]
-  
- 
+ ]   
     return jobsConfig
 }     
