From a1b427957c7a9a12d2423578118fb1c7f6479856 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Wed, 29 Jan 2025 17:57:00 +0200
Subject: [PATCH] [FSW-13558] fs1: specs h15 p1p36 update

Change-Id: I1b16958099118e656854274097a6576857fc4593
---

diff --git a/fs/fs1/fs1_reg_map.h b/fs/fs1/fs1_reg_map.h
index d3b9422..646472a 100644
--- a/fs/fs1/fs1_reg_map.h
+++ b/fs/fs1/fs1_reg_map.h
@@ -71,7 +71,7 @@
 
 #define IO0_H0_PSOC_CFG_MEM_COPY(reg)	(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE + (reg))
 #define IO0_H0_PSOC_CFG_MEM_COPY_LBW_TRAN_CHK(reg)	\
-		(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_MSTR_IF_LBW_TRAN_CHK_BASE + (reg))
+		(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_MSTR_IF_LBW_TSRL_T_BASE + (reg))
 
 #define mmCPU_BOOT_DEV_STS0		IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_DEV_STS0)
 #define mmCPU_BOOT_DEV_STS1		IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_DEV_STS1)
