From 2dd2bef1ca672be075ff0a436572c9d203d8ee37 Mon Sep 17 00:00:00 2001
From: vaggarwal <vaggarwal@habana.ai>
Date: Fri, 07 Feb 2025 06:38:42 +0200
Subject: [PATCH] updated last commit for pcie

Change-Id: I76e12433401fdb5aee0179c54f9700d4fd91d076
---

diff --git a/h15b/h15b_job_params_vaggarwal.groovy b/h15b/h15b_job_params_vaggarwal.groovy
index ce31c1f..1b32aaf 100644
--- a/h15b/h15b_job_params_vaggarwal.groovy
+++ b/h15b/h15b_job_params_vaggarwal.groovy
@@ -18,7 +18,7 @@
                 'RUN_ON_REPO_UPDATE': true
             ]
         ],
-         'h15b_fullchip_pcie_w_pif_regression': [
+         'h15b_fullchip_pcie_hbw_psoc_regression': [
             'jobDescription': 'Regression environment for IO DIE PCIE HBW datapath',
             'jenkinsView': 'VLSI-H15B',
             'jobParameters': [
