From a722c1b2ed16af9f1d2452fba96670f5ea49232c Mon Sep 17 00:00:00 2001
From: vaggarwal <vaggarwal@habana.ai>
Date: Fri, 07 Feb 2025 06:55:04 +0200
Subject: [PATCH] removed changes

Change-Id: Ib2f74429154042a2c88f567e9f48cb0ac608094e
---

diff --git a/h15b/h15b_job_params_vaggarwal.groovy b/h15b/h15b_job_params_vaggarwal.groovy
index 1b32aaf..8b31047 100644
--- a/h15b/h15b_job_params_vaggarwal.groovy
+++ b/h15b/h15b_job_params_vaggarwal.groovy
@@ -17,24 +17,6 @@
                 'ADDITIONAL_COVERAGE_PARAMS': '',
                 'RUN_ON_REPO_UPDATE': true
             ]
-        ],
-         'h15b_fullchip_pcie_hbw_psoc_regression': [
-            'jobDescription': 'Regression environment for IO DIE PCIE HBW datapath',
-            'jenkinsView': 'VLSI-H15B',
-            'jobParameters': [
-                'PROJECT_NAME': 'H15B',
-                'BRANCH_NAME': 'h15b_fullchip',
-                'BLOCK_NAME': 'io_die_direct',
-                'PATH_NAME': 'pcie_verif',
-                'ADDITIONAL_RUN_PARAMS': '-args_defs  "USE_PCIE_AGENT +PCIE_GEN6_CUSTOM_MODE +USE_PSOC_AGENT +PSOC +USE_IO_AXI_AGENT"  -stub_defs "USE_PCIE_AGENT +PSOC_AGENT +PSOC"  -af "+define+SKIP_MAIN_PHASE +define+SKIP_PCIE_RST +define+SKIP_PIF_BOOT" -use_pcie_ts_vip -lte 5',
-                'MAILING_LIST': 'vaggarwal@@habana.ai akagrawalx@habana.ai',
-                'CSV_FILE': 'pcie_verif/tb/regressions/fc_pcie_data_path_regression.csv',
-                'BUILD_PERIODICALLY': 'H 02 * * 1,3,5',
-                'RERUN_FAILED': true,
-                'RERUN_FAILED_PARAMS': '-v UVM_HIGH -vpd -lte 8',
-                'ADDITIONAL_COVERAGE_PARAMS': '',
-                'RUN_ON_REPO_UPDATE': true
-            ]
         ]
     ]
     return jobsConfig
