From 4a7b53b339eb7bca4c179e616a1003a77fd3e57e Mon Sep 17 00:00:00 2001
From: gshaked <gshaked@habana.ai>
Date: Mon, 17 Feb 2025 10:58:31 +0200
Subject: [PATCH] trigger seed

Change-Id: I420588be6206cd37057dfca61548cc0954c31b64
---

diff --git a/h15b/h15b_job_params_rkogan.groovy b/h15b/h15b_job_params_rkogan.groovy
index 86badf6..0140b94 100644
--- a/h15b/h15b_job_params_rkogan.groovy
+++ b/h15b/h15b_job_params_rkogan.groovy
@@ -3,3 +3,5 @@
     
     return jobsConfig
 }
+
+
