From e2982e7673c0f63606d262beab64e61bb3c16d63 Mon Sep 17 00:00:00 2001
From: gshaked <gshaked@habana.ai>
Date: Mon, 17 Feb 2025 10:51:34 +0200
Subject: [PATCH] trigger seed

Change-Id: Ia2a4c89027bea017e7933bd176f5081315fede30
---

diff --git a/dft_h9/dft_h9_job_params_aevgy.groovy b/dft_h9/dft_h9_job_params_aevgy.groovy
index 00b7d2f..389ae03 100644
--- a/dft_h9/dft_h9_job_params_aevgy.groovy
+++ b/dft_h9/dft_h9_job_params_aevgy.groovy
@@ -2360,3 +2360,4 @@
 
     return jobsConfig 
 }
+
