From be8cbff5fb26e25b1b482ac4984234b26a48e5b1 Mon Sep 17 00:00:00 2001
From: gshaked <gshaked@habana.ai>
Date: Mon, 17 Feb 2025 10:46:50 +0200
Subject: [PATCH] trigger seed

Change-Id: Ia4fa124e3d4f7e224751a7cf034a6a457aa40bc5
---

diff --git a/h6/h6_job_params_abekerman.groovy b/h6/h6_job_params_abekerman.groovy
index 4ad55a7..8e4daee 100644
--- a/h6/h6_job_params_abekerman.groovy
+++ b/h6/h6_job_params_abekerman.groovy
@@ -62,3 +62,4 @@
     return jobsConfig
 }
 
+
