From 04623e95810fb16878af13cb855271032443e57a Mon Sep 17 00:00:00 2001
From: gshaked <gshaked@habana.ai>
Date: Mon, 17 Feb 2025 10:55:37 +0200
Subject: [PATCH] trigger seed

Change-Id: Ic3afe1ad22998868f969686272f79ace766b27ac
---

diff --git a/dftfc/dftfc_job_params_lhajbi.groovy b/dftfc/dftfc_job_params_lhajbi.groovy
index 034f79c..3130947 100644
--- a/dftfc/dftfc_job_params_lhajbi.groovy
+++ b/dftfc/dftfc_job_params_lhajbi.groovy
@@ -145,4 +145,4 @@
         ]
     ]      
     return jobsConfig
-}
\ No newline at end of file
+}
