From fff2efcd01ca7cd49e4fb1e711e277596d8f8d92 Mon Sep 17 00:00:00 2001
From: Mateusz Anuszkiewicz <manuszkiewicz@habana.ai>
Date: Fri, 07 Mar 2025 13:43:51 +0200
Subject: [PATCH] [SW-0] Updating TC repo path

Change-Id: I91024d364c262558275e2ca4f39f08134459a89b
Signed-off-by: Mateusz Anuszkiewicz <manuszkiewicz@habana.ai>
---

diff --git a/templates/base/tencentos/scripts/common.sh b/templates/base/tencentos/scripts/common.sh
index adcfb2c..3262ca2 100644
--- a/templates/base/tencentos/scripts/common.sh
+++ b/templates/base/tencentos/scripts/common.sh
@@ -8,12 +8,12 @@
 cat << EOF >> /etc/yum.repos.d/TencentosMirror.repo
 [tencentos-mirror-srpms]
 name="Habana TC01 repo"
-baseurl=${ARTIFACTORY_URL}/tencentos-mirror/SRPMS/
+baseurl=${ARTIFACTORY_URL}/tencentos-mirror/3.1/TencentOS/SRPMS/
 enabled=1
 gpgcheck=0
 [tencentos-mirror-x86_64]
 name="Habana TC01 repo"
-baseurl=${ARTIFACTORY_URL}/tencentos-mirror/x86_64/
+baseurl=${ARTIFACTORY_URL}/tencentos-mirror/3.1/TencentOS/x86_64/
 enabled=1
 gpgcheck=0
 EOF
