From 9881bd69c475187b3ceb5963309a639a014fc98c Mon Sep 17 00:00:00 2001
From: Mateusz Anuszkiewicz <manuszkiewicz@habana.ai>
Date: Fri, 07 Mar 2025 13:57:49 +0200
Subject: [PATCH] [SW-221644] Updating Tencentos repos on vagrant

Change-Id: Id660018a97f5ed74417a02a70b785bc990da0ed2
Signed-off-by: Mateusz Anuszkiewicz <manuszkiewicz@habana.ai>
---

diff --git a/templates/base/tencentos/scripts/common.sh b/templates/base/tencentos/scripts/common.sh
index 3262ca2..c1b98bd 100644
--- a/templates/base/tencentos/scripts/common.sh
+++ b/templates/base/tencentos/scripts/common.sh
@@ -16,10 +16,23 @@
 baseurl=${ARTIFACTORY_URL}/tencentos-mirror/3.1/TencentOS/x86_64/
 enabled=1
 gpgcheck=0
+name="Habana Updates repo"
+baseurl=${ARTIFACTORY_URL}/artifactory/tencentos-mirror/3.1/Updates/x86_64/"
+enabled=1
+gpgcheck=0
+[TencentOS-AppStream]
+name="Habana TencentOS-AppStream repo"
+baseurl=${ARTIFACTORY_URL}/artifactory/tencentos-mirror/3.1/TencentOS-AppStream/x86_64/"
+enabled=1
+gpgcheck=0
+[EPEL8]
+name="Habana epel8 repo"
+baseurl=${ARTIFACTORY_URL}/artifactory/epel8-mirror"
+enabled=1
+gpgcheck=0
 EOF
-$pkg_cmd install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
 $pkg_cmd install -y wget nfs-utils lsof vim git pciutils redhat-lsb-core cloud-utils-growpart gdisk git-lfs
-$pkg_cmd --enablerepo=epel install -y \
+$pkg_cmd install -y \
     python3 \
     python3-libs \
     python3-devel \
