From 9547fe123c772aa6180416241c6447d46e73d8a2 Mon Sep 17 00:00:00 2001
From: Yoni Leibovitch <yleibovitch@habana.ai>
Date: Thu, 20 Mar 2025 10:54:05 +0200
Subject: [PATCH] disable FS1 NMS_CLS regressions

Change-Id: I614bf606b7965b85baf9f9f8871a5bf4da416ce8
---

diff --git a/h15b/h15b_job_params_yleibovitch.groovy b/h15b/h15b_job_params_yleibovitch.groovy
index e14bd14..7a2fcdb 100644
--- a/h15b/h15b_job_params_yleibovitch.groovy
+++ b/h15b/h15b_job_params_yleibovitch.groovy
@@ -11,7 +11,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-af " +define+TEMP_NMS_CLS_RUN_EXCLUDE_FLAGS" -args_def " NMS_CLS_ARCS_FARM_0_1_STUBS_SYNC_MNGR_AGENT_EDUP_AGENT" -lte 5',
                 'MAILING_LIST': 'yleibovitch@habana.ai mgerstenhaber@habana.ai enaim@habana.ai igeorgy@habana.ai',
                 'CSV_FILE': 'nms_verif/nms_cls_verif/tb/tests/nms_cls_arcs_agents_sms_agents_regression.csv',
-		        'BUILD_PERIODICALLY': 'H H * * 1,5',
+		        'BUILD_PERIODICALLY': '',
                 'RERUN_FAILED': true,
                 'RERUN_FAILED_PARAMS': ' -v UVM_HIGH -verdi',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
