From c3b943ed8509c0477ee47f62824d1e5ebfa00a6c Mon Sep 17 00:00:00 2001
From: kthuppali <kthuppali@habana.ai>
Date: Thu, 03 Apr 2025 08:00:28 +0300
Subject: [PATCH] lbw_regress_updates

Change-Id: I4499fd3a3e8ba5f4c8bbc093d3022064a0693126
---

diff --git a/h15a/h15a_job_params_kthuppali.groovy b/h15a/h15a_job_params_kthuppali.groovy
index 9bce5d9..441842e 100644
--- a/h15a/h15a_job_params_kthuppali.groovy
+++ b/h15a/h15a_job_params_kthuppali.groovy
@@ -11,7 +11,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_VD2D_MAC_0 -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_VD2D_MAC_0_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_vd2d_mac.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -29,7 +29,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_VD2D_PHY_0 -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_VD2D_PHY_0_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_vd2d_phy.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -47,25 +47,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_HD2D_MAC_0 -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_HD2D_MAC_0_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_hd2d_mac.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
-                'RERUN_FAILED': false,
-                'RERUN_FAILED_PARAMS': '',
-                'ADDITIONAL_COVERAGE_PARAMS': '',
-                'RUN_ON_REPO_UPDATE': false
-            ]
-        ],
-        'h15a_comp_die_direct_lbw_tests_HBM0_CMN01_regression': [
-            'jobDescription': 'Regression env for comp_die_direct_lbw HBM0_CMN01 test',
-            'jenkinsView': 'VLSI-H15A',
-            'jobParameters': [
-                'PROJECT_NAME': 'h15a',
-                'BRANCH_NAME': 'h15a_fullchip',
-                'BLOCK_NAME': 'compute_die_direct',
-                'PATH_NAME': 'fullchip_verif',
-                'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_HBM0_CMN01 -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_HBM0_CMN01_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
-                'MAILING_LIST': 'krithi.thuppali@intel.com',
-                'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_HBM0_CMN01.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -83,7 +65,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_HBM0_HCON0_HCON_0_CMN -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_HBM0_HCON0_HCON_0_CMN_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_HBM0_HCON0_HCON_0_CMN.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -101,7 +83,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_NOC_EZ_HBM0 -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_NOC_EZ_HBM0_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_NOC_EZ_HBM0.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -119,7 +101,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_NOC_NRTR_0 -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_NOC_NRTR_0_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_NOC_NRTR_0.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -137,7 +119,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_PSOC_to_GPIO -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_PSOC_to_GPIO_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_PSOC_to_GPIO.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -155,7 +137,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_EDMA -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_EDMA_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_EDMA.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
@@ -173,7 +155,7 @@
                 'ADDITIONAL_RUN_PARAMS': '-args_defs "USE_COMPUTE_PSOC_AGENT +REAL_PLL +PSOC +COMPUTE_DIE0 +USE_HBM_FORCES" -cl -rmem_256g -af_rmem_64g  -xterm -arctest compute_psoc_mng_w_rtt_build:tests_psoc/psoc_arcs_lbw_access_PDMA -arc_user_def "BOOT_ARC_FROM_SECURE_BOOT_ROM USE_PDMA_IP_AGENT" -af "+define+REAL_BOOT+USE_HBM_FORCES" -stub_defs "PSOC_AGENT +REAL_PLL +PSOC" -rf "+TRACE_CLK_EN"  -v UVM_HIGH -zip_log',
                 'MAILING_LIST': 'krithi.thuppali@intel.com',
                 'CSV_FILE': 'fullchip_verif/compute_die_direct_verif/tb/tests/compute_die_direct_lbw_separate_tests_regression_PDMA.csv',
-                'BUILD_PERIODICALLY': 'H H(0-23) * * *',
+                'BUILD_PERIODICALLY': 'H H(0-23) * * 5',
                 'RERUN_FAILED': false,
                 'RERUN_FAILED_PARAMS': '',
                 'ADDITIONAL_COVERAGE_PARAMS': '',
