From 17a99cae1478647793b438feec611363c93ebe98 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Mon, 14 Apr 2025 13:36:34 +0300
Subject: [PATCH] [FSW-11993] fs1: added board and dtsi for pm arc

Change-Id: I9d9192dc228b20fb3f0b84f953c9d1964778c0e5
---

diff --git a/fs/fs1/fs1_mem_addr_dt.h b/fs/fs1/fs1_mem_addr_dt.h
index 0e4230e..49d9e4d 100644
--- a/fs/fs1/fs1_mem_addr_dt.h
+++ b/fs/fs1/fs1_mem_addr_dt.h
@@ -55,6 +55,11 @@
 #define ARC_RST_DCCM_OFFSET	80000000
 #define ARC_RST_DCCM_SIZE	0x20
 
+#define ARC_PMFW_ICCM_OFFSET	0
+#define ARC_PMFW_ICCM_SIZE	0x40
+#define ARC_PMFW_DCCM_OFFSET	80000000
+#define ARC_PMFW_DCCM_SIZE	0x20
+
 #define ARC_RESET_ICCM_SECTION	1
 #define ARC_RESET_DCCM_SECTION	1
 #define ARC_RESET_ICCM_OFFSET	0
