From 442f1118a5e841839193dbb7c4718f09adff7271 Mon Sep 17 00:00:00 2001
From: Alexey Nikandrov <anikandrov@habana.ai>
Date: Wed, 07 May 2025 11:57:08 +0300
Subject: [PATCH] [SW-228173] package: hl-thunk add python test's scripts

Change-Id: I23055e15f701a366373a6df2686d43f2770f6f93
Signed-off-by: Alexey Nikandrov <anikandrov@habana.ai>
---

diff --git a/packages-specs/hlthunk/Makefile b/packages-specs/hlthunk/Makefile
index 8490623..feca272 100644
--- a/packages-specs/hlthunk/Makefile
+++ b/packages-specs/hlthunk/Makefile
@@ -59,6 +59,8 @@
 
 pre-build: pre-package
 	$(HIDE)echo "Pre-build"
+	$(HIDE)cd $(WORKSPACE)/hl-thunk && mkdir -p scripts-tmp && cp -r scripts/test_runner scripts-tmp/ && cp scripts/requirements.txt scripts-tmp/
 	$(HIDE)cd $(WORKSPACE)/hl-thunk && find . -type f -name '.gitignore' -exec rm -f {} \; && rm -rf tests/python scripts .ci .cd .git .gitreview
+	$(HIDE)cd $(WORKSPACE)/hl-thunk && mv scripts-tmp scripts
 
 build: init clean deps pre-build build-package
\ No newline at end of file
