From a3213dcf4d21ce28e98d9038bcf7081ad551bd9a Mon Sep 17 00:00:00 2001
From: Kamil Felskowski <kfelskowskix@habana.ai>
Date: Thu, 08 May 2025 13:10:31 +0300
Subject: [PATCH] [SW-221766] Add support for TC4 in CD tests

Change-Id: I8b14a73929a3bff4c29374a4e4e8a4b1846e2caf
---

diff --git a/tests/asic/PT-Suite/pt-run-docker-container.sh b/tests/asic/PT-Suite/pt-run-docker-container.sh
index 9418c6a..40433fe 100755
--- a/tests/asic/PT-Suite/pt-run-docker-container.sh
+++ b/tests/asic/PT-Suite/pt-run-docker-container.sh
@@ -37,6 +37,9 @@
     tencentos3.1)
         JOB_SCENARIO="pt_cd_tencentos31"
     ;;
+    tencentos4.2)
+        JOB_SCENARIO="pt_cd_tencentos42"
+    ;;
     suse15.5)
         JOB_SCENARIO="pt_cd_suse15"
     ;;
