From 63961e8bff451c4ffdbeb54ee3b1bf746ef5ce31 Mon Sep 17 00:00:00 2001
From: Witold Netel <wnetel@habana.ai>
Date: Wed, 14 May 2025 21:28:38 +0300
Subject: [PATCH] [SW-228823] Add --nobest flag to dnf update in rhel9.4 py312

Change-Id: I87c595413aa2d477b65f31e830fb8352508e9b49
Signed-off-by: Witold Netel <wnetel@habana.ai>
---

diff --git a/dockerfiles/pytorch/Dockerfile.rhel9.4-py312 b/dockerfiles/pytorch/Dockerfile.rhel9.4-py312
index 15e1754..fddea0e 100644
--- a/dockerfiles/pytorch/Dockerfile.rhel9.4-py312
+++ b/dockerfiles/pytorch/Dockerfile.rhel9.4-py312
@@ -31,7 +31,7 @@
     echo "baseurl=https://${ARTIFACTORY_URL}/artifactory/epel9-mirror" >> /etc/yum.repos.d/habana.repo && \
     echo 'gpgcheck=0'  >> /etc/yum.repos.d/habana.repo
 
-RUN dnf update -y && dnf install --nobest --nodocs --setopt=install_weak_deps=false --allowerasing -y \
+RUN dnf --nobest update -y && dnf install --nobest --nodocs --setopt=install_weak_deps=false --allowerasing -y \
         cairo-devel \
         gperftools-devel \
         iproute \
