From 3a75b9e57f3e406f0f4dc16850cad1beb7deed97 Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Wed, 21 May 2025 10:07:23 +0300
Subject: [PATCH] [SW-0] fwi: vi typo

typo mistake

Change-Id: I031d88a3819cf9493c88e9330829405f8e4cb968
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/scripts/fwi/hls_platform.rc b/scripts/fwi/hls_platform.rc
index 5e41848..cbad2e4 100644
--- a/scripts/fwi/hls_platform.rc
+++ b/scripts/fwi/hls_platform.rc
@@ -119,7 +119,7 @@
 	local vals=()
 
 	[ ${use_i2c_f:-0} -eq 1 ] && {
-		vals=($(bmc_i2c_send i${oam_id} 0x25 0x00))
+		vals=($(bmc_i2c_send ${oam_id} 0x25 0x00))
 	} || {
 		vals=($(bmc_raw_cmd 0xb ${oam_id}))
 		[ $? -eq 0 ] && {
