From 6f4f211a09c7695c9a2fc209ba72fbdd5560885d Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Sun, 25 May 2025 15:21:21 +0300
Subject: [PATCH] [JGS-TBD] jgs: add rma header to build jgs

Change-Id: I1644c6f0d49ad083013dd4cb17839f8e8cf6744d
---

diff --git a/fs/jgs/rma_if_dev.h b/fs/jgs/rma_if_dev.h
new file mode 100644
index 0000000..9d10065
--- /dev/null
+++ b/fs/jgs/rma_if_dev.h
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2024 Intel Ltd.
+ * All Rights Reserved.
+ */
+
+#ifndef _JGS_RMA_IF_DEV_H_
+#define _JGS_RMA_IF_DEV_H_
+
+/* TODO: set proper values for JGS */
+#define MAX_RMA_HEADER_ENTRIES			1
+#define MAX_SRAM_ECC_ENTRIES			1
+#define MAX_HBM_ECC_ENTRIES			1
+#define MAX_PARITY_ENTRIES			1
+#define MAX_ROW_REPLACE_ENTRIES			1
+#define MAX_PAGE_DISCARDS_ENTRIES		1
+#define MAX_HBM_LANE_REPAIR_ENTRIES		1
+#define MAX_HBM_SPARE_ROWS_ENTRIES		1
+
+#endif /* _JGS_RMA_IF_DEV_H_ */
