From c60c4ef5436caf2406347c805be5fd3f48706244 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 10 Jun 2025 09:22:59 +0300
Subject: [PATCH] arc specific changes

Change-Id: Ic7c85e8d97c01d8999a03a6af05b6e5a4c363d6f
---

diff --git a/fs/jgs/soc.h b/fs/jgs/soc.h
index 6348fc0..c453b8f 100644
--- a/fs/jgs/soc.h
+++ b/fs/jgs/soc.h
@@ -54,6 +54,8 @@
 #define CONV_ARC_REG_TO_GEN_MEM(addr)	(((addr) & MASK_28_BIT) + ((GEN_MEM_FULL_BASE) & MASK_32_BIT))
 #define CONV_ARC_REG_TO_CFG_MEM(addr)	(((addr) & MASK_28_BIT) + ((CFG_MEM_FULL_BASE) & MASK_32_BIT))
 
+#define ARC_REGION_UPD(REG_O, ARC_REGION)	((REG_O & ~(ARC_REGION_MASK)) | ARC_REGION) /* TODO: REGION Update should be based on 4mg group */
+
 #define FLASH_VALID_MASK		(SZ_2M - 1)
 #define SPI_FLASH_OFFSET		0xF000000
 #define SPI_FLASH_SIZE			(16 * SZ_1MB)
