From 020f50c3de554b166537983de8a00a3932450ebc Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 10 Jun 2025 09:34:36 +0300
Subject: [PATCH] make arch specific changes

Change-Id: If2304b5dd637c0355ffc8230eea3626d11ab6793
---

diff --git a/include/mem_io.h b/include/mem_io.h
index 84a812f..78bc04d 100644
--- a/include/mem_io.h
+++ b/include/mem_io.h
@@ -10,9 +10,6 @@
 #include <soc.h>
 #include <spi.h>
 
-#define ARC_REGION_UPD(REG_O, ARC_REGION)	((REG_O & ~(ARC_REGION_MASK)) | ARC_REGION) /* TODO: REGION Update should be based on 4mg group */
-
-
 /* In ARC address manipulation BIT[22:24] are for die id */
 #define HL_ARC_ADDR_DIE_ID_SHIFT		25
 #define HL_ARC_ADDR_DIE_ID_MASK			(0x7 << HL_ARC_ADDR_DIE_ID_SHIFT)
