From 211d615378a23ac1e3eb78f375ed0a27895e7297 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 10 Jun 2025 08:43:40 +0300
Subject: [PATCH] update arc region

Change-Id: Ic81a7e9f47912f26f43faa26a459428fd3f90d9d
---

diff --git a/Makefile b/Makefile
index e24b025..c6a1c72 100644
--- a/Makefile
+++ b/Makefile
@@ -92,7 +92,7 @@
 	@date +'#define PPBOOT_TIME "%T"' >> $@
 
 buildversion-habanalabs: include/baremetal.h $(AUTO_GENERATED_FILE)
-	echo "VERSION_STR" | gcc -include $< -I./ -I ./lib/common/ -I ./$(HL_DEV_ARCH)/ -I$(SPECS_EMBEDDED_DIR)/$(SPECS_EMB_DEV) -I$(SPECS_DIR)/$(BASE_DEV_NAME)/asic_reg -I$(SPECS_DIR)/ -I ./include/ -P -E -o - - | tail -1 | sed 's/\" \+\"//g;s/["\n]//g' | cut -d' ' -f2- | tr -d '\n' > $@
+	echo "VERSION_STR" | gcc -include $< -I./ -I ./lib/common/ -I ./$(HL_DEV_ARCH)/ -I$(SPECS_EMBEDDED_DIR)/fs/$(DEV_PREFIX) -I$(SPECS_DIR)/$(BASE_DEV_NAME)/asic_reg -I$(SPECS_DIR)/ -I ./include/ -P -E -o - - | tail -1 | sed 's/\" \+\"//g;s/["\n]//g' | cut -d' ' -f2- | tr -d '\n' > $@
 
 $(PCI_PHY_FW_TARGET): $(HL_DEV_ARCH)/pcie_phy_fw.i
 	@cut -d' ' -f2- $< | xxd -r -p | dd status=none bs=16 count=$$(($(PHY_FW_SIZE) / 16)) conv=sync,swab of=$@
diff --git a/include/baremetal.h b/include/baremetal.h
index 41a43be..a2034c0 100644
--- a/include/baremetal.h
+++ b/include/baremetal.h
@@ -10,6 +10,7 @@
 
 #include <common.h>
 #include <sizes.h>
+#include <soc.h>
 
 #define ARRAY_SIZE(x)		(sizeof(x) / sizeof((x)[0]))
 #define ROUND_UP4(n)		(((n) + sizeof(u32) - 1) & ~(sizeof(u32) - 1))
@@ -117,26 +118,26 @@
 
 static inline u8 regmap_readb(u32 offset)
 {
-	u32 reg = *((volatile u32 *)(HL_CFG_BASE + offset));
+	u32 reg = *((volatile u32 *)(ARC_REGION_UPD(CFG_SPACE_REG_BASE, offset)));
 
 	return (u8)(reg & 0xFF);
 }
 
 static inline void regmap_writeb(u8 val, u32 offset)
 {
-	*((volatile u32 *)(HL_CFG_BASE + offset)) = (u32)val;
+	*((volatile u32 *)(ARC_REGION_UPD(CFG_SPACE_REG_BASE, offset))) = (u32)val;
 }
 
 static inline u32 regmap_readl(u32 offset)
 {
-	u32 reg = *((volatile u32 *)(HL_CFG_BASE + offset));
+	u32 reg = *((volatile u32 *)(ARC_REGION_UPD(CFG_SPACE_REG_BASE, offset)));
 
 	return reg;
 }
 
 static inline void regmap_writel(u32 val, u32 offset)
 {
-	*((volatile u32 *)(HL_CFG_BASE + offset)) = val;
+	*((volatile u32 *)(ARC_REGION_UPD(CFG_SPACE_REG_BASE, offset))) = val;
 }
 
 static inline void regmap_rmwritel(u32 addr, u32 data, u32 mask)
