From 602dc417af83978b39511666482ef1c5c44d479f Mon Sep 17 00:00:00 2001
From: Rajesh Sanga <rsanga@habana.ai>
Date: Tue, 08 Jul 2025 14:57:47 +0300
Subject: [PATCH] [SW-231866] Enable the setuptools for the latest

Change-Id: I70d668491a19f16272c187ad77799c242e17f9f1
Signed-off-by: Rajesh Sanga <rsanga@habana.ai>
---

diff --git a/dockerfiles/base/Dockerfile.ubuntu22.04 b/dockerfiles/base/Dockerfile.ubuntu22.04
index 899eeb0..2ed3248 100644
--- a/dockerfiles/base/Dockerfile.ubuntu22.04
+++ b/dockerfiles/base/Dockerfile.ubuntu22.04
@@ -59,7 +59,7 @@
 RUN wget https://bootstrap.pypa.io/get-pip.py && \
     python3 get-pip.py && \
     rm -f get-pip.py && \
-    python3 -m pip install setuptools==76.1.0 wheel && \
+    python3 -m pip install setuptools==80.9.0 wheel && \
     python3 -m pip install --upgrade Jinja2 protobuf
 
 COPY install_efa.sh .
