From 550f170e9e2e7a84fe2a2dec377dc702858b8bd4 Mon Sep 17 00:00:00 2001
From: Ziemowit Orlikowski <zorlikow@habana.ai>
Date: Thu, 07 Aug 2025 11:24:52 +0300
Subject: [PATCH] [SW-237281] Delete unsupported vLLM images

Change-Id: I456b702143ed1d4f87e737ad1e0e50b538a71ea5
Signed-off-by: Ziemowit Orlikowski <zorlikow@habana.ai>
---

diff --git a/dockerfiles/vllm/Makefile b/dockerfiles/vllm/Makefile
index 07be88f..117f16c 100644
--- a/dockerfiles/vllm/Makefile
+++ b/dockerfiles/vllm/Makefile
@@ -35,14 +35,10 @@
 			$(HIDE)cp $(DOCKERFILE_PREFIX)/Dockerfile.ubuntu.pytorch.vllm $(BUILD_DIR)/Dockerfile
     else ifneq (,$(findstring rhel,$(BUILD_OS)))
 			$(HIDE)cp $(DOCKERFILE_PREFIX)/Dockerfile.rhel.tenc.pytorch.vllm $(BUILD_DIR)/Dockerfile
-    else ifneq (,$(findstring tencentos,$(BUILD_OS)))
-			$(HIDE)cp $(DOCKERFILE_PREFIX)/Dockerfile.rhel.tenc.pytorch.vllm $(BUILD_DIR)/Dockerfile
     else ifneq (,$(findstring navix,$(BUILD_OS)))
 			$(HIDE)cp $(DOCKERFILE_PREFIX)/Dockerfile.navix.pytorch.vllm $(BUILD_DIR)/Dockerfile
     else ifneq (,$(findstring opencloudos,$(BUILD_OS)))
 			$(HIDE)cp $(DOCKERFILE_PREFIX)/Dockerfile.opencloudos.pytorch.vllm $(BUILD_DIR)/Dockerfile
-    else ifneq (,$(findstring suse,$(BUILD_OS)))
-			$(HIDE)cp $(DOCKERFILE_PREFIX)/Dockerfile.suse.pytorch.vllm $(BUILD_DIR)/Dockerfile
     else
 			$(error Unsupported BUILD_OS: $(BUILD_OS))
     endif
