From adaff6b166591cb95b1f53043f09975a45e20ac5 Mon Sep 17 00:00:00 2001
From: HabDevopsInfra <svc_doinfgithub@habana.ai>
Date: Thu, 21 Aug 2025 16:45:11 +0300
Subject: [PATCH] Allowerasing rhel packages

Change-Id: I619e1304c9fa354464ae42bd5bff5daa14234c2a
---

diff --git a/installer/src/initialize.sh b/installer/src/initialize.sh
index ab89357..208c1a5 100644
--- a/installer/src/initialize.sh
+++ b/installer/src/initialize.sh
@@ -147,6 +147,7 @@
         __pkg__="dnf"
         __pkg_man__="rpm -qa"
         __suffix__=$(rpm --eval='%{dist}' | sed "s/\.centos//")
+        HABANALABS_PKG_INSTALL_OPTION="--allowerasing"
         case $OS_VERSION in
             "8.6")
                 MIN_PYTHON_VER="3.11"
