From 868c59db385d53907bc19c43cc7bdf9969924691 Mon Sep 17 00:00:00 2001
From: sakell <sakell@habana.ai>
Date: Thu, 04 Sep 2025 11:30:02 +0300
Subject: [PATCH] Updated confpy for v1.22.0

Change-Id: Iad0ad259da78ee472c606bc91dae0f6b7e6f8d8e
Signed-off-by: sakell <sakell@habana.ai>
---

diff --git a/source/Gaudi_OOB_Management/conf.py b/source/Gaudi_OOB_Management/conf.py
index 26dd392..4fe7ccd 100644
--- a/source/Gaudi_OOB_Management/conf.py
+++ b/source/Gaudi_OOB_Management/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.2'
+release = '1.22.0'
 
 
 
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.2\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
              BMC OOB Management Protocol for First-gen Gaudi}}
 
@@ -329,7 +329,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.2_BMC_OOB_Management_Protocol_for_First-gen_Gaudi_V0.1.tex', 'Intel Gaudi 1.21.2 First-gen Gaudi and BMC OOB Management Protocol',
+    (master_doc, 'Intel_Gaudi_1.22.0_BMC_OOB_Management_Protocol_for_First-gen_Gaudi_V0.1.tex', 'Intel Gaudi 1.22.0 First-gen Gaudi and BMC OOB Management Protocol',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
