From 12711c2ad5675a73faa6865c0b5df33f94577804 Mon Sep 17 00:00:00 2001
From: sakell <sakell@habana.ai>
Date: Thu, 04 Sep 2025 11:37:55 +0300
Subject: [PATCH] Updated confpy for v1.22.0

Change-Id: Ib8cba6140b2d37fc55b8673892842ff167a13560
Signed-off-by: sakell <sakell@habana.ai>
---

diff --git a/source/Arista_exporter/conf.py b/source/Arista_exporter/conf.py
index 340128f..300dd06 100644
--- a/source/Arista_exporter/conf.py
+++ b/source/Arista_exporter/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.2'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Arista_exporter/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/Arista_exporter/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Arista_exporter/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/Arista_exporter/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/Arista_exporter/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/Arista_exporter/Intel_gaudi_logo_big.png}
                 \includegraphics[width = 150mm]{}
                 \centering
             \end{figure}
diff --git a/source/GAUDI2_ODM_System_Tests/GAUDI2_ODM_System_Tests.rst b/source/GAUDI2_ODM_System_Tests/GAUDI2_ODM_System_Tests.rst
index 998fe6f..deb2f30 100644
--- a/source/GAUDI2_ODM_System_Tests/GAUDI2_ODM_System_Tests.rst
+++ b/source/GAUDI2_ODM_System_Tests/GAUDI2_ODM_System_Tests.rst
@@ -144,11 +144,11 @@
 | Criteria**      |                                                                                  |
 +-----------------+----------------------------------------------------------------------------------+
 | **Reference**   | `System Management Interface Tool User Guide (hl-smi Tool)                       |
-|                 | <https://docs.habana.ai/en/v1.21.4/Management_and_                               |
+|                 | <https://docs.habana.ai/en/v1.22.0/Management_and_                               |
 |                 | Monitoring/Embedded_System_Tools_Guide/System_Management_Interface_Tool.html>`_  |
 |                 |                                                                                  |
 |                 | `Platform Upgrade and Full System Installation                                   |
-|                 | <https://docs.habana.ai/en/v1.21.4/Installation_Guide/Platform_                  |
+|                 | <https://docs.habana.ai/en/v1.22.0/Installation_Guide/Platform_                  |
 |                 | Upgrade_and_Unboxing.html>`_                                                     |
 +-----------------+----------------------------------------------------------------------------------+
 | **Test          | If the test fails, perform the following:                                        |
@@ -419,7 +419,7 @@
 
 The hl_qual tool tests qualify the usage and integration of Gaudi hardware platforms in your server design.
 
-For further information, refer to `Qualification Tool (hl_qual) <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/index.html>`_ documentation.
+For further information, refer to `Qualification Tool (hl_qual) <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/index.html>`_ documentation.
 
 It is recommended to run the tests in the same order as described below.
 
@@ -505,7 +505,7 @@
 PCI Test (Serial Mode)
 ----------------------
 
-Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
+Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -541,7 +541,7 @@
 PCI Test (Parallel Mode)
 ------------------------
 
-Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
+Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -569,7 +569,7 @@
 Memory Bandwidth Test
 ---------------------
 
-Refer to `Memory Bandwidth Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#memory-bandwidth-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Memory Bandwidth Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#memory-bandwidth-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -590,7 +590,7 @@
 NIC E2E BW
 ----------
 
-Refer to `E2E Concurrency Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-e2e-concurrency-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `E2E Concurrency Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-e2e-concurrency-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -615,7 +615,7 @@
 SER Test for Serdes
 -------------------
 
-Refer to `SER Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-ser-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `SER Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-ser-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -662,7 +662,7 @@
 Serdes Base – Pairs Test
 ------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -688,7 +688,7 @@
 Serdes Base – All-Reduce Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -714,7 +714,7 @@
 Serdes Base – All-Gather Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -745,7 +745,7 @@
 Serdes Base – Loopback Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -771,7 +771,7 @@
 Serdes Base – Direct Bandwidth Test
 ------------------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -802,7 +802,7 @@
 Functional Test 2
 -----------------
 
-Refer to `Functional Test 2 Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#functional-test-2-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Functional Test 2 Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#functional-test-2-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -857,7 +857,7 @@
 Power Stress
 ------------
 
-Refer to `Power Stress Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#gaudi2-power-stress-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Power Stress Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#gaudi2-power-stress-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -892,7 +892,7 @@
 EDP
 ---
 
-Refer to `EDP Stress Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#edp-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `EDP Stress Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#edp-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -923,7 +923,7 @@
 ResNet-50 and TrainingApp
 --------------------------
 
-Refer to `ResNet-50 Training Stress Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#resnet-50-training-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `ResNet-50 Training Stress Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#resnet-50-training-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -953,7 +953,7 @@
 PyTorch ResNet-50
 ====================
 
-The test verifies accuracy and performance results of the ResNet-50 training. For more details, see `ResNet50 Keras Model References <https://github.com/HabanaAI/Model-References/tree/v1.21.4/PyTorch/computer_vision/classification/torchvision>`_ GitHub page.
+The test verifies accuracy and performance results of the ResNet-50 training. For more details, see `ResNet50 Keras Model References <https://github.com/HabanaAI/Model-References/tree/v1.22.0/PyTorch/computer_vision/classification/torchvision>`_ GitHub page.
 
 .. tabularcolumns:: |p{2cm}|p{2cm}|p{2cm}|p{2cm}|p{2cm}|p{4cm}|
 
diff --git a/source/GAUDI2_ODM_System_Tests/conf.py b/source/GAUDI2_ODM_System_Tests/conf.py
index 5c1e859..ef40574 100644
--- a/source/GAUDI2_ODM_System_Tests/conf.py
+++ b/source/GAUDI2_ODM_System_Tests/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_ODM_System_Tests/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_ODM_System_Tests/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_ODM_System_Tests/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_ODM_System_Tests/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             ODM System Tests for Gaudi 2}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_ODM_System_Tests/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_ODM_System_Tests/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_ODM_System_Tests_for_Gaudi2_V0.1.tex', 'Intel Gaudi 1.21.4 Gaudi 2 ODM System Tests',
+    (master_doc, 'Intel_Gaudi_1.22.0_ODM_System_Tests_for_Gaudi2_V0.1.tex', 'Intel Gaudi 1.22.0 Gaudi 2 ODM System Tests',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/GAUDI2_OOB_Management/conf.py b/source/GAUDI2_OOB_Management/conf.py
index c3f1253..4d5cd9b 100644
--- a/source/GAUDI2_OOB_Management/conf.py
+++ b/source/GAUDI2_OOB_Management/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             BMC OOB Management Protocol for Gaudi 2}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 159mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 159mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_BMC_OOB_Management_Protocol_for_Gaudi2_V0.1.tex', 'Intel Gaudi 1.21.4 Gaudi 2 and BMC OOB Management Protocol',
+    (master_doc, 'Intel_Gaudi_1.22.0_BMC_OOB_Management_Protocol_for_Gaudi2_V0.1.tex', 'Intel Gaudi 1.22.0 Gaudi 2 and BMC OOB Management Protocol',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/GAUDI2_Reliability_Availability_and_Serviceability/conf.py b/source/GAUDI2_Reliability_Availability_and_Serviceability/conf.py
index dde9bf5..5a23fb7 100644
--- a/source/GAUDI2_Reliability_Availability_and_Serviceability/conf.py
+++ b/source/GAUDI2_Reliability_Availability_and_Serviceability/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -134,7 +134,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -158,7 +158,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -179,7 +179,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -240,7 +240,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             Reliability Availability and Serviceability for Gaudi 2}}
 
@@ -252,7 +252,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -327,7 +327,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_Reliability_Availability_and_Serviceability_Gaudi2.tex', 'Intel Gaudi 1.21.4 Reliability Availability and Serviceability for Gaudi 2',
+    (master_doc, 'Intel_Gaudi_1.22.0_Reliability_Availability_and_Serviceability_Gaudi2.tex', 'Intel Gaudi 1.22.0 Reliability Availability and Serviceability for Gaudi 2',
      'Revision 1.0', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/GAUDI3_ODM_System_Tests/GAUDI3_ODM_System_Tests.rst b/source/GAUDI3_ODM_System_Tests/GAUDI3_ODM_System_Tests.rst
index c91cb79..73a7df6 100644
--- a/source/GAUDI3_ODM_System_Tests/GAUDI3_ODM_System_Tests.rst
+++ b/source/GAUDI3_ODM_System_Tests/GAUDI3_ODM_System_Tests.rst
@@ -145,11 +145,11 @@
 | Criteria**      |                                                                                  |
 +-----------------+----------------------------------------------------------------------------------+
 | **Reference**   | `System Management Interface Tool User Guide (hl-smi Tool)                       |
-|                 | <https://docs.habana.ai/en/v1.21.4/Management_and_                               |
+|                 | <https://docs.habana.ai/en/v1.22.0/Management_and_                               |
 |                 | Monitoring/Embedded_System_Tools_Guide/System_Management_Interface_Tool.html>`_  |
 |                 |                                                                                  |
 |                 | `Platform Upgrade and Full System Installation                                   |
-|                 | <https://docs.habana.ai/en/v1.21.4/Installation_Guide/Platform_                  |
+|                 | <https://docs.habana.ai/en/v1.22.0/Installation_Guide/Platform_                  |
 |                 | Upgrade_and_Unboxing.html>`_                                                     |
 +-----------------+----------------------------------------------------------------------------------+
 | **Test          | If the test fails, perform the following:                                        |
@@ -205,7 +205,7 @@
 
 The hl_qual tool tests qualify the usage and integration of Gaudi hardware platforms in your server design.
 
-For further information, refer to `Qualification Tool (hl_qual) <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/index.html>`_ documentation.
+For further information, refer to `Qualification Tool (hl_qual) <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/index.html>`_ documentation.
 
 It is recommended to run the tests in the same order as described below.
 
@@ -281,7 +281,7 @@
 PCI Test (Serial Mode)
 ----------------------
 
-Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
+Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -317,7 +317,7 @@
 Memory Bandwidth Test
 ---------------------
 
-Refer to `Memory Bandwidth Plugin <https://docs.habana.ai/en/1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#gaudi-memory-bandwith-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Memory Bandwidth Plugin <https://docs.habana.ai/en/1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#gaudi-memory-bandwith-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -341,7 +341,7 @@
 SER Test for Serdes
 -------------------
 
-Refer to `SER Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-ser-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `SER Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-ser-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -384,7 +384,7 @@
 Serdes Base – Pairs Test
 ------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -414,7 +414,7 @@
 Serdes Base – All-Reduce Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -440,7 +440,7 @@
 Serdes Base – All-Gather Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -466,7 +466,7 @@
 Serdes Base – Loopback Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -492,7 +492,7 @@
 Serdes Base – Direct Bandwidth Test
 ------------------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -519,7 +519,7 @@
 Functional Test 2
 -----------------
 
-Refer to `Functional Test 2 Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#functional-test-2-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Functional Test 2 Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#functional-test-2-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -568,7 +568,7 @@
 Power Stress
 ------------
 
-Refer to `Power Stress Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#gaudi2-power-stress-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Power Stress Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#gaudi2-power-stress-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -601,7 +601,7 @@
 EDP
 ---
 
-Refer to `EDP Stress Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#edp-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `EDP Stress Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#edp-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -622,7 +622,7 @@
 ResNet-50 and TrainingApp
 --------------------------
 
-Refer to `ResNet-50 Training Stress Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#resnet-50-training-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `ResNet-50 Training Stress Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#resnet-50-training-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
diff --git a/source/GAUDI3_ODM_System_Tests/conf.py b/source/GAUDI3_ODM_System_Tests/conf.py
index 0b49a70..6fb0a55 100644
--- a/source/GAUDI3_ODM_System_Tests/conf.py
+++ b/source/GAUDI3_ODM_System_Tests/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_ODM_System_Tests/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_ODM_System_Tests/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_ODM_System_Tests/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_ODM_System_Tests/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             ODM System Tests for Gaudi 3 OAM}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_ODM_System_Tests/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_ODM_System_Tests/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_ODM_System_Tests_for_Gaudi3_OAM_V0.1.tex', 'Intel Gaudi 1.21.4 Gaudi 3 OAM ODM System Tests',
+    (master_doc, 'Intel_Gaudi_1.22.0_ODM_System_Tests_for_Gaudi3_OAM_V0.1.tex', 'Intel Gaudi 1.22.0 Gaudi 3 OAM ODM System Tests',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/GAUDI3_OOB_Management/conf.py b/source/GAUDI3_OOB_Management/conf.py
index b5ad616..a8e0bf0 100644
--- a/source/GAUDI3_OOB_Management/conf.py
+++ b/source/GAUDI3_OOB_Management/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_OOB_Management/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_OOB_Management/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_OOB_Management/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_OOB_Management/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             OOB Management Specification for Gaudi 3}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 159mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_OOB_Management/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 159mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_OOB_Management/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_OOB_Management_Specification_for_Gaudi3_V0.1.tex', 'Intel Gaudi 1.21.4 OOB Management Specification for Gaudi 3',
+    (master_doc, 'Intel_Gaudi_1.22.0_OOB_Management_Specification_for_Gaudi3_V0.1.tex', 'Intel Gaudi 1.22.0 OOB Management Specification for Gaudi 3',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/GAUDI3_Reliability_Availability_and_Serviceability/conf.py b/source/GAUDI3_Reliability_Availability_and_Serviceability/conf.py
index da71df3..ad38d77 100644
--- a/source/GAUDI3_Reliability_Availability_and_Serviceability/conf.py
+++ b/source/GAUDI3_Reliability_Availability_and_Serviceability/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -134,7 +134,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -158,7 +158,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -179,7 +179,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -240,7 +240,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             Reliability Availability and Serviceability for Gaudi 3}}
 
@@ -252,7 +252,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI3_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI3_Reliability_Availability_and_Serviceability/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -327,7 +327,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_Reliability_Availability_and_Serviceability_Gaudi3.tex', 'Intel Gaudi 1.21.4 Reliability Availability and Serviceability for Gaudi 3',
+    (master_doc, 'Intel_Gaudi_1.22.0_Reliability_Availability_and_Serviceability_Gaudi3.tex', 'Intel Gaudi 1.22.0 Reliability Availability and Serviceability for Gaudi 3',
      'Revision 1.0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/Gaudi2_RMA_Events/conf.py b/source/Gaudi2_RMA_Events/conf.py
index aa45508..53ed3b0 100644
--- a/source/Gaudi2_RMA_Events/conf.py
+++ b/source/Gaudi2_RMA_Events/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Gaudi2_RMA_Events/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/Gaudi2_RMA_Events/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Gaudi2_RMA_Events/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/Gaudi2_RMA_Events/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/Gaudi2_RMA_Events/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/Gaudi2_RMA_Events/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_RMA_Events_Guidelines_Gaudi2.tex', 'Intel Gaudi 1.21.4 RMA Events Guidelines for Gaudi 2',
+    (master_doc, 'Intel_Gaudi_1.22.0_RMA_Events_Guidelines_Gaudi2.tex', 'Intel Gaudi 1.22.0 RMA Events Guidelines for Gaudi 2',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/Gaudi3_Inband_Telemetry_for_Hypervisor/conf.py b/source/Gaudi3_Inband_Telemetry_for_Hypervisor/conf.py
index 1fdfe22..41c609e 100644
--- a/source/Gaudi3_Inband_Telemetry_for_Hypervisor/conf.py
+++ b/source/Gaudi3_Inband_Telemetry_for_Hypervisor/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -141,7 +141,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Gaudi3_Inband_Telemetry_for_Hypervisor/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/Gaudi3_Inband_Telemetry_for_Hypervisor/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -165,7 +165,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Gaudi3_Inband_Telemetry_for_Hypervisor/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/Gaudi3_Inband_Telemetry_for_Hypervisor/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -186,7 +186,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -247,7 +247,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             Gaudi 3 In-Band Telemetry for Hypervisor}}
 
@@ -259,7 +259,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 159mm]{/home/auvarova/readthedocs/NDA_Docs/source/Gaudi3_Inband_Telemetry_for_Hypervisor/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 159mm]{/home/sabuakel/readthedocs/NDA_Docs/source/Gaudi3_Inband_Telemetry_for_Hypervisor/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -334,7 +334,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_In-Band_Telemetry_for_Hypervisor_for_Gaudi3.tex', 'Intel Gaudi 1.21.4 In-Band Telemetry for Hypervisor for Gaudi 3',
+    (master_doc, 'Intel_Gaudi_1.22.0_In-Band_Telemetry_for_Hypervisor_for_Gaudi3.tex', 'Intel Gaudi 1.22.0 In-Band Telemetry for Hypervisor for Gaudi 3',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/Gaudi3_Redfish_Data_Model/conf.py b/source/Gaudi3_Redfish_Data_Model/conf.py
index fd27d70..89402ad 100644
--- a/source/Gaudi3_Redfish_Data_Model/conf.py
+++ b/source/Gaudi3_Redfish_Data_Model/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             Redfish Data Model for Gaudi 3}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 159mm]{/home/auvarova/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 159mm]{/home/sabuakel/readthedocs/NDA_Docs/source/GAUDI2_OOB_Management/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_Redfish_Data_Model_for_Gaudi3.tex', 'Intel Gaudi 1.21.4 Redfish Data Model for Gaudi 3',
+    (master_doc, 'Intel_Gaudi_1.22.0_Redfish_Data_Model_for_Gaudi3.tex', 'Intel Gaudi 1.22.0 Redfish Data Model for Gaudi 3',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/Gaudi_OOB_Management/conf.py b/source/Gaudi_OOB_Management/conf.py
index 26dd392..4fe7ccd 100644
--- a/source/Gaudi_OOB_Management/conf.py
+++ b/source/Gaudi_OOB_Management/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.2'
+release = '1.22.0'
 
 
 
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.2\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
              BMC OOB Management Protocol for First-gen Gaudi}}
 
@@ -329,7 +329,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.2_BMC_OOB_Management_Protocol_for_First-gen_Gaudi_V0.1.tex', 'Intel Gaudi 1.21.2 First-gen Gaudi and BMC OOB Management Protocol',
+    (master_doc, 'Intel_Gaudi_1.22.0_BMC_OOB_Management_Protocol_for_First-gen_Gaudi_V0.1.tex', 'Intel Gaudi 1.22.0 First-gen Gaudi and BMC OOB Management Protocol',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/HL225_Porting_Guide_Gaudi2/Gaudi2_HL225_Porting_Guide.rst b/source/HL225_Porting_Guide_Gaudi2/Gaudi2_HL225_Porting_Guide.rst
index 7b47faf..60c35ef 100644
--- a/source/HL225_Porting_Guide_Gaudi2/Gaudi2_HL225_Porting_Guide.rst
+++ b/source/HL225_Porting_Guide_Gaudi2/Gaudi2_HL225_Porting_Guide.rst
@@ -501,7 +501,7 @@
 ----------------
 
 This section details the steps required to update the CPLD.
-All required .vme files can be found in the `Intel Gaudi vault <https://vault.habana.ai/artifactory/gaudi-odm/1.21.4/gaudi2/system/gaudi2-system-1.21.4-3.tgz>`_ and `Intel RDC <https://www.intel.com/content/www/us/en/secure/content-details/856022/gaudi-odm-sw.html?DocID=856022>`_.
+All required .vme files can be found in the `Intel Gaudi vault <https://vault.habana.ai/artifactory/gaudi-odm/1.22.0/gaudi2/system/gaudi2-system-1.22.0-740.tgz>`_ and `Intel RDC <https://www.intel.com/content/www/us/en/secure/content-details/856022/gaudi-odm-sw.html?DocID=856022>`_.
 For Lattice Programming, convert the .vme files to SVF file type.
 
 1. Before updating your secured CPLD, check which CFG is in use by reading the "FPGA status register" (address: 0x030) from the CPLD, bits 0-1:
@@ -1144,7 +1144,7 @@
 
 - :file:`gaudi2-fw_upd_open.itb` - enables the FW loader protocol
 
-These files are part of the OAM FW release package and can be found in the `Intel Gaudi vault <https://vault.habana.ai/artifactory/gaudi-odm/1.21.4/gaudi2/gaudi2-sec-fit/gaudi2-sec-fit-fw-61.0.0.tgz>`_ and `Intel RDC <https://www.intel.com/content/www/us/en/secure/content-details/856022/gaudi-odm-sw.html?DocID=856022>`_.
+These files are part of the OAM FW release package and can be found in the `Intel Gaudi vault <https://vault.habana.ai/artifactory/gaudi-odm/1.22.0/gaudi2/gaudi2-sec-fit/gaudi2-sec-fit-fw-61.0.0.tgz>`_ and `Intel RDC <https://www.intel.com/content/www/us/en/secure/content-details/856022/gaudi-odm-sw.html?DocID=856022>`_.
 
 To disable the FW loader protocol:
 
diff --git a/source/HL225_Porting_Guide_Gaudi2/conf.py b/source/HL225_Porting_Guide_Gaudi2/conf.py
index 7d4e008..86e6969 100644
--- a/source/HL225_Porting_Guide_Gaudi2/conf.py
+++ b/source/HL225_Porting_Guide_Gaudi2/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -134,7 +134,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/HL225_Porting_Guide_Gaudi2/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL225_Porting_Guide_Gaudi2/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -158,7 +158,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/HL225_Porting_Guide_Gaudi2/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL225_Porting_Guide_Gaudi2/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -179,7 +179,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             HL225 Porting User Guide}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/HL225_Porting_Guide_Gaudi2/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL225_Porting_Guide_Gaudi2/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_HL225_Porting_Guide_V0.1.tex', 'Intel Gaudi 1.21.4 HL225 Porting User Guide',
+    (master_doc, 'Intel_Gaudi_1.22.0_HL225_Porting_Guide_V0.1.tex', 'Intel Gaudi 1.22.0 HL225 Porting User Guide',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/HL325_Porting_Guide_Gaudi3/conf.py b/source/HL325_Porting_Guide_Gaudi3/conf.py
index fff2e07..9ffb705 100644
--- a/source/HL325_Porting_Guide_Gaudi3/conf.py
+++ b/source/HL325_Porting_Guide_Gaudi3/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -139,7 +139,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/HL325_Porting_Guide_Gaudi3/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL325_Porting_Guide_Gaudi3/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -163,7 +163,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/HL325_Porting_Guide_Gaudi3/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL325_Porting_Guide_Gaudi3/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -184,7 +184,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -246,7 +246,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             HL325 Porting User Guide}}
 
@@ -258,7 +258,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/HL325_Porting_Guide_Gaudi3/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL325_Porting_Guide_Gaudi3/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -333,7 +333,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_HL325_Porting_Guide_V0.1.tex', 'Intel Gaudi 1.21.4 HL325 Porting User Guide',
+    (master_doc, 'Intel_Gaudi_1.22.0_HL325_Porting_Guide_V0.1.tex', 'Intel Gaudi 1.22.0 HL325 Porting User Guide',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/HL338_ODM_System_Tests/HL338_ODM_System_Tests.rst b/source/HL338_ODM_System_Tests/HL338_ODM_System_Tests.rst
index f7df10b..3d4f403 100644
--- a/source/HL338_ODM_System_Tests/HL338_ODM_System_Tests.rst
+++ b/source/HL338_ODM_System_Tests/HL338_ODM_System_Tests.rst
@@ -145,11 +145,11 @@
 | Criteria**      |                                                                                  |
 +-----------------+----------------------------------------------------------------------------------+
 | **Reference**   | `System Management Interface Tool User Guide (hl-smi Tool)                       |
-|                 | <https://docs.habana.ai/en/v1.21.4/Management_and_                               |
+|                 | <https://docs.habana.ai/en/v1.22.0/Management_and_                               |
 |                 | Monitoring/Embedded_System_Tools_Guide/System_Management_Interface_Tool.html>`_  |
 |                 |                                                                                  |
 |                 | `Platform Upgrade and Full System Installation                                   |
-|                 | <https://docs.habana.ai/en/v1.21.4/Installation_Guide/Platform_                  |
+|                 | <https://docs.habana.ai/en/v1.22.0/Installation_Guide/Platform_                  |
 |                 | Upgrade_and_Unboxing.html>`_                                                     |
 +-----------------+----------------------------------------------------------------------------------+
 | **Test          | If the test fails, perform the following:                                        |
@@ -205,7 +205,7 @@
 
 The hl_qual tool tests qualify the usage and integration of Gaudi hardware platforms in your server design.
 
-For further information, refer to `Qualification Tool (hl_qual) <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/index.html>`_ documentation.
+For further information, refer to `Qualification Tool (hl_qual) <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/index.html>`_ documentation.
 
 It is recommended to run the tests in the same order as described below.
 
@@ -270,7 +270,7 @@
 PCI Test (Serial Mode)
 ----------------------
 
-Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
+Refer to `PCI Bandwidth Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#pci-bandwidth-test-design-considerations-and-requirements>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -306,7 +306,7 @@
 Memory Bandwidth Test
 ---------------------
 
-Refer to `Memory Bandwidth Plugin <https://docs.habana.ai/en/1.21.4/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#gaudi-memory-bandwith-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Memory Bandwidth Plugin <https://docs.habana.ai/en/1.22.0/Management_and_Monitoring/Qualification_Library/Bandwidth_Tests_Plugin.html#gaudi-memory-bandwith-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -327,7 +327,7 @@
 SER Test for Serdes
 -------------------
 
-Refer to `SER Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-ser-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `SER Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#gaudi2-ser-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -363,7 +363,7 @@
 Serdes Base – Pairs Test
 ------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -394,7 +394,7 @@
 Serdes Base – All-Reduce Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
 
@@ -427,7 +427,7 @@
 Serdes Base – All-Gather Test
 -----------------------------
 
-Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Serdes Base Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Connectivity_Serdes_Tests_Plugin.html#serdes-base-test-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -458,7 +458,7 @@
 Functional Test 2
 -----------------
 
-Refer to `Functional Test 2 Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#functional-test-2-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Functional Test 2 Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#functional-test-2-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -512,7 +512,7 @@
 Power Stress
 ------------
 
-Refer to `Power Stress Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#gaudi2-power-stress-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `Power Stress Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#gaudi2-power-stress-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -545,7 +545,7 @@
 EDP
 ---
 
-Refer to `EDP Stress Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#edp-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `EDP Stress Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Power_Stress_Tests_Plugin.html#edp-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
@@ -570,7 +570,7 @@
 ResNet-50 and TrainingApp
 --------------------------
 
-Refer to `ResNet-50 Training Stress Test Plugin <https://docs.habana.ai/en/v1.21.4/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#resnet-50-training-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
+Refer to `ResNet-50 Training Stress Test Plugin <https://docs.habana.ai/en/v1.22.0/Management_and_Monitoring/Qualification_Library/Functional_Tests_Plugin.html#resnet-50-training-stress-test-plugin-design-consideration-and-responsibilities>`_ for more details.
 
 
 .. tabularcolumns:: |p{4cm}|p{12cm}|
diff --git a/source/HL338_ODM_System_Tests/conf.py b/source/HL338_ODM_System_Tests/conf.py
index c150ffe..0b29acb 100644
--- a/source/HL338_ODM_System_Tests/conf.py
+++ b/source/HL338_ODM_System_Tests/conf.py
@@ -31,7 +31,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -135,7 +135,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/HL338_ODM_System_Tests/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL338_ODM_System_Tests/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -159,7 +159,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/HL338_ODM_System_Tests/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL338_ODM_System_Tests/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -180,7 +180,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sabuakel/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -241,7 +241,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             ODM System Tests for Gaudi 3 PCIe Card}}
 
@@ -253,7 +253,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/HL338_ODM_System_Tests/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sabuakel/readthedocs/NDA_Docs/source/HL338_ODM_System_Tests/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -328,7 +328,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_ODM_System_Tests_for_Gaudi3_PCIe_Card_V0.1.tex', 'Intel Gaudi 1.21.4 Gaudi 3 PCIe Card ODM System Tests',
+    (master_doc, 'Intel_Gaudi_1.22.0_ODM_System_Tests_for_Gaudi3_PCIe_Card_V0.1.tex', 'Intel Gaudi 1.22.0 Gaudi 3 PCIe Card ODM System Tests',
      'Revision 0.1', 'report')
 
 ]
\ No newline at end of file
diff --git a/source/RAS_APIs_Gaudi3/conf.py b/source/RAS_APIs_Gaudi3/conf.py
index ffb44d8..3bb8ec5 100644
--- a/source/RAS_APIs_Gaudi3/conf.py
+++ b/source/RAS_APIs_Gaudi3/conf.py
@@ -32,7 +32,7 @@
 
 
 # The full version, including alpha/beta/rc tags
-release = '1.21.4'
+release = '1.22.0'
 
 
 
@@ -154,7 +154,7 @@
         \fancyfoot[C]{Confidential and Proprietary Information}
         \fancyfoot[L]{Habana Labs}
         \fancyhead[L]{{\py@HeaderFamily \@title}} % here's the change
-            \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
+            \rhead{\includegraphics[width=1cm]{/home/sakell/readthedocs/NDA_Docs/source/Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
         \renewcommand{\headrulewidth}{0.4pt}
         \renewcommand{\footrulewidth}{0.4pt}
         }
@@ -178,7 +178,7 @@
             \fancyfoot[C]{Confidential and Proprietary Information}
             \fancyfoot[L]{Habana Labs}
             \fancyhead[L]{}
-                \rhead{\includegraphics[width=1cm]{/home/auvarova/readthedocs/NDA_Docs/source/Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
+                \rhead{\includegraphics[width=1cm]{/home/sakell/readthedocs/NDA_Docs/source/Reliability_Availability_and_Serviceability/Intel_gaudi_logo_small.png}}
         }
 
         \setlength{\headsep}{0.5in}
@@ -199,7 +199,7 @@
 
         \usepackage{amsmath,amsfonts,amssymb,amsthm}
         \usepackage{graphicx}
-        \graphicspath{ {C:/Users/auvarova/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
+        \graphicspath{ {C:/Users/sakell/Documents/GitHub/SynapseAI_Gaudi/source/latex_images} }
 
 
         %%% reduce spaces for Table of contents, and figures
@@ -260,7 +260,7 @@
 
 
             \vspace*{40mm} %%% * is used to give space from top
-            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.21.4\\
+            \textbf{\Huge\color{Habana_blue}{Intel Gaudi 1.22.0\\
             \vspace*{10mm}
             Reliability, Availability, and\\
             Serviceability (RAS) REST APIs for Gaudi 3}}
@@ -272,7 +272,7 @@
 
             \vspace{10mm}
             \begin{figure}[!h]
-                \includegraphics[width = 150mm]{/home/auvarova/readthedocs/NDA_Docs/source/Reliability_Availability_and_Serviceability/Intel_gaudi_logo_big.png}
+                \includegraphics[width = 150mm]{/home/sakell/readthedocs/NDA_Docs/source/Reliability_Availability_and_Serviceability/Intel_gaudi_logo_big.png}
                 \centering
             \end{figure}
 
@@ -347,7 +347,7 @@
 # (source start file, target name, title,
 #  author, documentclass [howto, manual, or own class]).
 latex_documents = [
-    (master_doc, 'Intel_Gaudi_1.21.4_RAS_REST_APIs_for_Gaudi3_V0.9.tex', 'Intel Gaudi 1.21.4 Reliability, Availability, and Serviceability (RAS) REST APIs for Gaudi 3',
+    (master_doc, 'Intel_Gaudi_1.22.0_RAS_REST_APIs_for_Gaudi3_V0.9.tex', 'Intel Gaudi 1.22.0 Reliability, Availability, and Serviceability (RAS) REST APIs for Gaudi 3',
      'Revision 0.9', 'report')
 
 ]
\ No newline at end of file
